// antd-token-previewer/lib/component-demos/treeSelect/disabled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _react = _interopRequireWildcard(require("react"));

var _antd = require("antd");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var TreeNode = _antd.TreeSelect.TreeNode,
    TreeSelect = _antd.TreeSelect._InternalPanelDoNotUseOrYouWillBeFired;

var Demo = function Demo() {
  var _useState = (0, _react.useState)(undefined),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      value = _useState2[0],
      setValue = _useState2[1];

  var onChange = function onChange() {
    setValue(value);
  };

  return /*#__PURE__*/_react.default.createElement(TreeSelect, {
    disabled: true,
    showSearch: true,
    style: {
      width: '100%'
    },
    value: value,
    dropdownStyle: {
      maxHeight: 400,
      overflow: 'auto'
    },
    placeholder: "Please select",
    allowClear: true,
    treeDefaultExpandAll: true,
    onChange: onChange
  }, /*#__PURE__*/_react.default.createElement(TreeNode, {
    value: "parent 1",
    title: "parent 1"
  }, /*#__PURE__*/_react.default.createElement(TreeNode, {
    value: "parent 1-0",
    title: "parent 1-0"
  }, /*#__PURE__*/_react.default.createElement(TreeNode, {
    value: "leaf1",
    title: "leaf1"
  }), /*#__PURE__*/_react.default.createElement(TreeNode, {
    value: "leaf2",
    title: "leaf2"
  })), /*#__PURE__*/_react.default.createElement(TreeNode, {
    value: "parent 1-1",
    title: "parent 1-1"
  }, /*#__PURE__*/_react.default.createElement(TreeNode, {
    value: "leaf3",
    title: 'leaf3'
  }))));
};

var componentDemo = {
  demo: /*#__PURE__*/_react.default.createElement(Demo, null),
  tokens: ['colorTextDisabled', 'colorBgContainerDisabled'],
  key: 'disabled'
};
var _default = componentDemo;
exports.default = _default;
