// @antv/xflow-core/es/index.js
// IOC
import * as ManaSyringe from 'mana-syringe';
export { ManaSyringe };
/** XFlow 基础 interface */
export { NsGraph } from './interface';
/*******************************************************
 * 核心组件：
 * 1. Application：XFlow
 * 2. Application Extension：XFlowAppExtensionModule
 * 3. Graph：XFlowCanvas
 *****************************************************/
export { 
// 组件
XFlow, XFlowCanvas, XFlowAppExtensionModule, 
// app context
XFlowAppProvider, XFlowAppContext, useXFlowApp, 
// config provider context：使用全局Config
useXflowPrefixCls, XFlowConfigProviderContext, 
// extension context: 注册扩展
ExtensionRegistryContext, useExtensionRegistry, } from './xflow-main';
/** widget：extension  */
export {} from './xflow-main';
/** graphProvider：注入Graph时 需要 */
export { IGraphProvider, createGraphConfig, GraphConfig } from './xflow-main';
/** app：用于extension扩展*/
export { IApplication, IApplicationContribution, } from './xflow-main';
/** Command Service */
export { ICommandHandler, IGraphCommandService, ICommandContextProvider, IGraphCommandContribution, IGraphCommandFactory, GraphCommandRegistry, commandRegistryModule, } from './command';
/** Command 常量 */
export { XFlowNodeCommands, XFlowEdgeCommands, XFlowGroupCommands, XFlowGraphCommands, XFlowModelCommands, 
/** 创建 Command hook config */
createCmdConfig, } from './command-contributions';
/** React Node Context */
export { AppContext, useAppContext, getNodeReactComponent } from './command-contributions';
/*******************************************************
 *  XFlow Hooks： 钩子函数
 *****************************************************/
export { 
// 扩展Hooks
IHookService, IHookContribution, 
// 创建 React config hook
createHookConfig, } from './hooks';
/*******************************************************
 *  Model Service：全局状态
 *****************************************************/
export { MODELS, IModelService, IModelContribution, createModelServiceConfig, } from './model-service';
export { RxModel, NsModel } from './common/rx-model';
export { useModel, createComponentModel, useModelAsync, useIsMountedRef, } from './common/rx-model-hook';
/*******************************************************
 *  Toolbar：工具栏
 *****************************************************/
/** Toolbar 配置 */
export { 
// component
ToolbarRegistry, 
// ioc 扩展
IToolbarService, IToolbarContribution, } from './toolbar';
/*******************************************************
 *  Menu：菜单
 *****************************************************/
/** Menu 配置 */
export { 
// 扩展 Service
IMenuService, IMenuContribution, 
// menu item
MenuItemType, 
// Component
MenuRegistry, } from './menu';
/*******************************************************
 *  KeyBindings：快捷键
 *****************************************************/
/** KeyBindings 配置 */
export { KeyBindings, createKeybindingConfig, KeybindingConfig, IKeyBindingContribution, } from './keybinding';
/*******************************************************
 *  UTILS：工具方法
 *****************************************************/
/** utils：dispose */
export { Disposable, DisposableCollection } from './common/disposable';
/** utils：defer 延迟 */
export { Deferred } from './common/deferred';
/** utils：延迟 */
export { delay } from './common/delay';
/** utils：uuid */
export { uuidv4 } from './common/uuid';
/** utils：绝对定位 */
export { usePositionStyle } from './common/position';
export {} from './common/types';
/** utils：insertCss */
export { insertCss, isReactComponent } from './common/utils';
/*******************************************************
 * Icon：Antd Icon
 *****************************************************/
/** ICON */
export { IconStore } from './antd-icons';
import * as XFlowConstants_1 from './constants';
export { XFlowConstants_1 as XFlowConstants };
