// @antv/l7-core/lib/services/debug/DebugService.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _typeof2 = _interopRequireDefault(require("@babel/runtime/helpers/typeof"));
var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _l7Utils = require("@antv/l7-utils");
var _eventemitter = require("eventemitter3");
var _inversify = require("inversify");
var _dec, _class;
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }
function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
var DebugService = (_dec = (0, _inversify.injectable)(), _dec(_class = /*#__PURE__*/function (_EventEmitter) {
  (0, _inherits2.default)(DebugService, _EventEmitter);
  var _super = _createSuper(DebugService);
  function DebugService() {
    var _this;
    (0, _classCallCheck2.default)(this, DebugService);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "renderMap", new Map());
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "enable", false);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "renderEnable", false);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "cacheLogs", {});
    return _this;
  }
  (0, _createClass2.default)(DebugService, [{
    key: "setEnable",
    value: function setEnable(flag) {
      this.enable = !!flag;
    }
  }, {
    key: "log",
    value: function log(key, values) {
      var _this2 = this;
      if (!this.enable) {
        return;
      }
      var keys = key.split('.'); // [12, init, layerInitStart]
      var parent = null;
      keys.forEach(function (k, i) {
        if (parent !== null) {
          if (!parent[k]) {
            parent[k] = {};
          }
          if (i !== keys.length - 1) {
            parent = parent[k];
          }
        } else {
          if (!_this2.cacheLogs[k]) {
            _this2.cacheLogs[k] = {};
          }
          if (i !== keys.length - 1) {
            parent = _this2.cacheLogs[k];
          }
        }
        if (i === keys.length - 1) {
          parent[k] = (0, _objectSpread2.default)((0, _objectSpread2.default)({
            time: Date.now()
          }, parent[k]), values);
        }
      });
    }
  }, {
    key: "getLog",
    value: function getLog(key) {
      var _this3 = this;
      switch ((0, _typeof2.default)(key)) {
        case 'string':
          return this.cacheLogs[key];
        case 'object':
          return key.map(function (k) {
            return _this3.cacheLogs[k];
          }).filter(function (o) {
            return o !== undefined;
          });
        case 'undefined':
          return this.cacheLogs;
      }
    }

    /**
     * 删除日志
     * @param key
     */
  }, {
    key: "removeLog",
    value: function removeLog(key) {
      delete this.cacheLogs[key];
    }
  }, {
    key: "generateRenderUid",
    value: function generateRenderUid() {
      if (this.renderEnable) {
        return (0, _l7Utils.guid)();
      } else {
        return '';
      }
    }
  }, {
    key: "renderDebug",
    value: function renderDebug(enable) {
      this.renderEnable = enable;
    }
  }, {
    key: "renderStart",
    value: function renderStart(id) {
      if (!this.renderEnable || !this.enable) {
        return;
      }
      var cacheRenderInfo = this.renderMap.get(id) || {};
      this.renderMap.set(id, (0, _objectSpread2.default)((0, _objectSpread2.default)({}, cacheRenderInfo), {}, {
        renderUid: id,
        renderStart: Date.now()
      }));
    }
  }, {
    key: "renderEnd",
    value: function renderEnd(id) {
      if (!this.renderEnable || !this.enable) {
        return;
      }
      var cacheRenderInfo = this.renderMap.get(id);
      if (cacheRenderInfo) {
        var renderStart = cacheRenderInfo.renderStart;
        var renderEnd = Date.now();
        this.emit('renderEnd', (0, _objectSpread2.default)((0, _objectSpread2.default)({}, cacheRenderInfo), {}, {
          renderEnd: renderEnd,
          renderDuration: renderEnd - renderStart
        }));
        this.renderMap.delete(id);
      }
    }
  }, {
    key: "destroy",
    value: function destroy() {
      this.cacheLogs = null;
      this.renderMap.clear();
    }
  }]);
  return DebugService;
}(_eventemitter.EventEmitter)) || _class);
exports.default = DebugService;
