// @antv/l7plot/dist/lib/plots/hexbin/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Hexbin = void 0;
var tslib_1 = require("tslib");
var util_1 = require("@antv/util");
var plot_1 = require("../../core/plot");
var constants_1 = require("./constants");
var hexbin_layer_1 = require("../../layers/hexbin-layer");
var layer_group_1 = require("../../core/layer/layer-group");
var Hexbin = /** @class */ (function (_super) {
    tslib_1.__extends(Hexbin, _super);
    function Hexbin() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /**
         * 地图类型
         */
        _this.type = plot_1.Plot.PlotType.Hexbin;
        return _this;
    }
    /**
     * 获取默认配置
     */
    Hexbin.prototype.getDefaultOptions = function () {
        return Hexbin.DefaultOptions;
    };
    /**
     * 创建图层
     */
    Hexbin.prototype.createLayers = function (source) {
        this.hexbinLayer = new hexbin_layer_1.HexbinLayer(tslib_1.__assign({ source: source }, (0, util_1.pick)(this.options, hexbin_layer_1.HexbinLayer.LayerOptionsKeys)));
        var layerGroup = new layer_group_1.LayerGroup([this.hexbinLayer]);
        if (this.options.label) {
            this.labelLayer = this.createLabelLayer(this.source, this.options.label, this.options);
            layerGroup.addLayer(this.labelLayer);
        }
        return layerGroup;
    };
    /**
     * 更新图层
     */
    Hexbin.prototype.updateLayers = function (options) {
        var heatMapLayerConfig = (0, util_1.pick)(options, hexbin_layer_1.HexbinLayer.LayerOptionsKeys);
        this.hexbinLayer.update(heatMapLayerConfig);
        this.updateLabelLayer(this.source, options.label, this.options, this.labelLayer);
    };
    /**
     * 默认配置项
     */
    Hexbin.DefaultOptions = constants_1.DEFAULT_OPTIONS;
    return Hexbin;
}(plot_1.Plot));
exports.Hexbin = Hexbin;
