// @antv/x6-react-components/lib/color-picker/index.js
"use strict";
/* eslint-disable jsx-a11y/click-events-have-key-events  */
/* eslint-disable jsx-a11y/no-static-element-interactions */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ColorPicker = void 0;
var react_1 = __importDefault(require("react"));
var classnames_1 = __importDefault(require("classnames"));
var antd_1 = require("antd");
var addEventListener_1 = __importDefault(require("rc-util/lib/Dom/addEventListener"));
var react_color_1 = require("react-color");
var ColorPicker = /** @class */ (function (_super) {
    __extends(ColorPicker, _super);
    function ColorPicker(props) {
        var _this = _super.call(this, props) || this;
        _this.onDocumentClick = function (e) {
            var target = e.target;
            if (target === _this.container || _this.container.contains(target)) {
                return;
            }
            _this.setState({ active: false });
            _this.unbindDocEvent();
        };
        _this.handleChange = function (value, event) {
            if (_this.props.onChange) {
                _this.props.onChange(value, event);
            }
            _this.setState({
                active: false,
                color: value.rgb,
            });
            _this.unbindDocEvent();
        };
        _this.handleClick = function (e) {
            e.stopPropagation();
            if (_this.state.active) {
                _this.setState({ active: false });
                _this.unbindDocEvent();
            }
            else {
                _this.setState({ active: true });
                if (!_this.removeDocClickEvent) {
                    _this.removeDocClickEvent = (0, addEventListener_1.default)(document.documentElement, 'click', _this.onDocumentClick).remove;
                }
            }
        };
        _this.refContainer = function (container) {
            _this.container = container;
        };
        _this.state = {
            active: false,
            color: props.color,
        };
        return _this;
    }
    ColorPicker.prototype.componentWillUnmount = function () {
        this.unbindDocEvent();
    };
    ColorPicker.prototype.unbindDocEvent = function () {
        if (this.removeDocClickEvent) {
            this.removeDocClickEvent();
            this.removeDocClickEvent = null;
        }
    };
    ColorPicker.prototype.renderPicker = function () {
        var _a = this.props, prefixCls = _a.prefixCls, disabled = _a.disabled, style = _a.style, props = __rest(_a, ["prefixCls", "disabled", "style"]);
        return (react_1.default.createElement(react_color_1.SketchPicker, __assign({ width: "240px" }, props, { onChange: this.handleChange })));
    };
    ColorPicker.prototype.render = function () {
        var _a;
        var color = this.state.color;
        var _b = this.props, disabled = _b.disabled, overlayProps = _b.overlayProps, style = _b.style;
        var baseCls = this.props.prefixCls + "-color-picker";
        var popoverProps = {};
        if (disabled) {
            popoverProps.visible = false;
            // Support for antd 5.0
            popoverProps.open = false;
        }
        else {
            popoverProps.visible = this.state.active;
            // Support for antd 5.0
            popoverProps.open = this.state.active;
        }
        var colorStr = typeof color === 'string'
            ? color
            : "rgba(" + color.r + "," + color.g + "," + color.b + "," + color.a + ")";
        return (react_1.default.createElement(antd_1.Popover, __assign({ placement: "topLeft" }, overlayProps, popoverProps, { content: this.renderPicker(), overlayClassName: baseCls + "-overlay" }),
            react_1.default.createElement("div", { style: style, ref: this.refContainer, onClick: this.handleClick, className: (0, classnames_1.default)(baseCls, (_a = {},
                    _a[baseCls + "-disabled"] = disabled,
                    _a)) },
                react_1.default.createElement("div", { className: baseCls + "-block", style: { backgroundColor: disabled ? '#c4c4c4' : colorStr } }))));
    };
    return ColorPicker;
}(react_1.default.Component));
exports.ColorPicker = ColorPicker;
(function (ColorPicker) {
    ColorPicker.defaultProps = {
        prefixCls: 'x6',
        color: '#1890FF',
    };
})(ColorPicker = exports.ColorPicker || (exports.ColorPicker = {}));
exports.ColorPicker = ColorPicker;
