// antd-token-previewer/lib/component-demos/alert/alert.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _antd = require("antd");

var Demo = function Demo() {
  return /*#__PURE__*/_react.default.createElement(_antd.Space, {
    direction: 'vertical'
  }, /*#__PURE__*/_react.default.createElement(_antd.Alert, {
    message: "Success Tips",
    type: "success",
    showIcon: true
  }), /*#__PURE__*/_react.default.createElement(_antd.Alert, {
    message: "Informational Notes",
    type: "info",
    showIcon: true
  }), /*#__PURE__*/_react.default.createElement(_antd.Alert, {
    message: "Warning",
    type: "warning",
    showIcon: true,
    closable: true
  }), /*#__PURE__*/_react.default.createElement(_antd.Alert, {
    message: "Error",
    type: "error",
    showIcon: true
  }));
};

var componentDemo = {
  demo: /*#__PURE__*/_react.default.createElement(Demo, null),
  tokens: ['colorIconHover', 'colorIcon', 'colorText'],
  key: 'alert'
};
var _default = componentDemo;
exports.default = _default;
