// antd-token-previewer/es/component-demos/list/list.js
import React from 'react';
import { List, Avatar } from 'antd';
var data = [{
  title: 'Ant Design Title 1'
}, {
  title: 'Ant Design Title 2'
}, {
  title: 'Ant Design Title 3'
}, {
  title: 'Ant Design Title 4'
}];

var Demo = function Demo() {
  return /*#__PURE__*/React.createElement(List, {
    itemLayout: "horizontal",
    dataSource: data,
    renderItem: function renderItem(item) {
      return /*#__PURE__*/React.createElement(List.Item, null, /*#__PURE__*/React.createElement(List.Item.Meta, {
        avatar: /*#__PURE__*/React.createElement(Avatar, {
          src: "https://joeschmoe.io/api/v1/random"
        }),
        title: /*#__PURE__*/React.createElement("a", {
          href: "https://ant.design"
        }, item.title),
        description: "Ant Design, a design language for background applications, is refined by Ant UED Team"
      }));
    }
  });
};

var componentDemo = {
  demo: /*#__PURE__*/React.createElement(Demo, null),
  tokens: [],
  key: 'default'
};
export default componentDemo;
