// @antv/l7-utils/lib/mini-adapter/HTMLElement.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/mini-adapter/HTMLElement.ts
var HTMLElement_exports = {};
__export(HTMLElement_exports, {
  HTMLElement: () => HTMLElement
});
module.exports = __toCommonJS(HTMLElement_exports);
var import_Element = require("./Element");
var Mixin = __toESM(require("./util/mixin"));
function noop() {
}
var HTMLElement = class extends import_Element.Element {
  constructor(tagName = "", level) {
    super();
    this.className = "";
    this.children = [];
    this.focus = noop;
    this.blur = noop;
    this.insertBefore = noop;
    this.appendChild = noop;
    this.removeChild = noop;
    this.remove = noop;
    this.innerHTML = "";
    this.tagName = tagName.toUpperCase();
    Mixin.parentNode(this, level);
    Mixin.style(this);
    Mixin.classList(this);
    Mixin.clientRegion(this);
    Mixin.offsetRegion(this);
    Mixin.scrollRegion(this);
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  HTMLElement
});
