// @antv/l7-utils/lib/mini-adapter/performance.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/mini-adapter/performance.ts
var performance_exports = {};
__export(performance_exports, {
  performance: () => performance
});
module.exports = __toCommonJS(performance_exports);
var performance = {
  mark: (mark) => null,
  clearMeasures: (measure) => null,
  clearMarks: (mark) => null,
  measure: (mark, create, load) => {
    return {
      duration: 0
    };
  },
  now: () => {
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  performance
});
