// @antv/l7-utils/lib/mini-adapter/WebGL.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/mini-adapter/WebGL.ts
var WebGL_exports = {};
__export(WebGL_exports, {
  WebGLRenderingContext: () => WebGLRenderingContext
});
module.exports = __toCommonJS(WebGL_exports);
var WebGLRenderingContext = {
  GCCSO_SHADER_BINARY_FJ: 37472,
  _3DC_XY_AMD: 34810,
  _3DC_X_AMD: 34809,
  ACTIVE_ATTRIBUTES: 35721,
  ACTIVE_ATTRIBUTE_MAX_LENGTH: 35722,
  ACTIVE_PROGRAM_EXT: 33369,
  ACTIVE_TEXTURE: 34016,
  ACTIVE_UNIFORMS: 35718,
  ACTIVE_UNIFORM_MAX_LENGTH: 35719,
  ALIASED_LINE_WIDTH_RANGE: 33902,
  ALIASED_POINT_SIZE_RANGE: 33901,
  ALL_COMPLETED_NV: 34034,
  ALL_SHADER_BITS_EXT: 4294967295,
  ALPHA: 6406,
  ALPHA16F_EXT: 34844,
  ALPHA32F_EXT: 34838,
  ALPHA8_EXT: 32828,
  ALPHA8_OES: 32828,
  ALPHA_BITS: 3413,
  ALPHA_TEST_FUNC_QCOM: 3009,
  ALPHA_TEST_QCOM: 3008,
  ALPHA_TEST_REF_QCOM: 3010,
  ALREADY_SIGNALED_APPLE: 37146,
  ALWAYS: 519,
  AMD_compressed_3DC_texture: 1,
  AMD_compressed_ATC_texture: 1,
  AMD_performance_monitor: 1,
  AMD_program_binary_Z400: 1,
  ANGLE_depth_texture: 1,
  ANGLE_framebuffer_blit: 1,
  ANGLE_framebuffer_multisample: 1,
  ANGLE_instanced_arrays: 1,
  ANGLE_pack_reverse_row_order: 1,
  ANGLE_program_binary: 1,
  ANGLE_texture_compression_dxt3: 1,
  ANGLE_texture_compression_dxt5: 1,
  ANGLE_texture_usage: 1,
  ANGLE_translated_shader_source: 1,
  ANY_SAMPLES_PASSED_CONSERVATIVE_EXT: 36202,
  ANY_SAMPLES_PASSED_EXT: 35887,
  APPLE_copy_texture_levels: 1,
  APPLE_framebuffer_multisample: 1,
  APPLE_rgb_422: 1,
  APPLE_sync: 1,
  APPLE_texture_format_BGRA8888: 1,
  APPLE_texture_max_level: 1,
  ARM_mali_program_binary: 1,
  ARM_mali_shader_binary: 1,
  ARM_rgba8: 1,
  ARRAY_BUFFER: 34962,
  ARRAY_BUFFER_BINDING: 34964,
  ATC_RGBA_EXPLICIT_ALPHA_AMD: 35987,
  ATC_RGBA_INTERPOLATED_ALPHA_AMD: 34798,
  ATC_RGB_AMD: 35986,
  ATTACHED_SHADERS: 35717,
  BACK: 1029,
  BGRA8_EXT: 37793,
  BGRA_EXT: 32993,
  BGRA_IMG: 32993,
  BINNING_CONTROL_HINT_QCOM: 36784,
  BLEND: 3042,
  BLEND_COLOR: 32773,
  BLEND_DST_ALPHA: 32970,
  BLEND_DST_RGB: 32968,
  BLEND_EQUATION: 32777,
  BLEND_EQUATION_ALPHA: 34877,
  BLEND_EQUATION_RGB: 32777,
  BLEND_SRC_ALPHA: 32971,
  BLEND_SRC_RGB: 32969,
  BLUE_BITS: 3412,
  BOOL: 35670,
  BOOL_VEC2: 35671,
  BOOL_VEC3: 35672,
  BOOL_VEC4: 35673,
  BUFFER: 33504,
  BUFFER_ACCESS_OES: 35003,
  BUFFER_MAPPED_OES: 35004,
  BUFFER_MAP_POINTER_OES: 35005,
  BUFFER_OBJECT_EXT: 37201,
  BUFFER_SIZE: 34660,
  BUFFER_USAGE: 34661,
  BYTE: 5120,
  CCW: 2305,
  CLAMP_TO_BORDER_NV: 33069,
  CLAMP_TO_EDGE: 33071,
  COLOR_ATTACHMENT0: 36064,
  COLOR_ATTACHMENT0_NV: 36064,
  COLOR_ATTACHMENT10_NV: 36074,
  COLOR_ATTACHMENT11_NV: 36075,
  COLOR_ATTACHMENT12_NV: 36076,
  COLOR_ATTACHMENT13_NV: 36077,
  COLOR_ATTACHMENT14_NV: 36078,
  COLOR_ATTACHMENT15_NV: 36079,
  COLOR_ATTACHMENT1_NV: 36065,
  COLOR_ATTACHMENT2_NV: 36066,
  COLOR_ATTACHMENT3_NV: 36067,
  COLOR_ATTACHMENT4_NV: 36068,
  COLOR_ATTACHMENT5_NV: 36069,
  COLOR_ATTACHMENT6_NV: 36070,
  COLOR_ATTACHMENT7_NV: 36071,
  COLOR_ATTACHMENT8_NV: 36072,
  COLOR_ATTACHMENT9_NV: 36073,
  COLOR_ATTACHMENT_EXT: 37104,
  COLOR_BUFFER_BIT: 16384,
  COLOR_BUFFER_BIT0_QCOM: 1,
  COLOR_BUFFER_BIT1_QCOM: 2,
  COLOR_BUFFER_BIT2_QCOM: 4,
  COLOR_BUFFER_BIT3_QCOM: 8,
  COLOR_BUFFER_BIT4_QCOM: 16,
  COLOR_BUFFER_BIT5_QCOM: 32,
  COLOR_BUFFER_BIT6_QCOM: 64,
  COLOR_BUFFER_BIT7_QCOM: 128,
  COLOR_CLEAR_VALUE: 3106,
  COLOR_EXT: 6144,
  COLOR_WRITEMASK: 3107,
  COMPARE_REF_TO_TEXTURE_EXT: 34894,
  COMPILE_STATUS: 35713,
  COMPRESSED_RGBA_ASTC_10x10_KHR: 37819,
  COMPRESSED_RGBA_ASTC_10x5_KHR: 37816,
  COMPRESSED_RGBA_ASTC_10x6_KHR: 37817,
  COMPRESSED_RGBA_ASTC_10x8_KHR: 37818,
  COMPRESSED_RGBA_ASTC_12x10_KHR: 37820,
  COMPRESSED_RGBA_ASTC_12x12_KHR: 37821,
  COMPRESSED_RGBA_ASTC_4x4_KHR: 37808,
  COMPRESSED_RGBA_ASTC_5x4_KHR: 37809,
  COMPRESSED_RGBA_ASTC_5x5_KHR: 37810,
  COMPRESSED_RGBA_ASTC_6x5_KHR: 37811,
  COMPRESSED_RGBA_ASTC_6x6_KHR: 37812,
  COMPRESSED_RGBA_ASTC_8x5_KHR: 37813,
  COMPRESSED_RGBA_ASTC_8x6_KHR: 37814,
  COMPRESSED_RGBA_ASTC_8x8_KHR: 37815,
  COMPRESSED_RGBA_PVRTC_2BPPV1_IMG: 35843,
  COMPRESSED_RGBA_PVRTC_2BPPV2_IMG: 37175,
  COMPRESSED_RGBA_PVRTC_4BPPV1_IMG: 35842,
  COMPRESSED_RGBA_PVRTC_4BPPV2_IMG: 37176,
  COMPRESSED_RGBA_S3TC_DXT1_EXT: 33777,
  COMPRESSED_RGBA_S3TC_DXT3_ANGLE: 33778,
  COMPRESSED_RGBA_S3TC_DXT5_ANGLE: 33779,
  COMPRESSED_RGB_PVRTC_2BPPV1_IMG: 35841,
  COMPRESSED_RGB_PVRTC_4BPPV1_IMG: 35840,
  COMPRESSED_RGB_S3TC_DXT1_EXT: 33776,
  COMPRESSED_SRGB8_ALPHA8_ASTC_10x10_KHR: 37851,
  COMPRESSED_SRGB8_ALPHA8_ASTC_10x5_KHR: 37848,
  COMPRESSED_SRGB8_ALPHA8_ASTC_10x6_KHR: 37849,
  COMPRESSED_SRGB8_ALPHA8_ASTC_10x8_KHR: 37850,
  COMPRESSED_SRGB8_ALPHA8_ASTC_12x10_KHR: 37852,
  COMPRESSED_SRGB8_ALPHA8_ASTC_12x12_KHR: 37853,
  COMPRESSED_SRGB8_ALPHA8_ASTC_4x4_KHR: 37840,
  COMPRESSED_SRGB8_ALPHA8_ASTC_5x4_KHR: 37841,
  COMPRESSED_SRGB8_ALPHA8_ASTC_5x5_KHR: 37842,
  COMPRESSED_SRGB8_ALPHA8_ASTC_6x5_KHR: 37843,
  COMPRESSED_SRGB8_ALPHA8_ASTC_6x6_KHR: 37844,
  COMPRESSED_SRGB8_ALPHA8_ASTC_8x5_KHR: 37845,
  COMPRESSED_SRGB8_ALPHA8_ASTC_8x6_KHR: 37846,
  COMPRESSED_SRGB8_ALPHA8_ASTC_8x8_KHR: 37847,
  COMPRESSED_SRGB_ALPHA_S3TC_DXT1_NV: 35917,
  COMPRESSED_SRGB_ALPHA_S3TC_DXT3_NV: 35918,
  COMPRESSED_SRGB_ALPHA_S3TC_DXT5_NV: 35919,
  COMPRESSED_SRGB_S3TC_DXT1_NV: 35916,
  COMPRESSED_TEXTURE_FORMATS: 34467,
  CONDITION_SATISFIED_APPLE: 37148,
  CONSTANT_ALPHA: 32771,
  CONSTANT_COLOR: 32769,
  CONTEXT_FLAG_DEBUG_BIT: 2,
  CONTEXT_ROBUST_ACCESS_EXT: 37107,
  COUNTER_RANGE_AMD: 35777,
  COUNTER_TYPE_AMD: 35776,
  COVERAGE_ALL_FRAGMENTS_NV: 36565,
  COVERAGE_ATTACHMENT_NV: 36562,
  COVERAGE_AUTOMATIC_NV: 36567,
  COVERAGE_BUFFERS_NV: 36563,
  COVERAGE_BUFFER_BIT_NV: 32768,
  COVERAGE_COMPONENT4_NV: 36561,
  COVERAGE_COMPONENT_NV: 36560,
  COVERAGE_EDGE_FRAGMENTS_NV: 36566,
  COVERAGE_SAMPLES_NV: 36564,
  CPU_OPTIMIZED_QCOM: 36785,
  CULL_FACE: 2884,
  CULL_FACE_MODE: 2885,
  CURRENT_PROGRAM: 35725,
  CURRENT_QUERY_EXT: 34917,
  CURRENT_VERTEX_ATTRIB: 34342,
  CW: 2304,
  DEBUG_CALLBACK_FUNCTION: 33348,
  DEBUG_CALLBACK_USER_PARAM: 33349,
  DEBUG_GROUP_STACK_DEPTH: 33389,
  DEBUG_LOGGED_MESSAGES: 37189,
  DEBUG_NEXT_LOGGED_MESSAGE_LENGTH: 33347,
  DEBUG_OUTPUT: 37600,
  DEBUG_OUTPUT_SYNCHRONOUS: 33346,
  DEBUG_SEVERITY_HIGH: 37190,
  DEBUG_SEVERITY_LOW: 37192,
  DEBUG_SEVERITY_MEDIUM: 37191,
  DEBUG_SEVERITY_NOTIFICATION: 33387,
  DEBUG_SOURCE_API: 33350,
  DEBUG_SOURCE_APPLICATION: 33354,
  DEBUG_SOURCE_OTHER: 33355,
  DEBUG_SOURCE_SHADER_COMPILER: 33352,
  DEBUG_SOURCE_THIRD_PARTY: 33353,
  DEBUG_SOURCE_WINDOW_SYSTEM: 33351,
  DEBUG_TYPE_DEPRECATED_BEHAVIOR: 33357,
  DEBUG_TYPE_ERROR: 33356,
  DEBUG_TYPE_MARKER: 33384,
  DEBUG_TYPE_OTHER: 33361,
  DEBUG_TYPE_PERFORMANCE: 33360,
  DEBUG_TYPE_POP_GROUP: 33386,
  DEBUG_TYPE_PORTABILITY: 33359,
  DEBUG_TYPE_PUSH_GROUP: 33385,
  DEBUG_TYPE_UNDEFINED_BEHAVIOR: 33358,
  DECR: 7683,
  DECR_WRAP: 34056,
  DELETE_STATUS: 35712,
  DEPTH24_STENCIL8_OES: 35056,
  DEPTH_ATTACHMENT: 36096,
  DEPTH_STENCIL_ATTACHMENT: 33306,
  DEPTH_BITS: 3414,
  DEPTH_BUFFER_BIT: 256,
  DEPTH_BUFFER_BIT0_QCOM: 256,
  DEPTH_BUFFER_BIT1_QCOM: 512,
  DEPTH_BUFFER_BIT2_QCOM: 1024,
  DEPTH_BUFFER_BIT3_QCOM: 2048,
  DEPTH_BUFFER_BIT4_QCOM: 4096,
  DEPTH_BUFFER_BIT5_QCOM: 8192,
  DEPTH_BUFFER_BIT6_QCOM: 16384,
  DEPTH_BUFFER_BIT7_QCOM: 32768,
  DEPTH_CLEAR_VALUE: 2931,
  DEPTH_COMPONENT: 6402,
  DEPTH_COMPONENT16: 33189,
  DEPTH_COMPONENT16_NONLINEAR_NV: 36396,
  DEPTH_COMPONENT16_OES: 33189,
  DEPTH_COMPONENT24_OES: 33190,
  DEPTH_COMPONENT32_OES: 33191,
  DEPTH_EXT: 6145,
  DEPTH_FUNC: 2932,
  DEPTH_RANGE: 2928,
  DEPTH_STENCIL: 34041,
  DEPTH_STENCIL_OES: 34041,
  DEPTH_TEST: 2929,
  DEPTH_WRITEMASK: 2930,
  DITHER: 3024,
  DMP_shader_binary: 1,
  DONT_CARE: 4352,
  DRAW_BUFFER0_NV: 34853,
  DRAW_BUFFER10_NV: 34863,
  DRAW_BUFFER11_NV: 34864,
  DRAW_BUFFER12_NV: 34865,
  DRAW_BUFFER13_NV: 34866,
  DRAW_BUFFER14_NV: 34867,
  DRAW_BUFFER15_NV: 34868,
  DRAW_BUFFER1_NV: 34854,
  DRAW_BUFFER2_NV: 34855,
  DRAW_BUFFER3_NV: 34856,
  DRAW_BUFFER4_NV: 34857,
  DRAW_BUFFER5_NV: 34858,
  DRAW_BUFFER6_NV: 34859,
  DRAW_BUFFER7_NV: 34860,
  DRAW_BUFFER8_NV: 34861,
  DRAW_BUFFER9_NV: 34862,
  DRAW_BUFFER_EXT: 3073,
  DRAW_FRAMEBUFFER_ANGLE: 36009,
  DRAW_FRAMEBUFFER_APPLE: 36009,
  DRAW_FRAMEBUFFER_BINDING_ANGLE: 36006,
  DRAW_FRAMEBUFFER_BINDING_APPLE: 36006,
  DRAW_FRAMEBUFFER_BINDING_NV: 36006,
  DRAW_FRAMEBUFFER_NV: 36009,
  DST_ALPHA: 772,
  DST_COLOR: 774,
  DYNAMIC_DRAW: 35048,
  ELEMENT_ARRAY_BUFFER: 34963,
  ELEMENT_ARRAY_BUFFER_BINDING: 34965,
  EQUAL: 514,
  ES_VERSION_2_0: 1,
  ETC1_RGB8_OES: 36196,
  ETC1_SRGB8_NV: 35054,
  EXTENSIONS: 7939,
  EXT_blend_minmax: 1,
  EXT_color_buffer_half_float: 1,
  EXT_debug_label: 1,
  EXT_debug_marker: 1,
  EXT_discard_framebuffer: 1,
  EXT_map_buffer_range: 1,
  EXT_multi_draw_arrays: 1,
  EXT_multisampled_render_to_texture: 1,
  EXT_multiview_draw_buffers: 1,
  EXT_occlusion_query_boolean: 1,
  EXT_read_format_bgra: 1,
  EXT_robustness: 1,
  EXT_sRGB: 1,
  EXT_separate_shader_objects: 1,
  EXT_shader_framebuffer_fetch: 1,
  EXT_shader_texture_lod: 1,
  EXT_shadow_samplers: 1,
  EXT_texture_compression_dxt1: 1,
  EXT_texture_filter_anisotropic: 1,
  EXT_texture_format_BGRA8888: 1,
  EXT_texture_rg: 1,
  EXT_texture_storage: 1,
  EXT_texture_type_2_10_10_10_REV: 1,
  EXT_unpack_subimage: 1,
  FALSE: 0,
  FASTEST: 4353,
  FENCE_CONDITION_NV: 34036,
  FENCE_STATUS_NV: 34035,
  FIXED: 5132,
  FJ_shader_binary_GCCSO: 1,
  FLOAT: 5126,
  FLOAT_MAT2: 35674,
  FLOAT_MAT3: 35675,
  FLOAT_MAT4: 35676,
  FLOAT_VEC2: 35664,
  FLOAT_VEC3: 35665,
  FLOAT_VEC4: 35666,
  FRAGMENT_SHADER: 35632,
  FRAGMENT_SHADER_BIT_EXT: 2,
  FRAGMENT_SHADER_DERIVATIVE_HINT_OES: 35723,
  FRAGMENT_SHADER_DISCARDS_SAMPLES_EXT: 35410,
  FRAMEBUFFER: 36160,
  FRAMEBUFFER_ATTACHMENT_ANGLE: 37795,
  FRAMEBUFFER_ATTACHMENT_COLOR_ENCODING_EXT: 33296,
  FRAMEBUFFER_ATTACHMENT_COMPONENT_TYPE_EXT: 33297,
  FRAMEBUFFER_ATTACHMENT_OBJECT_NAME: 36049,
  FRAMEBUFFER_ATTACHMENT_OBJECT_TYPE: 36048,
  FRAMEBUFFER_ATTACHMENT_TEXTURE_3D_ZOFFSET_OES: 36052,
  FRAMEBUFFER_ATTACHMENT_TEXTURE_CUBE_MAP_FACE: 36051,
  FRAMEBUFFER_ATTACHMENT_TEXTURE_LEVEL: 36050,
  FRAMEBUFFER_ATTACHMENT_TEXTURE_SAMPLES_EXT: 36204,
  FRAMEBUFFER_BINDING: 36006,
  FRAMEBUFFER_COMPLETE: 36053,
  FRAMEBUFFER_INCOMPLETE_ATTACHMENT: 36054,
  FRAMEBUFFER_INCOMPLETE_DIMENSIONS: 36057,
  FRAMEBUFFER_INCOMPLETE_MISSING_ATTACHMENT: 36055,
  FRAMEBUFFER_INCOMPLETE_MULTISAMPLE_ANGLE: 36182,
  FRAMEBUFFER_INCOMPLETE_MULTISAMPLE_APPLE: 36182,
  FRAMEBUFFER_INCOMPLETE_MULTISAMPLE_EXT: 36182,
  FRAMEBUFFER_INCOMPLETE_MULTISAMPLE_IMG: 37172,
  FRAMEBUFFER_INCOMPLETE_MULTISAMPLE_NV: 36182,
  FRAMEBUFFER_UNDEFINED_OES: 33305,
  FRAMEBUFFER_UNSUPPORTED: 36061,
  FRONT: 1028,
  FRONT_AND_BACK: 1032,
  FRONT_FACE: 2886,
  FUNC_ADD: 32774,
  FUNC_REVERSE_SUBTRACT: 32779,
  FUNC_SUBTRACT: 32778,
  GENERATE_MIPMAP_HINT: 33170,
  GEQUAL: 518,
  GPU_OPTIMIZED_QCOM: 36786,
  GREATER: 516,
  GREEN_BITS: 3411,
  GUILTY_CONTEXT_RESET_EXT: 33363,
  HALF_FLOAT_OES: 36193,
  HIGH_FLOAT: 36338,
  HIGH_INT: 36341,
  IMG_multisampled_render_to_texture: 1,
  IMG_program_binary: 1,
  IMG_read_format: 1,
  IMG_shader_binary: 1,
  IMG_texture_compression_pvrtc: 1,
  IMG_texture_compression_pvrtc2: 1,
  IMPLEMENTATION_COLOR_READ_FORMAT: 35739,
  IMPLEMENTATION_COLOR_READ_TYPE: 35738,
  INCR: 7682,
  INCR_WRAP: 34055,
  INFO_LOG_LENGTH: 35716,
  INNOCENT_CONTEXT_RESET_EXT: 33364,
  INT: 5124,
  INT_10_10_10_2_OES: 36343,
  INT_VEC2: 35667,
  INT_VEC3: 35668,
  INT_VEC4: 35669,
  INVALID_ENUM: 1280,
  INVALID_FRAMEBUFFER_OPERATION: 1286,
  INVALID_OPERATION: 1282,
  INVALID_VALUE: 1281,
  INVERT: 5386,
  KEEP: 7680,
  KHR_debug: 1,
  KHR_texture_compression_astc_ldr: 1,
  LEFT: 1030,
  LEQUAL: 515,
  LESS: 513,
  LINEAR: 9729,
  LINEAR_MIPMAP_LINEAR: 9987,
  LINEAR_MIPMAP_NEAREST: 9985,
  LINES: 1,
  LINE_LOOP: 2,
  LINE_STRIP: 3,
  LINE_WIDTH: 2849,
  LINK_STATUS: 35714,
  LOSE_CONTEXT_ON_RESET_EXT: 33362,
  LOW_FLOAT: 36336,
  LOW_INT: 36339,
  LUMINANCE: 6409,
  LUMINANCE16F_EXT: 34846,
  LUMINANCE32F_EXT: 34840,
  LUMINANCE4_ALPHA4_OES: 32835,
  LUMINANCE8_ALPHA8_EXT: 32837,
  LUMINANCE8_ALPHA8_OES: 32837,
  LUMINANCE8_EXT: 32832,
  LUMINANCE8_OES: 32832,
  LUMINANCE_ALPHA: 6410,
  LUMINANCE_ALPHA16F_EXT: 34847,
  LUMINANCE_ALPHA32F_EXT: 34841,
  MALI_PROGRAM_BINARY_ARM: 36705,
  MALI_SHADER_BINARY_ARM: 36704,
  MAP_FLUSH_EXPLICIT_BIT_EXT: 16,
  MAP_INVALIDATE_BUFFER_BIT_EXT: 8,
  MAP_INVALIDATE_RANGE_BIT_EXT: 4,
  MAP_READ_BIT_EXT: 1,
  MAP_UNSYNCHRONIZED_BIT_EXT: 32,
  MAP_WRITE_BIT_EXT: 2,
  MAX_3D_TEXTURE_SIZE_OES: 32883,
  MAX_COLOR_ATTACHMENTS_NV: 36063,
  MAX_COMBINED_TEXTURE_IMAGE_UNITS: 35661,
  MAX_CUBE_MAP_TEXTURE_SIZE: 34076,
  MAX_DEBUG_GROUP_STACK_DEPTH: 33388,
  MAX_DEBUG_LOGGED_MESSAGES: 37188,
  MAX_DEBUG_MESSAGE_LENGTH: 37187,
  MAX_DRAW_BUFFERS_NV: 34852,
  MAX_EXT: 32776,
  MAX_FRAGMENT_UNIFORM_VECTORS: 36349,
  MAX_LABEL_LENGTH: 33512,
  MAX_MULTIVIEW_BUFFERS_EXT: 37106,
  MAX_RENDERBUFFER_SIZE: 34024,
  MAX_SAMPLES_ANGLE: 36183,
  MAX_SAMPLES_APPLE: 36183,
  MAX_SAMPLES_EXT: 36183,
  MAX_SAMPLES_IMG: 37173,
  MAX_SAMPLES_NV: 36183,
  MAX_SERVER_WAIT_TIMEOUT_APPLE: 37137,
  MAX_TEXTURE_IMAGE_UNITS: 34930,
  MAX_TEXTURE_MAX_ANISOTROPY_EXT: 34047,
  MAX_TEXTURE_SIZE: 3379,
  MAX_VARYING_VECTORS: 36348,
  MAX_VERTEX_ATTRIBS: 34921,
  MAX_VERTEX_TEXTURE_IMAGE_UNITS: 35660,
  MAX_VERTEX_UNIFORM_VECTORS: 36347,
  MAX_VIEWPORT_DIMS: 3386,
  MEDIUM_FLOAT: 36337,
  MEDIUM_INT: 36340,
  MIN_EXT: 32775,
  MIRRORED_REPEAT: 33648,
  MULTISAMPLE_BUFFER_BIT0_QCOM: 16777216,
  MULTISAMPLE_BUFFER_BIT1_QCOM: 33554432,
  MULTISAMPLE_BUFFER_BIT2_QCOM: 67108864,
  MULTISAMPLE_BUFFER_BIT3_QCOM: 134217728,
  MULTISAMPLE_BUFFER_BIT4_QCOM: 268435456,
  MULTISAMPLE_BUFFER_BIT5_QCOM: 536870912,
  MULTISAMPLE_BUFFER_BIT6_QCOM: 1073741824,
  MULTISAMPLE_BUFFER_BIT7_QCOM: 2147483648,
  MULTIVIEW_EXT: 37105,
  NEAREST: 9728,
  NEAREST_MIPMAP_LINEAR: 9986,
  NEAREST_MIPMAP_NEAREST: 9984,
  NEVER: 512,
  NICEST: 4354,
  NONE: 0,
  NOTEQUAL: 517,
  NO_ERROR: 0,
  NO_RESET_NOTIFICATION_EXT: 33377,
  NUM_COMPRESSED_TEXTURE_FORMATS: 34466,
  NUM_PROGRAM_BINARY_FORMATS_OES: 34814,
  NUM_SHADER_BINARY_FORMATS: 36345,
  NV_coverage_sample: 1,
  NV_depth_nonlinear: 1,
  NV_draw_buffers: 1,
  NV_draw_instanced: 1,
  NV_fbo_color_attachments: 1,
  NV_fence: 1,
  NV_framebuffer_blit: 1,
  NV_framebuffer_multisample: 1,
  NV_generate_mipmap_sRGB: 1,
  NV_instanced_arrays: 1,
  NV_read_buffer: 1,
  NV_read_buffer_front: 1,
  NV_read_depth: 1,
  NV_read_depth_stencil: 1,
  NV_read_stencil: 1,
  NV_sRGB_formats: 1,
  NV_shadow_samplers_array: 1,
  NV_shadow_samplers_cube: 1,
  NV_texture_border_clamp: 1,
  NV_texture_compression_s3tc_update: 1,
  NV_texture_npot_2D_mipmap: 1,
  OBJECT_TYPE_APPLE: 37138,
  OES_EGL_image: 1,
  OES_EGL_image_external: 1,
  OES_compressed_ETC1_RGB8_texture: 1,
  OES_compressed_paletted_texture: 1,
  OES_depth24: 1,
  OES_depth32: 1,
  OES_depth_texture: 1,
  OES_element_index_uint: 1,
  OES_fbo_render_mipmap: 1,
  OES_fragment_precision_high: 1,
  OES_get_program_binary: 1,
  OES_mapbuffer: 1,
  OES_packed_depth_stencil: 1,
  OES_required_internalformat: 1,
  OES_rgb8_rgba8: 1,
  OES_standard_derivatives: 1,
  OES_stencil1: 1,
  OES_stencil4: 1,
  OES_surfaceless_context: 1,
  OES_texture_3D: 1,
  OES_texture_float: 1,
  OES_texture_float_linear: 1,
  OES_texture_half_float: 1,
  OES_texture_half_float_linear: 1,
  OES_texture_npot: 1,
  OES_vertex_array_object: 1,
  OES_vertex_half_float: 1,
  OES_vertex_type_10_10_10_2: 1,
  ONE: 1,
  ONE_MINUS_CONSTANT_ALPHA: 32772,
  ONE_MINUS_CONSTANT_COLOR: 32770,
  ONE_MINUS_DST_ALPHA: 773,
  ONE_MINUS_DST_COLOR: 775,
  ONE_MINUS_SRC_ALPHA: 771,
  ONE_MINUS_SRC_COLOR: 769,
  OUT_OF_MEMORY: 1285,
  PACK_ALIGNMENT: 3333,
  PACK_REVERSE_ROW_ORDER_ANGLE: 37796,
  PALETTE4_R5_G6_B5_OES: 35730,
  PALETTE4_RGB5_A1_OES: 35732,
  PALETTE4_RGB8_OES: 35728,
  PALETTE4_RGBA4_OES: 35731,
  PALETTE4_RGBA8_OES: 35729,
  PALETTE8_R5_G6_B5_OES: 35735,
  PALETTE8_RGB5_A1_OES: 35737,
  PALETTE8_RGB8_OES: 35733,
  PALETTE8_RGBA4_OES: 35736,
  PALETTE8_RGBA8_OES: 35734,
  PERCENTAGE_AMD: 35779,
  PERFMON_GLOBAL_MODE_QCOM: 36768,
  PERFMON_RESULT_AMD: 35782,
  PERFMON_RESULT_AVAILABLE_AMD: 35780,
  PERFMON_RESULT_SIZE_AMD: 35781,
  POINTS: 0,
  POLYGON_OFFSET_FACTOR: 32824,
  POLYGON_OFFSET_FILL: 32823,
  POLYGON_OFFSET_UNITS: 10752,
  PROGRAM: 33506,
  PROGRAM_BINARY_ANGLE: 37798,
  PROGRAM_BINARY_FORMATS_OES: 34815,
  PROGRAM_BINARY_LENGTH_OES: 34625,
  PROGRAM_OBJECT_EXT: 35648,
  PROGRAM_PIPELINE_BINDING_EXT: 33370,
  PROGRAM_PIPELINE_OBJECT_EXT: 35407,
  PROGRAM_SEPARABLE_EXT: 33368,
  QCOM_alpha_test: 1,
  QCOM_binning_control: 1,
  QCOM_driver_control: 1,
  QCOM_extended_get: 1,
  QCOM_extended_get2: 1,
  QCOM_perfmon_global_mode: 1,
  QCOM_tiled_rendering: 1,
  QCOM_writeonly_rendering: 1,
  QUERY: 33507,
  QUERY_OBJECT_EXT: 37203,
  QUERY_RESULT_AVAILABLE_EXT: 34919,
  QUERY_RESULT_EXT: 34918,
  R16F_EXT: 33325,
  R32F_EXT: 33326,
  R8_EXT: 33321,
  READ_BUFFER_EXT: 3074,
  READ_BUFFER_NV: 3074,
  READ_FRAMEBUFFER_ANGLE: 36008,
  READ_FRAMEBUFFER_APPLE: 36008,
  READ_FRAMEBUFFER_BINDING_ANGLE: 36010,
  READ_FRAMEBUFFER_BINDING_APPLE: 36010,
  READ_FRAMEBUFFER_BINDING_NV: 36010,
  READ_FRAMEBUFFER_NV: 36008,
  RED_BITS: 3410,
  RED_EXT: 6403,
  RENDERBUFFER: 36161,
  RENDERBUFFER_ALPHA_SIZE: 36179,
  RENDERBUFFER_BINDING: 36007,
  RENDERBUFFER_BLUE_SIZE: 36178,
  RENDERBUFFER_DEPTH_SIZE: 36180,
  RENDERBUFFER_GREEN_SIZE: 36177,
  RENDERBUFFER_HEIGHT: 36163,
  RENDERBUFFER_INTERNAL_FORMAT: 36164,
  RENDERBUFFER_RED_SIZE: 36176,
  RENDERBUFFER_SAMPLES_ANGLE: 36011,
  RENDERBUFFER_SAMPLES_APPLE: 36011,
  RENDERBUFFER_SAMPLES_EXT: 36011,
  RENDERBUFFER_SAMPLES_IMG: 37171,
  RENDERBUFFER_SAMPLES_NV: 36011,
  RENDERBUFFER_STENCIL_SIZE: 36181,
  RENDERBUFFER_WIDTH: 36162,
  RENDERER: 7937,
  RENDER_DIRECT_TO_FRAMEBUFFER_QCOM: 36787,
  REPEAT: 10497,
  REPLACE: 7681,
  REQUIRED_TEXTURE_IMAGE_UNITS_OES: 36200,
  RESET_NOTIFICATION_STRATEGY_EXT: 33366,
  RG16F_EXT: 33327,
  RG32F_EXT: 33328,
  RG8_EXT: 33323,
  RGB: 6407,
  RGB10_A2_EXT: 32857,
  RGB10_EXT: 32850,
  RGB16F_EXT: 34843,
  RGB32F_EXT: 34837,
  RGB565: 36194,
  RGB565_OES: 36194,
  RGB5_A1: 32855,
  RGB5_A1_OES: 32855,
  RGB8_OES: 32849,
  RGBA: 6408,
  RGBA16F_EXT: 34842,
  RGBA32F_EXT: 34836,
  RGBA4: 32854,
  RGBA4_OES: 32854,
  RGBA8_OES: 32856,
  RGB_422_APPLE: 35359,
  RG_EXT: 33319,
  RIGHT: 1031,
  SAMPLER: 33510,
  SAMPLER_2D: 35678,
  SAMPLER_2D_ARRAY_SHADOW_NV: 36292,
  SAMPLER_2D_SHADOW_EXT: 35682,
  SAMPLER_3D_OES: 35679,
  SAMPLER_CUBE: 35680,
  SAMPLER_CUBE_SHADOW_NV: 36293,
  SAMPLER_EXTERNAL_OES: 36198,
  SAMPLES: 32937,
  SAMPLE_ALPHA_TO_COVERAGE: 32926,
  SAMPLE_BUFFERS: 32936,
  SAMPLE_COVERAGE: 32928,
  SAMPLE_COVERAGE_INVERT: 32939,
  SAMPLE_COVERAGE_VALUE: 32938,
  SCISSOR_BOX: 3088,
  SCISSOR_TEST: 3089,
  SGX_BINARY_IMG: 35850,
  SGX_PROGRAM_BINARY_IMG: 37168,
  SHADER: 33505,
  SHADER_BINARY_DMP: 37456,
  SHADER_BINARY_FORMATS: 36344,
  SHADER_BINARY_VIV: 36804,
  SHADER_COMPILER: 36346,
  SHADER_OBJECT_EXT: 35656,
  SHADER_SOURCE_LENGTH: 35720,
  SHADER_TYPE: 35663,
  SHADING_LANGUAGE_VERSION: 35724,
  SHORT: 5122,
  SIGNALED_APPLE: 37145,
  SLUMINANCE8_ALPHA8_NV: 35909,
  SLUMINANCE8_NV: 35911,
  SLUMINANCE_ALPHA_NV: 35908,
  SLUMINANCE_NV: 35910,
  SRC_ALPHA: 770,
  SRC_ALPHA_SATURATE: 776,
  SRC_COLOR: 768,
  SRGB8_ALPHA8_EXT: 35907,
  SRGB8_NV: 35905,
  SRGB_ALPHA_EXT: 35906,
  SRGB_EXT: 35904,
  STACK_OVERFLOW: 1283,
  STACK_UNDERFLOW: 1284,
  STATE_RESTORE: 35804,
  STATIC_DRAW: 35044,
  STENCIL_ATTACHMENT: 36128,
  STENCIL_BACK_FAIL: 34817,
  STENCIL_BACK_FUNC: 34816,
  STENCIL_BACK_PASS_DEPTH_FAIL: 34818,
  STENCIL_BACK_PASS_DEPTH_PASS: 34819,
  STENCIL_BACK_REF: 36003,
  STENCIL_BACK_VALUE_MASK: 36004,
  STENCIL_BACK_WRITEMASK: 36005,
  STENCIL_BITS: 3415,
  STENCIL_BUFFER_BIT: 1024,
  STENCIL_BUFFER_BIT0_QCOM: 65536,
  STENCIL_BUFFER_BIT1_QCOM: 131072,
  STENCIL_BUFFER_BIT2_QCOM: 262144,
  STENCIL_BUFFER_BIT3_QCOM: 524288,
  STENCIL_BUFFER_BIT4_QCOM: 1048576,
  STENCIL_BUFFER_BIT5_QCOM: 2097152,
  STENCIL_BUFFER_BIT6_QCOM: 4194304,
  STENCIL_BUFFER_BIT7_QCOM: 8388608,
  STENCIL_CLEAR_VALUE: 2961,
  STENCIL_EXT: 6146,
  STENCIL_FAIL: 2964,
  STENCIL_FUNC: 2962,
  STENCIL_INDEX1_OES: 36166,
  STENCIL_INDEX4_OES: 36167,
  STENCIL_INDEX: 6401,
  STENCIL_INDEX8: 36168,
  STENCIL_PASS_DEPTH_FAIL: 2965,
  STENCIL_PASS_DEPTH_PASS: 2966,
  STENCIL_REF: 2967,
  STENCIL_TEST: 2960,
  STENCIL_VALUE_MASK: 2963,
  STENCIL_WRITEMASK: 2968,
  STREAM_DRAW: 35040,
  SUBPIXEL_BITS: 3408,
  SYNC_CONDITION_APPLE: 37139,
  SYNC_FENCE_APPLE: 37142,
  SYNC_FLAGS_APPLE: 37141,
  SYNC_FLUSH_COMMANDS_BIT_APPLE: 1,
  SYNC_GPU_COMMANDS_COMPLETE_APPLE: 37143,
  SYNC_OBJECT_APPLE: 35411,
  SYNC_STATUS_APPLE: 37140,
  TEXTURE: 5890,
  TEXTURE0: 33984,
  TEXTURE1: 33985,
  TEXTURE10: 33994,
  TEXTURE11: 33995,
  TEXTURE12: 33996,
  TEXTURE13: 33997,
  TEXTURE14: 33998,
  TEXTURE15: 33999,
  TEXTURE16: 34e3,
  TEXTURE17: 34001,
  TEXTURE18: 34002,
  TEXTURE19: 34003,
  TEXTURE2: 33986,
  TEXTURE20: 34004,
  TEXTURE21: 34005,
  TEXTURE22: 34006,
  TEXTURE23: 34007,
  TEXTURE24: 34008,
  TEXTURE25: 34009,
  TEXTURE26: 34010,
  TEXTURE27: 34011,
  TEXTURE28: 34012,
  TEXTURE29: 34013,
  TEXTURE3: 33987,
  TEXTURE30: 34014,
  TEXTURE31: 34015,
  TEXTURE4: 33988,
  TEXTURE5: 33989,
  TEXTURE6: 33990,
  TEXTURE7: 33991,
  TEXTURE8: 33992,
  TEXTURE9: 33993,
  TEXTURE_2D: 3553,
  TEXTURE_3D_OES: 32879,
  TEXTURE_BINDING_2D: 32873,
  TEXTURE_BINDING_3D_OES: 32874,
  TEXTURE_BINDING_CUBE_MAP: 34068,
  TEXTURE_BINDING_EXTERNAL_OES: 36199,
  TEXTURE_BORDER_COLOR_NV: 4100,
  TEXTURE_COMPARE_FUNC_EXT: 34893,
  TEXTURE_COMPARE_MODE_EXT: 34892,
  TEXTURE_CUBE_MAP: 34067,
  TEXTURE_CUBE_MAP_NEGATIVE_X: 34070,
  TEXTURE_CUBE_MAP_NEGATIVE_Y: 34072,
  TEXTURE_CUBE_MAP_NEGATIVE_Z: 34074,
  TEXTURE_CUBE_MAP_POSITIVE_X: 34069,
  TEXTURE_CUBE_MAP_POSITIVE_Y: 34071,
  TEXTURE_CUBE_MAP_POSITIVE_Z: 34073,
  TEXTURE_DEPTH_QCOM: 35796,
  TEXTURE_EXTERNAL_OES: 36197,
  TEXTURE_FORMAT_QCOM: 35798,
  TEXTURE_HEIGHT_QCOM: 35795,
  TEXTURE_IMAGE_VALID_QCOM: 35800,
  TEXTURE_IMMUTABLE_FORMAT_EXT: 37167,
  TEXTURE_INTERNAL_FORMAT_QCOM: 35797,
  TEXTURE_MAG_FILTER: 10240,
  TEXTURE_MAX_ANISOTROPY_EXT: 34046,
  TEXTURE_MAX_LEVEL_APPLE: 33085,
  TEXTURE_MIN_FILTER: 10241,
  TEXTURE_NUM_LEVELS_QCOM: 35801,
  TEXTURE_OBJECT_VALID_QCOM: 35803,
  TEXTURE_SAMPLES_IMG: 37174,
  TEXTURE_TARGET_QCOM: 35802,
  TEXTURE_TYPE_QCOM: 35799,
  TEXTURE_USAGE_ANGLE: 37794,
  TEXTURE_WIDTH_QCOM: 35794,
  TEXTURE_WRAP_R_OES: 32882,
  TEXTURE_WRAP_S: 10242,
  TEXTURE_WRAP_T: 10243,
  TIMEOUT_EXPIRED_APPLE: 37147,
  /* eslint-disable @typescript-eslint/no-loss-of-precision */
  TIMEOUT_IGNORED_APPLE: 18446744073709552e3,
  TRANSLATED_SHADER_SOURCE_LENGTH_ANGLE: 37792,
  TRIANGLES: 4,
  TRIANGLE_FAN: 6,
  TRIANGLE_STRIP: 5,
  TRUE: 1,
  UNKNOWN_CONTEXT_RESET_EXT: 33365,
  UNPACK_ALIGNMENT: 3317,
  UNPACK_ROW_LENGTH: 3314,
  UNPACK_SKIP_PIXELS: 3316,
  UNPACK_SKIP_ROWS: 3315,
  UNSIGNALED_APPLE: 37144,
  UNSIGNED_BYTE: 5121,
  UNSIGNED_INT: 5125,
  UNSIGNED_INT64_AMD: 35778,
  UNSIGNED_INT_10_10_10_2_OES: 36342,
  UNSIGNED_INT_24_8_OES: 34042,
  UNSIGNED_INT_2_10_10_10_REV_EXT: 33640,
  UNSIGNED_NORMALIZED_EXT: 35863,
  UNSIGNED_SHORT: 5123,
  UNSIGNED_SHORT_1_5_5_5_REV_EXT: 33638,
  UNSIGNED_SHORT_4_4_4_4: 32819,
  UNSIGNED_SHORT_4_4_4_4_REV_EXT: 33637,
  UNSIGNED_SHORT_4_4_4_4_REV_IMG: 33637,
  UNSIGNED_SHORT_5_5_5_1: 32820,
  UNSIGNED_SHORT_5_6_5: 33635,
  UNSIGNED_SHORT_8_8_APPLE: 34234,
  UNSIGNED_SHORT_8_8_REV_APPLE: 34235,
  VALIDATE_STATUS: 35715,
  VENDOR: 7936,
  VERSION: 7938,
  VERTEX_ARRAY_BINDING_OES: 34229,
  VERTEX_ARRAY_OBJECT_EXT: 37204,
  VERTEX_ATTRIB_ARRAY_BUFFER_BINDING: 34975,
  VERTEX_ATTRIB_ARRAY_DIVISOR_ANGLE: 35070,
  VERTEX_ATTRIB_ARRAY_DIVISOR_NV: 35070,
  VERTEX_ATTRIB_ARRAY_ENABLED: 34338,
  VERTEX_ATTRIB_ARRAY_NORMALIZED: 34922,
  VERTEX_ATTRIB_ARRAY_POINTER: 34373,
  VERTEX_ATTRIB_ARRAY_SIZE: 34339,
  VERTEX_ATTRIB_ARRAY_STRIDE: 34340,
  VERTEX_ATTRIB_ARRAY_TYPE: 34341,
  VERTEX_SHADER: 35633,
  VERTEX_SHADER_BIT_EXT: 1,
  VIEWPORT: 2978,
  VIV_shader_binary: 1,
  WAIT_FAILED_APPLE: 37149,
  WRITEONLY_RENDERING_QCOM: 34851,
  WRITE_ONLY_OES: 35001,
  Z400_BINARY_AMD: 34624,
  ZERO: 0,
  RASTERIZER_DISCARD: 35977,
  UNPACK_FLIP_Y_WEBGL: 37440,
  UNPACK_PREMULTIPLY_ALPHA_WEBGL: 37441,
  CONTEXT_LOST_WEBGL: 37442,
  UNPACK_COLORSPACE_CONVERSION_WEBGL: 37443,
  BROWSER_DEFAULT_WEBGL: 37444
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  WebGLRenderingContext
});
