// antd-token-previewer/lib/component-demos/autoComplete/auto-complete.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _react = _interopRequireWildcard(require("react"));

var _antd = require("antd");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var mockVal = function mockVal(str) {
  var repeat = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 1;
  return {
    value: str.repeat(repeat)
  };
};

var Complete = function Complete() {
  var _useState = (0, _react.useState)(''),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      value = _useState2[0],
      setValue = _useState2[1];

  var _useState3 = (0, _react.useState)([]),
      _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
      options = _useState4[0],
      setOptions = _useState4[1];

  var onSearch = function onSearch(searchText) {
    setOptions(!searchText ? [] : [mockVal(searchText), mockVal(searchText, 2), mockVal(searchText, 3)]);
  };

  var onSelect = function onSelect(data) {
    // eslint-disable-next-line no-console
    console.log('onSelect', data);
  };

  var onChange = function onChange(data) {
    setValue(data);
  };

  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, ' ', /*#__PURE__*/_react.default.createElement(_antd.AutoComplete, {
    options: options,
    style: {
      width: 200
    },
    onSelect: onSelect,
    onSearch: onSearch,
    placeholder: "input here"
  }), ' ', /*#__PURE__*/_react.default.createElement("br", null), " ", /*#__PURE__*/_react.default.createElement("br", null), ' ', /*#__PURE__*/_react.default.createElement(_antd.AutoComplete, {
    value: value,
    options: options,
    style: {
      width: 200
    },
    onSelect: onSelect,
    onSearch: onSearch,
    onChange: onChange,
    placeholder: "control mode"
  }), ' ');
};

var Demo = function Demo() {
  return /*#__PURE__*/_react.default.createElement(Complete, null);
};

var componentDemo = {
  demo: /*#__PURE__*/_react.default.createElement(Demo, null),
  tokens: [],
  key: 'autoComplete'
};
var _default = componentDemo;
exports.default = _default;
