// @antv/x6-react-shape/lib/view.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReactShapeView = void 0;
var react_1 = __importDefault(require("react"));
var react_dom_1 = __importDefault(require("react-dom"));
var x6_1 = require("@antv/x6");
var portal_1 = require("./portal");
var wrap_1 = require("./wrap");
var ReactShapeView = /** @class */ (function (_super) {
    __extends(ReactShapeView, _super);
    function ReactShapeView() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ReactShapeView.prototype.init = function () {
        var _this = this;
        _super.prototype.init.call(this);
        this.cell.on('removed', function () {
            portal_1.Portal.disconnect(_this.cell.id);
        });
    };
    ReactShapeView.prototype.getComponentContainer = function () {
        return this.cell.prop('useForeignObject') === false
            ? this.selectors.content
            : this.selectors.foContent;
    };
    ReactShapeView.prototype.confirmUpdate = function (flag) {
        var _this = this;
        var ret = _super.prototype.confirmUpdate.call(this, flag);
        return this.handleAction(ret, ReactShapeView.action, function () {
            if (x6_1.Scheduler) {
                x6_1.Scheduler.scheduleTask(function () {
                    _this.renderReactComponent();
                });
            }
            else {
                _this.renderReactComponent();
            }
        });
    };
    ReactShapeView.prototype.renderReactComponent = function () {
        this.unmountReactComponent();
        var root = this.getComponentContainer();
        var node = this.cell;
        var graph = this.graph;
        if (root) {
            var component = this.graph.hook.getReactComponent(node);
            var elem = react_1.default.createElement(wrap_1.Wrap, { graph: graph, node: node, component: component });
            if (portal_1.Portal.isActive()) {
                portal_1.Portal.connect(this.cell.id, react_dom_1.default.createPortal(elem, root));
            }
            else {
                react_dom_1.default.render(elem, root);
            }
        }
    };
    ReactShapeView.prototype.unmountReactComponent = function () {
        var root = this.getComponentContainer();
        if (root) {
            react_dom_1.default.unmountComponentAtNode(root);
        }
        return root;
    };
    ReactShapeView.prototype.unmount = function () {
        portal_1.Portal.disconnect(this.cell.id);
        this.unmountReactComponent();
        _super.prototype.unmount.call(this);
        return this;
    };
    return ReactShapeView;
}(x6_1.NodeView));
exports.ReactShapeView = ReactShapeView;
(function (ReactShapeView) {
    ReactShapeView.action = 'react';
    ReactShapeView.config({
        bootstrap: [ReactShapeView.action],
        actions: {
            component: ReactShapeView.action,
        },
    });
    x6_1.NodeView.registry.register('react-shape-view', ReactShapeView, true);
})(ReactShapeView = exports.ReactShapeView || (exports.ReactShapeView = {}));
exports.ReactShapeView = ReactShapeView;
