// antd-token-previewer/lib/component-demos/mentions/danger.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _antd = require("antd");

var Option = _antd.Mentions.Option;

function onChange() {}

function onSelect() {}

var Demo = function Demo() {
  return /*#__PURE__*/_react.default.createElement(_antd.Mentions, {
    style: {
      width: '100%'
    },
    onChange: onChange,
    onSelect: onSelect,
    status: 'error',
    defaultValue: "@afc163"
  }, /*#__PURE__*/_react.default.createElement(Option, {
    value: "afc163"
  }, "afc163"), /*#__PURE__*/_react.default.createElement(Option, {
    value: "zombieJ"
  }, "zombieJ"), /*#__PURE__*/_react.default.createElement(Option, {
    value: "yesmeck"
  }, "yesmeck"));
};

var componentDemo = {
  demo: /*#__PURE__*/_react.default.createElement(Demo, null),
  tokens: ['colorError', 'colorErrorOutline', 'colorErrorBorder', 'colorErrorHover'],
  key: 'danger'
};
var _default = componentDemo;
exports.default = _default;
