// @antv/l7-component/lib/control/exportImage.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ExportImage = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));
var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));
var _get2 = _interopRequireDefault(require("@babel/runtime/helpers/get"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _icon = require("../utils/icon");
var _buttonControl = _interopRequireDefault(require("./baseControl/buttonControl"));
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }
function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
var ExportImage = /*#__PURE__*/function (_ButtonControl) {
  (0, _inherits2.default)(ExportImage, _ButtonControl);
  var _super = _createSuper(ExportImage);
  function ExportImage() {
    var _this;
    (0, _classCallCheck2.default)(this, ExportImage);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onClick", /*#__PURE__*/(0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee() {
      var onExport;
      return _regenerator.default.wrap(function _callee$(_context) {
        while (1) switch (_context.prev = _context.next) {
          case 0:
            onExport = _this.controlOption.onExport;
            if (!(onExport === null || onExport === void 0)) {
              _context.next = 5;
              break;
            }
            void 0;
            _context.next = 10;
            break;
          case 5:
            _context.t0 = onExport;
            _context.next = 8;
            return _this.getImage();
          case 8:
            _context.t1 = _context.sent;
            (0, _context.t0)(_context.t1);
          case 10:
          case "end":
            return _context.stop();
        }
      }, _callee);
    })));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "mergeImage", /*#__PURE__*/(0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee2() {
      var _this$mapsService$get, _this$mapsService$get2;
      var imageType,
        _ref3,
        _ref3$width,
        width,
        _ref3$height,
        height,
        canvas,
        context,
        _len2,
        base64List,
        _key2,
        imgList,
        _args2 = arguments;
      return _regenerator.default.wrap(function _callee2$(_context2) {
        while (1) switch (_context2.prev = _context2.next) {
          case 0:
            imageType = _this.controlOption.imageType;
            _ref3 = (_this$mapsService$get = (_this$mapsService$get2 = _this.mapsService.getContainer()) === null || _this$mapsService$get2 === void 0 ? void 0 : _this$mapsService$get2.getBoundingClientRect()) !== null && _this$mapsService$get !== void 0 ? _this$mapsService$get : {}, _ref3$width = _ref3.width, width = _ref3$width === void 0 ? 0 : _ref3$width, _ref3$height = _ref3.height, height = _ref3$height === void 0 ? 0 : _ref3$height;
            canvas = document.createElement('canvas');
            canvas.width = width;
            canvas.height = height;
            context = canvas.getContext('2d');
            for (_len2 = _args2.length, base64List = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
              base64List[_key2] = _args2[_key2];
            }
            _context2.next = 9;
            return Promise.all(base64List.map(function (base64) {
              return new Promise(function (resolve) {
                var img = new Image();
                img.onload = function () {
                  resolve(img);
                };
                img.src = base64;
              });
            }));
          case 9:
            imgList = _context2.sent;
            imgList.forEach(function (img) {
              context === null || context === void 0 ? void 0 : context.drawImage(img, 0, 0, width, height);
            });
            return _context2.abrupt("return", canvas.toDataURL("image/".concat(imageType)));
          case 12:
          case "end":
            return _context2.stop();
        }
      }, _callee2);
    })));
    return _this;
  }
  (0, _createClass2.default)(ExportImage, [{
    key: "onAdd",
    value: function onAdd() {
      var button = (0, _get2.default)((0, _getPrototypeOf2.default)(ExportImage.prototype), "onAdd", this).call(this);
      button.addEventListener('click', this.onClick);
      return button;
    }
  }, {
    key: "getDefault",
    value: function getDefault(option) {
      return (0, _objectSpread2.default)((0, _objectSpread2.default)({}, (0, _get2.default)((0, _getPrototypeOf2.default)(ExportImage.prototype), "getDefault", this).call(this, option)), {}, {
        title: '导出图片',
        btnIcon: (0, _icon.createL7Icon)('l7-icon-export-picture'),
        imageType: 'png'
      });
    }
  }, {
    key: "getImage",
    value: function () {
      var _getImage = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee3() {
        var mapImage, layerImage;
        return _regenerator.default.wrap(function _callee3$(_context3) {
          while (1) switch (_context3.prev = _context3.next) {
            case 0:
              _context3.next = 2;
              return this.mapsService.exportMap('png');
            case 2:
              mapImage = _context3.sent;
              _context3.next = 5;
              return this.scene.exportPng('png');
            case 5:
              layerImage = _context3.sent;
              return _context3.abrupt("return", this.mergeImage.apply(this, (0, _toConsumableArray2.default)([mapImage, layerImage].filter(function (base64) {
                return base64;
              }))));
            case 7:
            case "end":
              return _context3.stop();
          }
        }, _callee3, this);
      }));
      function getImage() {
        return _getImage.apply(this, arguments);
      }
      return getImage;
    }()
  }]);
  return ExportImage;
}(_buttonControl.default);
exports.default = exports.ExportImage = ExportImage;
