// @antv/l7-component/lib/control/layerSwitch.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.LayerSwitch = void 0;
var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));
var _get2 = _interopRequireDefault(require("@babel/runtime/helpers/get"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _icon = require("../utils/icon");
var _selectControl = _interopRequireDefault(require("./baseControl/selectControl"));
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }
function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
var LayerSwitch = /*#__PURE__*/function (_SelectControl) {
  (0, _inherits2.default)(LayerSwitch, _SelectControl);
  var _super = _createSuper(LayerSwitch);
  function LayerSwitch() {
    var _this;
    (0, _classCallCheck2.default)(this, LayerSwitch);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onLayerChange", function () {
      var _this$controlOption$l;
      if ((_this$controlOption$l = _this.controlOption.layers) !== null && _this$controlOption$l !== void 0 && _this$controlOption$l.length) {
        return;
      }
      _this.selectValue = _this.getLayerVisible();
      _this.setOptions({
        options: _this.getLayerOptions()
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onLayerVisibleChane", function () {
      _this.setSelectValue(_this.getLayerVisible());
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onSelectChange", function () {
      _this.layers.forEach(function (layer) {
        var needShow = _this.selectValue.includes(layer.name);
        var isShow = layer.isVisible();
        if (needShow && !isShow) {
          layer.show();
        }
        if (!needShow && isShow) {
          layer.hide();
        }
      });
    });
    return _this;
  }
  (0, _createClass2.default)(LayerSwitch, [{
    key: "layers",
    get: function get() {
      var layerService = this.layerService;
      var layers = this.controlOption.layers;
      if (Array.isArray(layers) && layers.length) {
        var layerInstances = [];
        layers.forEach(function (layer) {
          if (layer instanceof Object) {
            layerInstances.push(layer);
          }
          if (typeof layer === 'string') {
            var targetLayer = layerService.getLayer(layer) || layerService.getLayerByName(layer);
            if (targetLayer) {
              layerInstances.push(targetLayer);
            }
          }
        });
        return layerInstances;
      }
      return layerService.getLayers() || [];
    }
  }, {
    key: "getDefault",
    value: function getDefault(option) {
      return (0, _objectSpread2.default)((0, _objectSpread2.default)({}, (0, _get2.default)((0, _getPrototypeOf2.default)(LayerSwitch.prototype), "getDefault", this).call(this, option)), {}, {
        title: '图层控制',
        btnIcon: (0, _icon.createL7Icon)('l7-icon-layer'),
        options: []
      });
    }
  }, {
    key: "getLayerVisible",
    value: function getLayerVisible() {
      return this.layers.filter(function (layer) {
        return layer.isVisible();
      }).map(function (layer) {
        return layer.name;
      });
    }
  }, {
    key: "getLayerOptions",
    value: function getLayerOptions() {
      return this.layers.map(function (layer) {
        return {
          text: layer.name,
          value: layer.name
        };
      });
    }
  }, {
    key: "setOptions",
    value: function setOptions(option) {
      var isLayerChange = this.checkUpdateOption(option, ['layers']);
      (0, _get2.default)((0, _getPrototypeOf2.default)(LayerSwitch.prototype), "setOptions", this).call(this, option);
      if (isLayerChange) {
        this.selectValue = this.getLayerVisible();
        this.controlOption.options = this.getLayerOptions();
        this.popper.setContent(this.getPopperContent(this.controlOption.options));
      }
    }
  }, {
    key: "onAdd",
    value: function onAdd() {
      var _this$controlOption$o;
      if (!((_this$controlOption$o = this.controlOption.options) !== null && _this$controlOption$o !== void 0 && _this$controlOption$o.length)) {
        this.controlOption.options = this.getLayerOptions();
      }
      if (!this.controlOption.defaultValue) {
        this.controlOption.defaultValue = this.getLayerVisible();
      }
      this.on('selectChange', this.onSelectChange);
      this.layerService.on('layerChange', this.onLayerChange);
      return (0, _get2.default)((0, _getPrototypeOf2.default)(LayerSwitch.prototype), "onAdd", this).call(this);
    }
  }, {
    key: "onRemove",
    value: function onRemove() {
      this.off('selectChange', this.onSelectChange);
      this.layerService.off('layerChange', this.onLayerChange);
    }
  }, {
    key: "getIsMultiple",
    value: function getIsMultiple() {
      return true;
    }
  }]);
  return LayerSwitch;
}(_selectControl.default);
exports.default = exports.LayerSwitch = LayerSwitch;
