// @antv/l7-component/lib/control/logo.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Logo = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _get2 = _interopRequireDefault(require("@babel/runtime/helpers/get"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _l7Utils = require("@antv/l7-utils");
var _baseControl = require("./baseControl");
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }
function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
var Logo = /*#__PURE__*/function (_Control) {
  (0, _inherits2.default)(Logo, _Control);
  var _super = _createSuper(Logo);
  function Logo() {
    (0, _classCallCheck2.default)(this, Logo);
    return _super.apply(this, arguments);
  }
  (0, _createClass2.default)(Logo, [{
    key: "getDefault",
    value: function getDefault() {
      return {
        position: _baseControl.PositionType.BOTTOMLEFT,
        name: 'logo',
        href: 'https://l7.antv.antgroup.com/',
        img: 'https://gw.alipayobjects.com/mdn/rms_816329/afts/img/A*GRb1TKp4HcMAAAAAAAAAAAAAARQnAQ'
      };
    }
  }, {
    key: "onAdd",
    value: function onAdd() {
      var container = _l7Utils.DOM.create('div', 'l7-control-logo');
      this.setLogoContent(container);
      return container;
    }
  }, {
    key: "onRemove",
    value: function onRemove() {
      return null;
    }
  }, {
    key: "setOptions",
    value: function setOptions(option) {
      (0, _get2.default)((0, _getPrototypeOf2.default)(Logo.prototype), "setOptions", this).call(this, option);
      if (this.checkUpdateOption(option, ['img', 'href'])) {
        _l7Utils.DOM.clearChildren(this.container);
        this.setLogoContent(this.container);
      }
    }
  }, {
    key: "setLogoContent",
    value: function setLogoContent(container) {
      var _this$controlOption = this.controlOption,
        href = _this$controlOption.href,
        img = _this$controlOption.img;
      var imgDOM = _l7Utils.DOM.create('img');
      imgDOM.setAttribute('src', img);
      imgDOM.setAttribute('aria-label', 'AntV logo');
      _l7Utils.DOM.setUnDraggable(imgDOM);
      if (href) {
        var anchorDOM = _l7Utils.DOM.create('a', 'l7-control-logo-link');
        anchorDOM.target = '_blank';
        anchorDOM.href = href;
        anchorDOM.rel = 'noopener nofollow';
        anchorDOM.setAttribute('rel', 'noopener nofollow');
        anchorDOM.appendChild(imgDOM);
        container.appendChild(anchorDOM);
      } else {
        container.appendChild(imgDOM);
      }
    }
  }]);
  return Logo;
}(_baseControl.Control);
exports.default = exports.Logo = Logo;
