// @antv/l7-component/lib/control/mapTheme.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.MapTheme = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));
var _get2 = _interopRequireDefault(require("@babel/runtime/helpers/get"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _constants = require("../constants");
var _icon = require("../utils/icon");
var _selectControl = _interopRequireDefault(require("./baseControl/selectControl"));
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }
function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
var MapTheme = /*#__PURE__*/function (_SelectControl) {
  (0, _inherits2.default)(MapTheme, _SelectControl);
  var _super = _createSuper(MapTheme);
  function MapTheme() {
    var _this;
    (0, _classCallCheck2.default)(this, MapTheme);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onMapThemeChange", function () {
      _this.mapsService.setMapStyle(_this.selectValue[0]);
    });
    return _this;
  }
  (0, _createClass2.default)(MapTheme, [{
    key: "getDefault",
    value: function getDefault(option) {
      return (0, _objectSpread2.default)((0, _objectSpread2.default)({}, (0, _get2.default)((0, _getPrototypeOf2.default)(MapTheme.prototype), "getDefault", this).call(this, option)), {}, {
        title: '地图样式',
        btnIcon: (0, _icon.createL7Icon)('l7-icon-color'),
        options: []
      });
    }
  }, {
    key: "getStyleOptions",
    value: function getStyleOptions() {
      var mapStyleConfig = this.mapsService.getType() === 'mapbox' ? _constants.MapboxMapStyleConfig : _constants.GaodeMapStyleConfig;
      return Object.entries(this.mapsService.getMapStyleConfig()).filter(function (_ref) {
        var _ref2 = (0, _slicedToArray2.default)(_ref, 2),
          key = _ref2[0],
          value = _ref2[1];
        return typeof value === 'string' && key !== 'blank';
      }).map(function (_ref3) {
        var _mapStyleConfig$key;
        var _ref4 = (0, _slicedToArray2.default)(_ref3, 2),
          key = _ref4[0],
          value = _ref4[1];
        // @ts-ignore
        var _ref5 = (_mapStyleConfig$key = mapStyleConfig[key]) !== null && _mapStyleConfig$key !== void 0 ? _mapStyleConfig$key : {},
          text = _ref5.text,
          img = _ref5.img;
        return {
          text: text !== null && text !== void 0 ? text : key,
          value: value,
          img: img,
          key: key
        };
      });
    }
  }, {
    key: "getMapStyle",
    value: function getMapStyle() {
      return this.mapsService.getMapStyle();
    }
  }, {
    key: "onAdd",
    value: function onAdd() {
      var _this$controlOption$o,
        _this2 = this;
      if (!((_this$controlOption$o = this.controlOption.options) !== null && _this$controlOption$o !== void 0 && _this$controlOption$o.length)) {
        this.controlOption.options = this.getStyleOptions();
      }
      if (this.controlOption.defaultValue) {
        var _this$controlOption$o2, _this$controlOption$o3;
        var defaultValue = this.controlOption.defaultValue;
        this.controlOption.defaultValue = (_this$controlOption$o2 = (_this$controlOption$o3 = this.controlOption.options.find(function (item) {
          return item.key === defaultValue;
        })) === null || _this$controlOption$o3 === void 0 ? void 0 : _this$controlOption$o3.value) !== null && _this$controlOption$o2 !== void 0 ? _this$controlOption$o2 : defaultValue;
      } else {
        var defaultStyle = this.getMapStyle();
        if (defaultStyle) {
          this.controlOption.defaultValue = defaultStyle;
        } else {
          // @ts-ignore
          this.mapsService.map.once('styledata', function () {
            var mapboxStyle = _this2.mapsService.getMapStyle();
            _this2.controlOption.defaultValue = mapboxStyle;
            _this2.setSelectValue(mapboxStyle, false);
          });
        }
      }
      this.on('selectChange', this.onMapThemeChange);
      return (0, _get2.default)((0, _getPrototypeOf2.default)(MapTheme.prototype), "onAdd", this).call(this);
    }
  }, {
    key: "getIsMultiple",
    value: function getIsMultiple() {
      return false;
    }
  }]);
  return MapTheme;
}(_selectControl.default);
exports.default = exports.MapTheme = MapTheme;
