// antd/lib/table/style/bordered.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault")["default"];
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _extends4 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var genBorderedStyle = function genBorderedStyle(token) {
  var _table, _ref4, _extends2, _ref6, _ref7;
  var componentCls = token.componentCls;
  var tableBorder = token.lineWidth + "px " + token.lineType + " " + token.tableBorderColor;
  var getSizeBorderStyle = function getSizeBorderStyle(size, paddingVertical, paddingHorizontal) {
    return (0, _defineProperty2["default"])({}, "&" + componentCls + "-" + size, (0, _defineProperty2["default"])({}, "> " + componentCls + "-container", (0, _defineProperty2["default"])({}, "> " + componentCls + "-content, > " + componentCls + "-body", {
      '> table > tbody > tr > td': (0, _defineProperty2["default"])({}, "> " + componentCls + "-expanded-row-fixed", {
        margin: "-" + paddingVertical + "px -" + (paddingHorizontal + token.lineWidth) + "px"
      })
    })));
  };
  return (0, _defineProperty2["default"])({}, componentCls + "-wrapper", (_ref7 = {}, (0, _defineProperty2["default"])(_ref7, "" + componentCls + componentCls + "-bordered", (0, _extends4["default"])((0, _extends4["default"])((0, _extends4["default"])((_extends2 = {}, (0, _defineProperty2["default"])(_extends2, "> " + componentCls + "-title", {
    border: tableBorder,
    borderBottom: 0
  }), (0, _defineProperty2["default"])(_extends2, "> " + componentCls + "-container", (_ref4 = {
    borderInlineStart: tableBorder
  }, (0, _defineProperty2["default"])(_ref4, "\n            > " + componentCls + "-content,\n            > " + componentCls + "-header,\n            > " + componentCls + "-body,\n            > " + componentCls + "-summary\n          ", {
    '> table': (_table = {}, (0, _defineProperty2["default"])(_table, "\n                > thead > tr > th,\n                > tbody > tr > td,\n                > tfoot > tr > th,\n                > tfoot > tr > td\n              ", {
      borderInlineEnd: tableBorder
    }), (0, _defineProperty2["default"])(_table, '> thead', {
      '> tr:not(:last-child) > th': {
        borderBottom: tableBorder
      },
      '> tr > th::before': {
        backgroundColor: 'transparent !important'
      }
    }), (0, _defineProperty2["default"])(_table, "\n                > thead > tr,\n                > tbody > tr,\n                > tfoot > tr\n              ", (0, _defineProperty2["default"])({}, "> " + componentCls + "-cell-fix-right-first::after", {
      borderInlineEnd: tableBorder
    })), (0, _defineProperty2["default"])(_table, '> table > tbody > tr > td', (0, _defineProperty2["default"])({}, "> " + componentCls + "-expanded-row-fixed", {
      margin: "-" + token.tablePaddingVertical + "px -" + (token.tablePaddingHorizontal + token.lineWidth) + "px",
      '&::after': {
        position: 'absolute',
        top: 0,
        insetInlineEnd: token.lineWidth,
        bottom: 0,
        borderInlineEnd: tableBorder,
        content: '""'
      }
    })), _table)
  }), (0, _defineProperty2["default"])(_ref4, "\n            > " + componentCls + "-content,\n            > " + componentCls + "-header\n          ", {
    '> table': {
      borderTop: tableBorder
    }
  }), _ref4)), (0, _defineProperty2["default"])(_extends2, "&" + componentCls + "-scroll-horizontal", (0, _defineProperty2["default"])({}, "> " + componentCls + "-container > " + componentCls + "-body", {
    '> table > tbody': (0, _defineProperty2["default"])({}, "\n                > tr" + componentCls + "-expanded-row,\n                > tr" + componentCls + "-placeholder\n              ", {
      '> td': {
        borderInlineEnd: 0
      }
    })
  })), _extends2), getSizeBorderStyle('middle', token.tablePaddingVerticalMiddle, token.tablePaddingHorizontalMiddle)), getSizeBorderStyle('small', token.tablePaddingVerticalSmall, token.tablePaddingHorizontalSmall)), (0, _defineProperty2["default"])({}, "> " + componentCls + "-footer", {
    border: tableBorder,
    borderTop: 0
  }))), (0, _defineProperty2["default"])(_ref7, componentCls + "-cell", (_ref6 = {}, (0, _defineProperty2["default"])(_ref6, componentCls + "-container:first-child", {
    // :first-child to avoid the case when bordered and title is set
    borderTop: 0
  }), (0, _defineProperty2["default"])(_ref6, '&-scrollbar:not([rowspan])', {
    boxShadow: "0 " + token.lineWidth + "px 0 " + token.lineWidth + "px " + token.tableHeaderBg
  }), _ref6)), _ref7));
};
var _default = genBorderedStyle;
exports["default"] = _default;
