// @antv/xflow/lib/core/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IMenuContribution = exports.IMenuService = exports.IToolbarContribution = exports.IToolbarService = exports.ToolbarRegistry = exports.useIsMountedRef = exports.useModelAsync = exports.createComponentModel = exports.useModel = exports.NsModel = exports.RxModel = exports.createModelServiceConfig = exports.IModelContribution = exports.IModelService = exports.MODELS = exports.createHookConfig = exports.IHookContribution = exports.IHookService = exports.getNodeReactComponent = exports.useAppContext = exports.AppContext = exports.createCmdConfig = exports.XFlowModelCommands = exports.XFlowGraphCommands = exports.XFlowGroupCommands = exports.XFlowEdgeCommands = exports.XFlowNodeCommands = exports.commandRegistryModule = exports.GraphCommandRegistry = exports.IGraphCommandFactory = exports.IGraphCommandContribution = exports.ICommandContextProvider = exports.IGraphCommandService = exports.ICommandHandler = exports.IApplicationContribution = exports.IApplication = exports.createGraphConfig = exports.IGraphProvider = exports.useExtensionRegistry = exports.ExtensionRegistryContext = exports.XFlowConfigProviderContext = exports.useXflowPrefixCls = exports.useXFlowApp = exports.XFlowAppContext = exports.XFlowAppProvider = exports.XFlowAppExtensionModule = exports.XFlowCanvas = exports.XFlow = exports.NsGraph = exports.ManaSyringe = void 0;
exports.XFlowConstants = exports.IconStore = exports.isReactComponent = exports.insertCss = exports.usePositionStyle = exports.uuidv4 = exports.delay = exports.Deferred = exports.DisposableCollection = exports.Disposable = exports.IKeyBindingContribution = exports.KeybindingConfig = exports.createKeybindingConfig = exports.KeyBindings = exports.MenuRegistry = exports.MenuItemType = void 0;
var xflow_core_1 = require("@antv/xflow-core");
Object.defineProperty(exports, "ManaSyringe", { enumerable: true, get: function () { return xflow_core_1.ManaSyringe; } });
/** XFlow 基础 interface */
var xflow_core_2 = require("@antv/xflow-core");
Object.defineProperty(exports, "NsGraph", { enumerable: true, get: function () { return xflow_core_2.NsGraph; } });
/*******************************************************
 * 核心组件：
 * 1. Application：XFlow
 * 2. Application Extension：XFlowAppExtensionModule
 * 3. Graph：XFlowCanvas
 *****************************************************/
var xflow_core_3 = require("@antv/xflow-core");
// 组件
Object.defineProperty(exports, "XFlow", { enumerable: true, get: function () { return xflow_core_3.XFlow; } });
Object.defineProperty(exports, "XFlowCanvas", { enumerable: true, get: function () { return xflow_core_3.XFlowCanvas; } });
Object.defineProperty(exports, "XFlowAppExtensionModule", { enumerable: true, get: function () { return xflow_core_3.XFlowAppExtensionModule; } });
// app context
Object.defineProperty(exports, "XFlowAppProvider", { enumerable: true, get: function () { return xflow_core_3.XFlowAppProvider; } });
Object.defineProperty(exports, "XFlowAppContext", { enumerable: true, get: function () { return xflow_core_3.XFlowAppContext; } });
Object.defineProperty(exports, "useXFlowApp", { enumerable: true, get: function () { return xflow_core_3.useXFlowApp; } });
// config provider context：使用全局Config
Object.defineProperty(exports, "useXflowPrefixCls", { enumerable: true, get: function () { return xflow_core_3.useXflowPrefixCls; } });
Object.defineProperty(exports, "XFlowConfigProviderContext", { enumerable: true, get: function () { return xflow_core_3.XFlowConfigProviderContext; } });
// extension context: 注册扩展
Object.defineProperty(exports, "ExtensionRegistryContext", { enumerable: true, get: function () { return xflow_core_3.ExtensionRegistryContext; } });
Object.defineProperty(exports, "useExtensionRegistry", { enumerable: true, get: function () { return xflow_core_3.useExtensionRegistry; } });
/** graphProvider：注入Graph时 需要 */
var xflow_core_4 = require("@antv/xflow-core");
Object.defineProperty(exports, "IGraphProvider", { enumerable: true, get: function () { return xflow_core_4.IGraphProvider; } });
Object.defineProperty(exports, "createGraphConfig", { enumerable: true, get: function () { return xflow_core_4.createGraphConfig; } });
/** app：用于extension扩展*/
var xflow_core_5 = require("@antv/xflow-core");
Object.defineProperty(exports, "IApplication", { enumerable: true, get: function () { return xflow_core_5.IApplication; } });
Object.defineProperty(exports, "IApplicationContribution", { enumerable: true, get: function () { return xflow_core_5.IApplicationContribution; } });
/** Command Service */
var xflow_core_6 = require("@antv/xflow-core");
Object.defineProperty(exports, "ICommandHandler", { enumerable: true, get: function () { return xflow_core_6.ICommandHandler; } });
Object.defineProperty(exports, "IGraphCommandService", { enumerable: true, get: function () { return xflow_core_6.IGraphCommandService; } });
Object.defineProperty(exports, "ICommandContextProvider", { enumerable: true, get: function () { return xflow_core_6.ICommandContextProvider; } });
Object.defineProperty(exports, "IGraphCommandContribution", { enumerable: true, get: function () { return xflow_core_6.IGraphCommandContribution; } });
Object.defineProperty(exports, "IGraphCommandFactory", { enumerable: true, get: function () { return xflow_core_6.IGraphCommandFactory; } });
Object.defineProperty(exports, "GraphCommandRegistry", { enumerable: true, get: function () { return xflow_core_6.GraphCommandRegistry; } });
Object.defineProperty(exports, "commandRegistryModule", { enumerable: true, get: function () { return xflow_core_6.commandRegistryModule; } });
/** Command 常量 */
var xflow_core_7 = require("@antv/xflow-core");
Object.defineProperty(exports, "XFlowNodeCommands", { enumerable: true, get: function () { return xflow_core_7.XFlowNodeCommands; } });
Object.defineProperty(exports, "XFlowEdgeCommands", { enumerable: true, get: function () { return xflow_core_7.XFlowEdgeCommands; } });
Object.defineProperty(exports, "XFlowGroupCommands", { enumerable: true, get: function () { return xflow_core_7.XFlowGroupCommands; } });
Object.defineProperty(exports, "XFlowGraphCommands", { enumerable: true, get: function () { return xflow_core_7.XFlowGraphCommands; } });
Object.defineProperty(exports, "XFlowModelCommands", { enumerable: true, get: function () { return xflow_core_7.XFlowModelCommands; } });
/** 创建 Command hook config */
Object.defineProperty(exports, "createCmdConfig", { enumerable: true, get: function () { return xflow_core_7.createCmdConfig; } });
/** React Node Context */
var xflow_core_8 = require("@antv/xflow-core");
Object.defineProperty(exports, "AppContext", { enumerable: true, get: function () { return xflow_core_8.AppContext; } });
Object.defineProperty(exports, "useAppContext", { enumerable: true, get: function () { return xflow_core_8.useAppContext; } });
Object.defineProperty(exports, "getNodeReactComponent", { enumerable: true, get: function () { return xflow_core_8.getNodeReactComponent; } });
/*******************************************************
 *  XFlow Hooks： 钩子函数
 *****************************************************/
var xflow_core_9 = require("@antv/xflow-core");
// 扩展Hooks
Object.defineProperty(exports, "IHookService", { enumerable: true, get: function () { return xflow_core_9.IHookService; } });
Object.defineProperty(exports, "IHookContribution", { enumerable: true, get: function () { return xflow_core_9.IHookContribution; } });
// 创建 React config hook
Object.defineProperty(exports, "createHookConfig", { enumerable: true, get: function () { return xflow_core_9.createHookConfig; } });
/*******************************************************
 *  Model Service：全局状态
 *****************************************************/
var xflow_core_10 = require("@antv/xflow-core");
Object.defineProperty(exports, "MODELS", { enumerable: true, get: function () { return xflow_core_10.MODELS; } });
Object.defineProperty(exports, "IModelService", { enumerable: true, get: function () { return xflow_core_10.IModelService; } });
Object.defineProperty(exports, "IModelContribution", { enumerable: true, get: function () { return xflow_core_10.IModelContribution; } });
Object.defineProperty(exports, "createModelServiceConfig", { enumerable: true, get: function () { return xflow_core_10.createModelServiceConfig; } });
var xflow_core_11 = require("@antv/xflow-core");
Object.defineProperty(exports, "RxModel", { enumerable: true, get: function () { return xflow_core_11.RxModel; } });
Object.defineProperty(exports, "NsModel", { enumerable: true, get: function () { return xflow_core_11.NsModel; } });
var xflow_core_12 = require("@antv/xflow-core");
Object.defineProperty(exports, "useModel", { enumerable: true, get: function () { return xflow_core_12.useModel; } });
Object.defineProperty(exports, "createComponentModel", { enumerable: true, get: function () { return xflow_core_12.createComponentModel; } });
Object.defineProperty(exports, "useModelAsync", { enumerable: true, get: function () { return xflow_core_12.useModelAsync; } });
Object.defineProperty(exports, "useIsMountedRef", { enumerable: true, get: function () { return xflow_core_12.useIsMountedRef; } });
/*******************************************************
 *  Toolbar：工具栏
 *****************************************************/
/** Toolbar 配置 */
var xflow_core_13 = require("@antv/xflow-core");
// component
Object.defineProperty(exports, "ToolbarRegistry", { enumerable: true, get: function () { return xflow_core_13.ToolbarRegistry; } });
// ioc 扩展
Object.defineProperty(exports, "IToolbarService", { enumerable: true, get: function () { return xflow_core_13.IToolbarService; } });
Object.defineProperty(exports, "IToolbarContribution", { enumerable: true, get: function () { return xflow_core_13.IToolbarContribution; } });
/*******************************************************
 *  Menu：菜单
 *****************************************************/
/** Menu 配置 */
var xflow_core_14 = require("@antv/xflow-core");
// 扩展 Service
Object.defineProperty(exports, "IMenuService", { enumerable: true, get: function () { return xflow_core_14.IMenuService; } });
Object.defineProperty(exports, "IMenuContribution", { enumerable: true, get: function () { return xflow_core_14.IMenuContribution; } });
// menu item
Object.defineProperty(exports, "MenuItemType", { enumerable: true, get: function () { return xflow_core_14.MenuItemType; } });
// Component
Object.defineProperty(exports, "MenuRegistry", { enumerable: true, get: function () { return xflow_core_14.MenuRegistry; } });
/*******************************************************
 *  KeyBindings：快捷键
 *****************************************************/
/** KeyBindings 配置 */
var xflow_core_15 = require("@antv/xflow-core");
Object.defineProperty(exports, "KeyBindings", { enumerable: true, get: function () { return xflow_core_15.KeyBindings; } });
Object.defineProperty(exports, "createKeybindingConfig", { enumerable: true, get: function () { return xflow_core_15.createKeybindingConfig; } });
Object.defineProperty(exports, "KeybindingConfig", { enumerable: true, get: function () { return xflow_core_15.KeybindingConfig; } });
Object.defineProperty(exports, "IKeyBindingContribution", { enumerable: true, get: function () { return xflow_core_15.IKeyBindingContribution; } });
/*******************************************************
 *  UTILS：工具方法
 *****************************************************/
/** utils：dispose */
var xflow_core_16 = require("@antv/xflow-core");
Object.defineProperty(exports, "Disposable", { enumerable: true, get: function () { return xflow_core_16.Disposable; } });
Object.defineProperty(exports, "DisposableCollection", { enumerable: true, get: function () { return xflow_core_16.DisposableCollection; } });
/** utils：defer 延迟 */
var xflow_core_17 = require("@antv/xflow-core");
Object.defineProperty(exports, "Deferred", { enumerable: true, get: function () { return xflow_core_17.Deferred; } });
/** utils：延迟 */
var xflow_core_18 = require("@antv/xflow-core");
Object.defineProperty(exports, "delay", { enumerable: true, get: function () { return xflow_core_18.delay; } });
/** utils：uuid */
var xflow_core_19 = require("@antv/xflow-core");
Object.defineProperty(exports, "uuidv4", { enumerable: true, get: function () { return xflow_core_19.uuidv4; } });
/** utils：绝对定位 */
var xflow_core_20 = require("@antv/xflow-core");
Object.defineProperty(exports, "usePositionStyle", { enumerable: true, get: function () { return xflow_core_20.usePositionStyle; } });
/** utils：insertCss */
var xflow_core_21 = require("@antv/xflow-core");
Object.defineProperty(exports, "insertCss", { enumerable: true, get: function () { return xflow_core_21.insertCss; } });
Object.defineProperty(exports, "isReactComponent", { enumerable: true, get: function () { return xflow_core_21.isReactComponent; } });
/*******************************************************
 * Icon：Antd Icon
 *****************************************************/
/** ICON */
var xflow_core_22 = require("@antv/xflow-core");
Object.defineProperty(exports, "IconStore", { enumerable: true, get: function () { return xflow_core_22.IconStore; } });
/** 全局常量 */
var xflow_core_23 = require("@antv/xflow-core");
Object.defineProperty(exports, "XFlowConstants", { enumerable: true, get: function () { return xflow_core_23.XFlowConstants; } });
