// @antv/xflow-core/lib/xflow-main/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IApplicationContribution = exports.IApplication = exports.GraphConfig = exports.createGraphConfig = exports.IGraphProvider = exports.useExtensionRegistry = exports.ExtensionRegistryContext = exports.useXflowPrefixCls = exports.XFlowConfigProviderContext = exports.useXFlowApp = exports.XFlowAppContext = exports.XFlowAppProvider = exports.XFlowAppExtensionModule = exports.XFlowCanvas = exports.XFlow = void 0;
var components_1 = require("./components");
Object.defineProperty(exports, "XFlow", { enumerable: true, get: function () { return components_1.XFlow; } });
var canvas_1 = require("./components/canvas");
Object.defineProperty(exports, "XFlowCanvas", { enumerable: true, get: function () { return canvas_1.XFlowCanvas; } });
/** 通过context获取app */
var app_context_1 = require("./components/app-context");
Object.defineProperty(exports, "XFlowAppProvider", { enumerable: true, get: function () { return app_context_1.XFlowAppProvider; } });
Object.defineProperty(exports, "XFlowAppContext", { enumerable: true, get: function () { return app_context_1.XFlowAppContext; } });
Object.defineProperty(exports, "useXFlowApp", { enumerable: true, get: function () { return app_context_1.useXFlowApp; } });
/** 通过context 注册Extension模块 */
var extension_context_1 = require("./components/extension-context");
Object.defineProperty(exports, "useExtensionRegistry", { enumerable: true, get: function () { return extension_context_1.useExtensionRegistry; } });
Object.defineProperty(exports, "ExtensionRegistryContext", { enumerable: true, get: function () { return extension_context_1.ExtensionRegistryContext; } });
/** XFlowAppExtension */
var app_extension_module_1 = require("./components/app-extension-module");
Object.defineProperty(exports, "XFlowAppExtensionModule", { enumerable: true, get: function () { return app_extension_module_1.XFlowAppExtensionModule; } });
/** classname  prefix  context */
var global_config_context_1 = require("./components/global-config-context");
Object.defineProperty(exports, "useXflowPrefixCls", { enumerable: true, get: function () { return global_config_context_1.useXflowPrefixCls; } });
Object.defineProperty(exports, "XFlowConfigProviderContext", { enumerable: true, get: function () { return global_config_context_1.XFlowConfigProviderContext; } });
/** Graph 配置 */
var graph_1 = require("./graph");
Object.defineProperty(exports, "IGraphProvider", { enumerable: true, get: function () { return graph_1.IGraphProvider; } });
Object.defineProperty(exports, "createGraphConfig", { enumerable: true, get: function () { return graph_1.createGraphConfig; } });
Object.defineProperty(exports, "GraphConfig", { enumerable: true, get: function () { return graph_1.GraphConfig; } });
/** Application 扩展依赖 */
var application_1 = require("./application");
Object.defineProperty(exports, "IApplication", { enumerable: true, get: function () { return application_1.FrontendApplication; } });
Object.defineProperty(exports, "IApplicationContribution", { enumerable: true, get: function () { return application_1.IFrontendApplicationContribution; } });
/** Application 扩展模块注册 */
var interface_1 = require("./interface");
