// antd-token-previewer/es/token-panel-pro/AliasPanel.js
import _toConsumableArray from "@babel/runtime/helpers/esm/toConsumableArray";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import { CaretRightOutlined, QuestionCircleOutlined, RightOutlined, ShrinkOutlined } from '@ant-design/icons';
import { Button, Collapse, Empty, Tooltip } from 'antd';
import classNames from 'classnames';
import useMergedState from 'rc-util/es/hooks/useMergedState';
import React, { useMemo } from 'react';
import { Pick } from "../icons";
import { mapRelatedAlias, seedRelatedAlias } from "../meta/TokenRelation";
import makeStyle from "../utils/makeStyle";
import { getRelatedComponents } from "../utils/statistic";
import TokenDetail from "./TokenDetail";
var Panel = Collapse.Panel;
var useStyle = makeStyle('TokenPanelProAlias', function (token) {
  var _$concat, _tokenPanelProAli, _tokenPanelProCol;

  return {
    '.token-panel-pro-color-alias': (_tokenPanelProCol = {
      display: 'flex',
      flexDirection: 'column',
      marginTop: 45,
      borderTop: "1px solid ".concat(token.colorSplit),
      '.token-panel-pro-color-alias-title': {
        display: 'flex',
        alignItems: 'center',
        padding: '0 16px',
        flex: '0 0 60px',
        '&-text': {
          fontSize: token.fontSizeLG,
          fontWeight: token.fontWeightStrong
        }
      },
      '.token-panel-pro-color-alias-description': {
        color: token.colorTextTertiary,
        fontSize: token.fontSizeSM,
        lineHeight: token.lineHeightSM,
        padding: '0 16px 12px'
      }
    }, _defineProperty(_tokenPanelProCol, ".token-panel-pro-alias-collapse".concat(token.rootCls, "-collapse"), (_tokenPanelProAli = {}, _defineProperty(_tokenPanelProAli, "> ".concat(token.rootCls, "-collapse-item > ").concat(token.rootCls, "-collapse-content > ").concat(token.rootCls, "-collapse-content-box"), {
      paddingBlock: '0'
    }), _defineProperty(_tokenPanelProAli, "> ".concat(token.rootCls, "-collapse-item"), _defineProperty({}, "> ".concat(token.rootCls, "-collapse-header"), (_$concat = {
      alignItems: 'center',
      padding: '8px 16px'
    }, _defineProperty(_$concat, "> ".concat(token.rootCls, "-collapse-header-text"), {
      flex: 1,
      '.token-panel-pro-token-collapse-map-collapse-count': {
        color: token.colorTextSecondary,
        display: 'inline-block',
        fontSize: 12,
        lineHeight: '16px',
        padding: '0 6px',
        backgroundColor: token.colorFillAlter,
        borderRadius: 999
      }
    }), _defineProperty(_$concat, '.token-panel-pro-token-picked', {
      color: token.colorPrimary
    }), _$concat))), _tokenPanelProAli)), _defineProperty(_tokenPanelProCol, '.token-panel-pro-color-alias-expand', {
      height: '100%',
      width: 20,
      transform: 'translateX(-50%)',
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center',
      '&:hover': {
        '.token-panel-pro-color-alias-expand-handler': {
          opacity: 1
        }
      },
      '.token-panel-pro-color-alias-expand-handler': {
        height: 100,
        width: 16,
        borderRadius: 999,
        border: "1px solid ".concat(token.colorSplit),
        backgroundColor: '#fff',
        margin: 'auto',
        cursor: 'pointer',
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center',
        opacity: 0,
        transition: 'box-shadow 0.2s',
        '&:hover': {
          boxShadow: token.boxShadow
        }
      }
    }), _tokenPanelProCol)
  };
});

var AliasPanel = function AliasPanel(_ref) {
  var className = _ref.className,
      activeSeeds = _ref.activeSeeds,
      theme = _ref.theme,
      style = _ref.style,
      selectedTokens = _ref.selectedTokens,
      onTokenSelect = _ref.onTokenSelect,
      customOpen = _ref.open,
      onOpenChange = _ref.onOpenChange,
      description = _ref.description;

  var _useStyle = useStyle(),
      _useStyle2 = _slicedToArray(_useStyle, 2),
      wrapSSR = _useStyle2[0],
      hashId = _useStyle2[1];

  var _useMergedState = useMergedState(customOpen !== null && customOpen !== void 0 ? customOpen : true, {
    value: customOpen,
    onChange: onOpenChange
  }),
      _useMergedState2 = _slicedToArray(_useMergedState, 2),
      open = _useMergedState2[0],
      setOpen = _useMergedState2[1];

  var shownAlias = useMemo(function () {
    var _selectedTokens$map;

    return selectedTokens !== null && selectedTokens !== void 0 && (_selectedTokens$map = selectedTokens.map) !== null && _selectedTokens$map !== void 0 && _selectedTokens$map.length ? Array.from(new Set(selectedTokens === null || selectedTokens === void 0 ? void 0 : selectedTokens.map.reduce(function (result, map) {
      var _map;

      return result.concat.apply(result, _toConsumableArray((_map = mapRelatedAlias[map]) !== null && _map !== void 0 ? _map : []));
    }, []))) : activeSeeds === null || activeSeeds === void 0 ? void 0 : activeSeeds.reduce(function (result, item) {
      var _item;

      return result.concat((_item = seedRelatedAlias[item]) !== null && _item !== void 0 ? _item : []);
    }, []);
  }, [selectedTokens, activeSeeds]);
  return wrapSSR( /*#__PURE__*/React.createElement("div", {
    className: classNames(className, 'token-panel-pro-color-alias', hashId),
    style: style
  }, open ? /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement("div", {
    className: "token-panel-pro-color-alias-title"
  }, /*#__PURE__*/React.createElement("span", {
    className: "token-panel-pro-color-alias-title-text"
  }, "Alias Token"), /*#__PURE__*/React.createElement(Tooltip, {
    placement: "topLeft",
    arrowPointAtCenter: true,
    title: "\u522B\u540D\u53D8\u91CF\uFF08Alias Token\uFF09\u662F Map Token \u7684\u522B\u540D\u3002Alias Token \u7528\u4E8E\u6279\u91CF\u63A7\u5236\u67D0\u4E9B\u5171\u6027\u7EC4\u4EF6\u7684\u6837\u5F0F\u3002"
  }, /*#__PURE__*/React.createElement(QuestionCircleOutlined, {
    style: {
      fontSize: 14,
      marginLeft: 4
    }
  })), /*#__PURE__*/React.createElement(Button, {
    type: "text",
    icon: /*#__PURE__*/React.createElement(ShrinkOutlined, null),
    style: {
      marginLeft: 'auto'
    },
    onClick: function onClick() {
      return setOpen(false);
    }
  })), description && /*#__PURE__*/React.createElement("div", {
    className: "token-panel-pro-color-alias-description"
  }, description), /*#__PURE__*/React.createElement("div", {
    style: {
      flex: 1,
      overflow: 'auto'
    }
  }, /*#__PURE__*/React.createElement(Collapse, {
    className: "token-panel-pro-alias-collapse",
    ghost: true,
    expandIcon: function expandIcon(_ref2) {
      var isActive = _ref2.isActive;
      return /*#__PURE__*/React.createElement(CaretRightOutlined, {
        rotate: isActive ? 90 : 0,
        style: {
          fontSize: 12
        }
      });
    }
  }, shownAlias === null || shownAlias === void 0 ? void 0 : shownAlias.map(function (aliasToken) {
    var _selectedTokens$alias;

    return /*#__PURE__*/React.createElement(Panel, {
      header: /*#__PURE__*/React.createElement("div", {
        style: {
          display: 'flex',
          alignItems: 'center'
        }
      }, /*#__PURE__*/React.createElement("span", {
        style: {
          marginRight: 8
        }
      }, aliasToken), /*#__PURE__*/React.createElement("span", {
        className: "token-panel-pro-token-collapse-map-collapse-count"
      }, getRelatedComponents(aliasToken).length), /*#__PURE__*/React.createElement("div", {
        style: {
          padding: 4,
          marginLeft: 'auto'
        },
        onClick: function onClick(e) {
          e.stopPropagation();
          onTokenSelect === null || onTokenSelect === void 0 ? void 0 : onTokenSelect(aliasToken, 'alias');
        }
      }, /*#__PURE__*/React.createElement(Pick, {
        className: classNames('token-panel-pro-token-pick', {
          'token-panel-pro-token-picked': selectedTokens === null || selectedTokens === void 0 ? void 0 : (_selectedTokens$alias = selectedTokens.alias) === null || _selectedTokens$alias === void 0 ? void 0 : _selectedTokens$alias.includes(aliasToken)
        })
      }))),
      key: aliasToken
    }, /*#__PURE__*/React.createElement(TokenDetail, {
      style: {
        paddingBottom: 10
      },
      themes: [theme],
      path: ['token'],
      tokenName: aliasToken
    }));
  })), !(shownAlias !== null && shownAlias !== void 0 && shownAlias.length) && /*#__PURE__*/React.createElement(Empty, {
    image: Empty.PRESENTED_IMAGE_SIMPLE,
    description: "\u6682\u65E0\u76F8\u5173 Alias Token"
  }))) : /*#__PURE__*/React.createElement("div", {
    className: "token-panel-pro-color-alias-expand"
  }, /*#__PURE__*/React.createElement("div", {
    className: "token-panel-pro-color-alias-expand-handler",
    onClick: function onClick() {
      return setOpen(true);
    }
  }, /*#__PURE__*/React.createElement(RightOutlined, {
    style: {
      fontSize: 12
    }
  })))));
};

export default AliasPanel;
