// antd-token-previewer/es/token-panel-pro/TokenPreview.js
import React from 'react';
import getColorBgImg from "../utils/getColorBgImg";
import getDesignToken from "../utils/getDesignToken";

var TokenPreview = function TokenPreview(_ref) {
  var theme = _ref.theme,
      tokenName = _ref.tokenName,
      type = _ref.type;

  if (type === 'Color') {
    return /*#__PURE__*/React.createElement("div", {
      style: {
        background: "".concat(getColorBgImg(false), " 0% 0% / 28px"),
        width: '100%',
        height: '100%',
        position: 'relative'
      }
    }, /*#__PURE__*/React.createElement("div", {
      style: {
        height: '100%',
        width: '100%',
        backgroundColor: getDesignToken(theme)[tokenName],
        transition: 'background-color 0.2s'
      }
    }));
  }

  if (type === 'FontSize') {
    return /*#__PURE__*/React.createElement("div", {
      style: {
        width: '100%',
        height: '100%',
        position: 'relative',
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center',
        overflow: 'hidden',
        background: "".concat(getColorBgImg(false), " 0% 0% / 28px"),
        fontSize: getDesignToken(theme)[tokenName],
        fontWeight: 700
      }
    }, /*#__PURE__*/React.createElement("span", null, "Aa"));
  }

  if (type === 'LineHeight') {
    return /*#__PURE__*/React.createElement("div", {
      style: {
        width: '100%',
        height: '100%',
        position: 'relative',
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center',
        fontWeight: 700,
        overflow: 'hidden',
        background: "".concat(getColorBgImg(false), " 0% 0% / 28px")
      }
    }, /*#__PURE__*/React.createElement("span", {
      style: {
        fontSize: getDesignToken(theme)[tokenName.replace('lineHeight', 'fontSize')],
        lineHeight: getDesignToken(theme)[tokenName],
        background: '#fff2f0',
        paddingInline: 8
      }
    }, "Aa"));
  }

  if (type === 'Margin') {
    var margin = getDesignToken(theme)[tokenName];
    return /*#__PURE__*/React.createElement("div", {
      style: {
        width: '100%',
        height: '100%',
        position: 'relative',
        overflow: 'hidden',
        background: "".concat(getColorBgImg(false), " 0% 0% / 28px")
      }
    }, /*#__PURE__*/React.createElement("div", {
      style: {
        width: '100%',
        height: '100%',
        overflow: 'hidden',
        background: '#fff1b8',
        transform: 'translate(10%, 10%) scale(0.8)'
      }
    }, /*#__PURE__*/React.createElement("div", {
      style: {
        marginLeft: margin,
        marginTop: margin,
        width: "calc(100% - ".concat(margin, "px)"),
        height: "calc(100% - ".concat(margin, "px)"),
        background: '#bae0ff'
      }
    })));
  }

  if (type === 'Padding') {
    var padding = getDesignToken(theme)[tokenName];
    return /*#__PURE__*/React.createElement("div", {
      style: {
        width: '100%',
        height: '100%',
        position: 'relative',
        overflow: 'hidden',
        background: "".concat(getColorBgImg(false), " 0% 0% / 28px")
      }
    }, /*#__PURE__*/React.createElement("div", {
      style: {
        width: '100%',
        height: '100%',
        overflow: 'hidden',
        background: '#d9f7be',
        transform: 'translate(10%, 10%) scale(0.8)',
        paddingLeft: padding,
        paddingTop: padding
      }
    }, /*#__PURE__*/React.createElement("div", {
      style: {
        width: "100%",
        height: "100%",
        background: '#bae0ff'
      }
    })));
  }

  if (type === 'BorderRadius') {
    return /*#__PURE__*/React.createElement("div", {
      style: {
        width: '100%',
        height: '100%',
        position: 'relative',
        overflow: 'hidden',
        background: "".concat(getColorBgImg(false), " 0% 0% / 28px")
      }
    }, /*#__PURE__*/React.createElement("div", {
      style: {
        width: '100%',
        height: '100%',
        overflow: 'hidden',
        transform: 'translate(30%, 30%)',
        border: '2px solid rgba(0,0,0,0.45)',
        background: '#fff',
        borderRadius: getDesignToken(theme)[tokenName]
      }
    }));
  }

  if (type === 'BoxShadow') {
    return /*#__PURE__*/React.createElement("div", {
      style: {
        width: '100%',
        height: '100%',
        position: 'relative',
        overflow: 'hidden',
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center',
        background: "".concat(getColorBgImg(false), " 0% 0% / 28px")
      }
    }, /*#__PURE__*/React.createElement("div", {
      style: {
        width: '60%',
        height: '50%',
        borderRadius: 6,
        background: '#fff',
        border: '1px solid #d9d9d9',
        boxShadow: getDesignToken(theme)[tokenName]
      }
    }));
  }

  return null;
};

export default TokenPreview;
