// antd-token-previewer/es/component-demos/treeSelect/tree-select.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import React, { useState } from 'react';
import { TreeSelect as _TreeSelect } from 'antd';
var TreeNode = _TreeSelect.TreeNode,
    TreeSelect = _TreeSelect._InternalPanelDoNotUseOrYouWillBeFired;

var Demo = function Demo() {
  var _useState = useState(undefined),
      _useState2 = _slicedToArray(_useState, 2),
      value = _useState2[0],
      setValue = _useState2[1];

  var onChange = function onChange() {
    setValue(value);
  };

  return /*#__PURE__*/React.createElement(TreeSelect, {
    showSearch: true,
    style: {
      width: '100%'
    },
    value: value,
    dropdownStyle: {
      maxHeight: 400,
      overflow: 'auto'
    },
    placeholder: "Please select",
    allowClear: true,
    treeDefaultExpandAll: true,
    onChange: onChange
  }, /*#__PURE__*/React.createElement(TreeNode, {
    value: "parent 1",
    title: "parent 1"
  }, /*#__PURE__*/React.createElement(TreeNode, {
    value: "parent 1-0",
    title: "parent 1-0"
  }, /*#__PURE__*/React.createElement(TreeNode, {
    value: "leaf1",
    title: "leaf1"
  }), /*#__PURE__*/React.createElement(TreeNode, {
    value: "leaf2",
    title: "leaf2"
  })), /*#__PURE__*/React.createElement(TreeNode, {
    value: "parent 1-1",
    title: "parent 1-1"
  }, /*#__PURE__*/React.createElement(TreeNode, {
    value: "leaf3",
    title: 'leaf3'
  }))));
};

var componentDemo = {
  demo: /*#__PURE__*/React.createElement(Demo, null),
  tokens: ['colorPrimary', 'colorPrimaryActive', 'controlOutline', 'colorBgElevated', 'colorBgContainer'],
  key: 'default'
};
export default componentDemo;
