// antd-token-previewer/es/component-demos/upload/danger.js
import { UploadOutlined } from '@ant-design/icons';
import { Upload, Button } from 'antd';
import React from 'react';

var Demo = function Demo() {
  return /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(Upload, {
    defaultFileList: [{
      uid: '3',
      name: 'zzz.png',
      status: 'error',
      response: 'Server Error 500',
      url: 'http://www.baidu.com/zzz.png'
    }]
  }, /*#__PURE__*/React.createElement(Button, {
    icon: /*#__PURE__*/React.createElement(UploadOutlined, null)
  }, "Upload")), /*#__PURE__*/React.createElement(Upload, {
    listType: 'picture',
    defaultFileList: [{
      uid: '3',
      name: 'zzz.png',
      status: 'error',
      response: 'Server Error 500'
    }]
  }, /*#__PURE__*/React.createElement(Button, {
    icon: /*#__PURE__*/React.createElement(UploadOutlined, null)
  }, "Upload")));
};

var componentDemo = {
  demo: /*#__PURE__*/React.createElement(Demo, null),
  tokens: ['colorError', 'colorErrorBg'],
  key: 'danger'
};
export default componentDemo;
