// @antv/l7plot/dist/lib/component/legend.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Legend = void 0;
var tslib_1 = require("tslib");
var l7_component_1 = require("@antv/l7-component");
var l7plot_component_1 = require("@antv/l7plot-component");
var Legend = /** @class */ (function (_super) {
    tslib_1.__extends(Legend, _super);
    function Legend(options) {
        var _this = _super.call(this, options) || this;
        /**
         * legendComponents 实例
         */
        _this.legendComponents = [];
        _this.options = options;
        _this.legendComponents = _this.initLegendComponents(options.items);
        return _this;
    }
    Legend.prototype.initLegendComponents = function (legendItem) {
        var legendComponents = [];
        for (var index = 0; index < legendItem.length; index++) {
            var item = legendItem[index];
            if (item.type === 'category') {
                var legend = new l7plot_component_1.CategoryLegend({
                    title: item.options.title,
                    items: item.options.items,
                    className: item.options.className,
                    customContent: item.options.customContent,
                    domStyles: item.options.domStyles,
                });
                legendComponents.push(legend);
            }
            else if (item.type === 'continue') {
                var legend = new l7plot_component_1.ContinueLegend({
                    title: item.options.title,
                    min: item.options.min,
                    max: item.options.max,
                    colors: item.options.colors,
                    className: item.options.className,
                    customContent: item.options.customContent,
                    domStyles: item.options.domStyles,
                });
                legendComponents.push(legend);
            }
        }
        return legendComponents;
    };
    /**
     * 获取默认配置
     */
    Legend.prototype.getDefaultOptions = function () {
        return {
            position: 'bottomleft',
        };
    };
    Legend.prototype.onAdd = function () {
        var container = window.document.createElement('div');
        container.className = 'l7plot-legend-container';
        this.legendComponents.forEach(function (legendComponent) {
            var legend = legendComponent.getContainer();
            container.appendChild(legend);
        });
        return container;
    };
    Legend.prototype.onRemove = function () {
        this.legendComponents.forEach(function (legendComponent) {
            legendComponent.destroy();
        });
    };
    return Legend;
}(l7_component_1.Control));
exports.Legend = Legend;
