// antd-token-previewer/es/token-panel/token-item/index.js
import _toConsumableArray from "@babel/runtime/helpers/esm/toConsumableArray";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import _extends from "@babel/runtime/helpers/esm/extends";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
var _excluded = ["info", "visible", "tokenName", "style", "dark"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

import { CaretRightOutlined } from '@ant-design/icons';
import { Collapse, Space } from 'antd';
import classNames from 'classnames';
import React, { useEffect, useMemo } from 'react';
import ColorPreview from "../../ColorPreview";
import { Pick } from "../../icons";
import TokenInput from "../../TokenInput";
import getValueByPath from "../../utils/getValueByPath";
import isColor from "../../utils/isColor";
import makeStyle from "../../utils/makeStyle";
import { getRelatedComponents } from "../../utils/statistic";
var Panel = Collapse.Panel;

var AdditionInfo = function AdditionInfo(_ref) {
  var info = _ref.info,
      visible = _ref.visible,
      tokenName = _ref.tokenName,
      style = _ref.style,
      dark = _ref.dark,
      rest = _objectWithoutProperties(_ref, _excluded);

  if (typeof info === 'string' && isColor(info)) {
    return /*#__PURE__*/React.createElement(ColorPreview, {
      dark: dark,
      color: String(info),
      style: _objectSpread({
        display: visible ? 'block' : 'none'
      }, style)
    });
  }

  if (info.toString().length < 6 && String(info) !== '') {
    return /*#__PURE__*/React.createElement("div", _extends({
      style: _objectSpread({
        height: 20,
        overflow: 'hidden',
        backgroundColor: 'rgba(0,0,0,0.04)',
        borderRadius: '8px',
        display: visible ? 'block' : 'none',
        padding: '0 6px',
        lineHeight: '20px'
      }, style)
    }, rest), info);
  }

  return null;
};

var ShowUsageButton = function ShowUsageButton(_ref2) {
  var selected = _ref2.selected,
      toggleSelected = _ref2.toggleSelected;
  return /*#__PURE__*/React.createElement(Pick, {
    style: {
      color: selected ? '#1890ff' : undefined,
      cursor: 'pointer',
      fontSize: 16,
      transition: 'color 0.3s',
      marginInlineStart: 12,
      verticalAlign: 'middle'
    },
    onClick: function onClick() {
      return toggleSelected(!selected);
    }
  });
};

var useStyle = makeStyle('TokenItem', function (token) {
  var _previewerTokenIte;

  return _defineProperty({}, "".concat(token.rootCls, "-collapse.previewer-token-item-collapse"), _defineProperty({}, ".previewer-token-item".concat(token.rootCls, "-collapse-item"), (_previewerTokenIte = {
    transition: "background-color ".concat(token.motionDurationSlow),
    borderRadius: {
      _skip_check_: true,
      value: "4px !important"
    }
  }, _defineProperty(_previewerTokenIte, "&:not(".concat(token.rootCls, "-collapse-item-active):hover"), {
    backgroundColor: '#f5f5f5'
  }), _defineProperty(_previewerTokenIte, "> ".concat(token.rootCls, "-collapse-header"), {
    padding: '12px 8px'
  }), _defineProperty(_previewerTokenIte, "".concat(token.rootCls, "-collapse-header-text"), {
    flex: 1,
    width: 0
  }), _defineProperty(_previewerTokenIte, "".concat(token.rootCls, "-collapse-content-box"), {
    padding: '0 4px'
  }), _defineProperty(_previewerTokenIte, "".concat(token.rootCls, "-collapse-expand-icon"), {
    paddingInlineEnd: "".concat(token.paddingXXS, "px !important")
  }), _defineProperty(_previewerTokenIte, '.previewer-token-count', {
    height: 16,
    fontSize: token.fontSizeSM,
    lineHeight: '16px',
    borderRadius: 100,
    paddingInline: token.paddingXXS * 1.5,
    color: token.colorTextSecondary,
    backgroundColor: token.colorFillAlter
  }), _defineProperty(_previewerTokenIte, '.previewer-token-item-name', {
    transition: 'color 0.3s'
  }), _defineProperty(_previewerTokenIte, '.previewer-token-item-highlighted.previewer-token-item-name', {
    color: "".concat(token.colorPrimary, " !important")
  }), _defineProperty(_previewerTokenIte, '&:hover .previewer-token-preview', {
    '> .previewer-color-preview:not(:last-child)': {
      transform: 'translateX(-100%)',
      marginInlineEnd: 4
    }
  }), _defineProperty(_previewerTokenIte, '.previewer-token-preview', {
    display: 'flex',
    alignItems: 'center',
    position: 'relative',
    '> .previewer-color-preview': {
      position: 'absolute',
      insetInlineEnd: 0,
      top: 0,
      bottom: 0,
      margin: 'auto'
    },
    '> .previewer-color-preview:not(:last-child)': {
      transform: 'translateX(-50%)',
      marginInlineEnd: 0,
      transition: 'transform 0.3s, margin-right 0.3s'
    },
    '> *:not(:last-child)': {
      marginInlineEnd: 4
    }
  }), _previewerTokenIte)));
});
export var getTokenItemId = function getTokenItemId(token) {
  return "previewer-token-panel-item-".concat(token);
};
export default (function (_ref4) {
  var tokenName = _ref4.tokenName,
      active = _ref4.active,
      onActiveChange = _ref4.onActiveChange,
      onTokenChange = _ref4.onTokenChange,
      tokenPath = _ref4.tokenPath,
      selectedTokens = _ref4.selectedTokens,
      themes = _ref4.themes,
      onTokenSelect = _ref4.onTokenSelect,
      enableTokenSelect = _ref4.enableTokenSelect,
      hideUsageCount = _ref4.hideUsageCount,
      fallback = _ref4.fallback;

  var _React$useState = React.useState(false),
      _React$useState2 = _slicedToArray(_React$useState, 2),
      infoVisible = _React$useState2[0],
      setInfoVisible = _React$useState2[1];

  var _useStyle = useStyle(),
      _useStyle2 = _slicedToArray(_useStyle, 2),
      wrapSSR = _useStyle2[0],
      hashId = _useStyle2[1];

  useEffect(function () {
    if (active) {
      setInfoVisible(true);
    }
  }, [active]);

  var handleTokenChange = function handleTokenChange(theme, value) {
    onTokenChange === null || onTokenChange === void 0 ? void 0 : onTokenChange(theme, tokenName, value);
  };

  var count = useMemo(function () {
    return getRelatedComponents(tokenName).length;
  }, [tokenName]);
  return wrapSSR( /*#__PURE__*/React.createElement("div", {
    onMouseEnter: function onMouseEnter() {
      return onActiveChange === null || onActiveChange === void 0 ? void 0 : onActiveChange(false);
    }
  }, /*#__PURE__*/React.createElement(Collapse, {
    collapsible: "header",
    ghost: true,
    onChange: function onChange(key) {
      return setInfoVisible(key.length > 0);
    },
    className: classNames('previewer-token-item-collapse', hashId),
    expandIcon: function expandIcon(_ref5) {
      var isActive = _ref5.isActive;
      return /*#__PURE__*/React.createElement(CaretRightOutlined, {
        rotate: isActive ? 90 : 0,
        style: {
          fontSize: 12,
          cursor: 'pointer'
        }
      });
    },
    activeKey: infoVisible ? tokenName : undefined
  }, /*#__PURE__*/React.createElement(Panel, {
    key: tokenName,
    className: "previewer-token-item",
    header: /*#__PURE__*/React.createElement("div", {
      style: {
        display: 'flex',
        alignItems: 'center',
        gap: 8
      },
      id: getTokenItemId(tokenName)
    }, /*#__PURE__*/React.createElement("span", {
      style: {
        flex: 1,
        width: 0,
        display: 'flex',
        overflow: 'hidden',
        alignItems: 'center'
      }
    }, /*#__PURE__*/React.createElement("span", {
      title: tokenName,
      className: classNames('previewer-token-item-name', {
        'previewer-token-item-highlighted': active
      }),
      style: {
        marginInlineEnd: '5px',
        overflow: 'hidden',
        textOverflow: 'ellipsis'
      }
    }, tokenName), !hideUsageCount && /*#__PURE__*/React.createElement("span", {
      className: "previewer-token-count"
    }, count)), !infoVisible && /*#__PURE__*/React.createElement("div", {
      className: "previewer-token-preview",
      style: {
        minWidth: themes.length * 20 + (themes.length - 1) * 4
      }
    }, themes.map(function (_ref6, index) {
      var _ref7, _getValueByPath;

      var config = _ref6.config,
          key = _ref6.key;
      return /*#__PURE__*/React.createElement(AdditionInfo, {
        key: key,
        dark: key === 'dark',
        tokenName: tokenName,
        info: (_ref7 = (_getValueByPath = getValueByPath(config, [].concat(_toConsumableArray(tokenPath), [tokenName]))) !== null && _getValueByPath !== void 0 ? _getValueByPath : fallback === null || fallback === void 0 ? void 0 : fallback(config)[tokenName]) !== null && _ref7 !== void 0 ? _ref7 : '',
        visible: !infoVisible,
        style: {
          zIndex: 10 - index
        }
      });
    }))),
    extra: enableTokenSelect ? /*#__PURE__*/React.createElement(ShowUsageButton, {
      selected: !!(selectedTokens !== null && selectedTokens !== void 0 && selectedTokens.includes(tokenName)),
      toggleSelected: function toggleSelected() {
        onTokenSelect === null || onTokenSelect === void 0 ? void 0 : onTokenSelect(tokenName);
      }
    }) : undefined
  }, /*#__PURE__*/React.createElement(Space, {
    direction: "vertical",
    style: {
      background: '#fafafa',
      borderRadius: 4,
      padding: 8,
      width: '100%'
    }
  }, themes.map(function (theme) {
    var _getValueByPath2;

    return /*#__PURE__*/React.createElement("div", {
      key: theme.key
    }, /*#__PURE__*/React.createElement(TokenInput, {
      hideTheme: themes.length === 1,
      theme: theme,
      onChange: function onChange(value) {
        return handleTokenChange(theme, value);
      },
      value: (_getValueByPath2 = getValueByPath(theme.config, [].concat(_toConsumableArray(tokenPath), [tokenName]))) !== null && _getValueByPath2 !== void 0 ? _getValueByPath2 : fallback === null || fallback === void 0 ? void 0 : fallback(theme.config)[tokenName]
    }));
  }))))));
});
