// antd-token-previewer/es/component-demos/select/select.js
import React from 'react';
import { Select as _Select, Space } from 'antd';
var Option = _Select.Option,
    Select = _Select._InternalPanelDoNotUseOrYouWillBeFired;

function handleChange() {}

var Demo = function Demo() {
  return /*#__PURE__*/React.createElement(Space, {
    align: 'start'
  }, /*#__PURE__*/React.createElement(Select, {
    defaultValue: "lucy",
    style: {
      width: 120
    },
    onChange: handleChange
  }, /*#__PURE__*/React.createElement(Option, {
    value: "jack"
  }, "Jack"), " ", /*#__PURE__*/React.createElement(Option, {
    value: "lucy"
  }, "Lucy"), /*#__PURE__*/React.createElement(Option, {
    value: "disabled",
    disabled: true
  }, "Disabled"), /*#__PURE__*/React.createElement(Option, {
    value: "Yiminghe"
  }, "yiminghe")), /*#__PURE__*/React.createElement(Select, {
    defaultValue: "lucy",
    style: {
      width: 120
    },
    disabled: true
  }, /*#__PURE__*/React.createElement(Option, {
    value: "lucy"
  }, "Lucy")), /*#__PURE__*/React.createElement(Select, {
    defaultValue: "lucy",
    style: {
      width: 120
    },
    loading: true
  }, /*#__PURE__*/React.createElement(Option, {
    value: "lucy"
  }, "Lucy")), /*#__PURE__*/React.createElement(Select, {
    defaultValue: "lucy",
    style: {
      width: 120
    },
    allowClear: true
  }, /*#__PURE__*/React.createElement(Option, {
    value: "lucy"
  }, "Lucy")));
};

var componentDemo = {
  demo: /*#__PURE__*/React.createElement(Demo, null),
  tokens: ['controlOutline', 'colorPrimary', 'colorPrimaryHover', 'colorText', 'colorBgElevated', 'colorBgContainer'],
  key: 'select'
};
export default componentDemo;
