// @antv/l7-layers/lib/tile/tileFactory/DebugTile.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _line = _interopRequireDefault(require("../../line"));
var _point = _interopRequireDefault(require("../../point"));
var _Tile2 = _interopRequireDefault(require("./Tile"));
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }
function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
var DebugTile = /*#__PURE__*/function (_Tile) {
  (0, _inherits2.default)(DebugTile, _Tile);
  var _super = _createSuper(DebugTile);
  function DebugTile() {
    (0, _classCallCheck2.default)(this, DebugTile);
    return _super.apply(this, arguments);
  }
  (0, _createClass2.default)(DebugTile, [{
    key: "initTileLayer",
    value: function () {
      var _initTileLayer = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee() {
        var sourceOptions, pointData, lineLayer, pointLayer;
        return _regenerator.default.wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              sourceOptions = this.getSourceOption();
              pointData = sourceOptions.data.features[0].properties;
              lineLayer = new _line.default().source(sourceOptions.data, sourceOptions.options).size(1).shape('line').color('red');
              pointLayer = new _point.default({
                minZoom: this.z - 1,
                maxZoom: this.z + 1
              }).source([pointData], {
                parser: {
                  type: 'json',
                  x: 'textLng',
                  y: 'textLat'
                }
              }).size(20).color('red').shape(this.key).style({
                stroke: '#fff',
                strokeWidth: 2
              });
              _context.next = 6;
              return this.addLayer(lineLayer);
            case 6:
              _context.next = 8;
              return this.addLayer(pointLayer);
            case 8:
              this.isLoaded = true;
            case 9:
            case "end":
              return _context.stop();
          }
        }, _callee, this);
      }));
      function initTileLayer() {
        return _initTileLayer.apply(this, arguments);
      }
      return initTileLayer;
    }()
  }, {
    key: "getSourceOption",
    value: function getSourceOption() {
      var rawSource = this.parent.getSource();
      var features = this.sourceTile.data.layers.testTile.features;
      return {
        data: {
          type: 'FeatureCollection',
          features: features
        },
        options: {
          parser: {
            type: 'geojson'
          },
          transforms: rawSource.transforms
        }
      };
    }
  }]);
  return DebugTile;
}(_Tile2.default);
exports.default = DebugTile;
