// @antv/l7-layers/lib/tile/tileFactory/VectorTile.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _Tile2 = _interopRequireDefault(require("./Tile"));
var _util = require("./util");
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }
function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
var VectorTile = /*#__PURE__*/function (_Tile) {
  (0, _inherits2.default)(VectorTile, _Tile);
  var _super = _createSuper(VectorTile);
  function VectorTile() {
    (0, _classCallCheck2.default)(this, VectorTile);
    return _super.apply(this, arguments);
  }
  (0, _createClass2.default)(VectorTile, [{
    key: "initTileLayer",
    value: function () {
      var _initTileLayer = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee() {
        var attributes, layerOptions, vectorLayer, sourceOptions, layer;
        return _regenerator.default.wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              attributes = this.parent.getLayerAttributeConfig();
              layerOptions = this.getLayerOptions();
              vectorLayer = (0, _util.getTileLayer)(this.parent.type);
              sourceOptions = this.getSourceOption();
              if (sourceOptions) {
                _context.next = 7;
                break;
              }
              this.isLoaded = true;
              return _context.abrupt("return");
            case 7:
              layer = new vectorLayer((0, _objectSpread2.default)({}, layerOptions)).source(sourceOptions.data, sourceOptions.options); // 初始化数据映射
              Object.keys(attributes).forEach(function (type) {
                var _attributes$attr, _attributes$attr2;
                var attr = type;
                // @ts-ignore
                layer[attr]((_attributes$attr = attributes[attr]) === null || _attributes$attr === void 0 ? void 0 : _attributes$attr.field, (_attributes$attr2 = attributes[attr]) === null || _attributes$attr2 === void 0 ? void 0 : _attributes$attr2.values);
              });
              _context.next = 11;
              return this.addLayer(layer);
            case 11:
              if (!layerOptions.tileMask) {
                _context.next = 14;
                break;
              }
              _context.next = 14;
              return this.addTileMask();
            case 14:
              this.setLayerMinMaxZoom(layer);
              this.isLoaded = true;
            case 16:
            case "end":
              return _context.stop();
          }
        }, _callee, this);
      }));
      function initTileLayer() {
        return _initTileLayer.apply(this, arguments);
      }
      return initTileLayer;
    }()
  }, {
    key: "getSourceOption",
    value: function getSourceOption() {
      var rawSource = this.parent.getSource();
      var _this$parent$getLayer = this.parent.getLayerConfig(),
        _this$parent$getLayer2 = _this$parent$getLayer.sourceLayer,
        sourceLayer = _this$parent$getLayer2 === void 0 ? 'defaultLayer' : _this$parent$getLayer2,
        _this$parent$getLayer3 = _this$parent$getLayer.featureId,
        featureId = _this$parent$getLayer3 === void 0 ? 'id' : _this$parent$getLayer3;
      var features = this.getFeatures(sourceLayer);
      return {
        data: {
          type: 'FeatureCollection',
          features: features
        },
        options: {
          parser: {
            type: 'geojson',
            featureId: featureId
          },
          transforms: rawSource.transforms
        }
      };
    }
  }, {
    key: "setLayerMinMaxZoom",
    value: function setLayerMinMaxZoom(layer) {
      // 文本图层设置，可见范围
      if (layer.getModelType() === 'text') {
        layer.updateLayerConfig({
          maxZoom: this.z + 1,
          minZoom: this.z - 1
        });
      }
    }
    // 获取瓦片数据
  }, {
    key: "getFeatures",
    value: function getFeatures(sourceLayer) {
      var source = this.sourceTile.data;
      return source.getTileData(sourceLayer);
    }

    /**
     * 在一个 Tile 中可能存在一个相同 ID 的 feature
     * @param id
     * @returns
     */
  }, {
    key: "getFeatureById",
    value: function getFeatureById(id) {
      var layer = this.getMainLayer();
      if (!layer) {
        return [];
      }
      var res = layer.getSource().data.dataArray.filter(function (d) {
        return d._id === id;
      });
      return res;
    }
  }]);
  return VectorTile;
}(_Tile2.default);
exports.default = VectorTile;
