// @ant-design/flowchart/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CanvasService = exports.GroupService = exports.EdgeService = exports.NodeService = exports.EditorPanels = exports.FormWrapper = exports.FormPanel = exports.createKeybindingConfig = exports.XFlowEdgeCommands = exports.useXFlowApp = exports.XFlowAppProvider = exports.IconStore = exports.FormItemWrapper = exports.usePanelContext = exports.XFlowGraphCommands = exports.XFlowNodeCommands = exports.WorkspacePanel = exports.Flowchart = exports.ToolbarPanel = void 0;
var xflow_1 = require("@antv/xflow");
Object.defineProperty(exports, "XFlowNodeCommands", { enumerable: true, get: function () { return xflow_1.XFlowNodeCommands; } });
Object.defineProperty(exports, "XFlowGraphCommands", { enumerable: true, get: function () { return xflow_1.XFlowGraphCommands; } });
Object.defineProperty(exports, "IconStore", { enumerable: true, get: function () { return xflow_1.IconStore; } });
Object.defineProperty(exports, "WorkspacePanel", { enumerable: true, get: function () { return xflow_1.WorkspacePanel; } });
Object.defineProperty(exports, "usePanelContext", { enumerable: true, get: function () { return xflow_1.usePanelContext; } });
Object.defineProperty(exports, "FormItemWrapper", { enumerable: true, get: function () { return xflow_1.FormItemWrapper; } });
Object.defineProperty(exports, "XFlowAppProvider", { enumerable: true, get: function () { return xflow_1.XFlowAppProvider; } });
Object.defineProperty(exports, "useXFlowApp", { enumerable: true, get: function () { return xflow_1.useXFlowApp; } });
Object.defineProperty(exports, "XFlowEdgeCommands", { enumerable: true, get: function () { return xflow_1.XFlowEdgeCommands; } });
Object.defineProperty(exports, "createKeybindingConfig", { enumerable: true, get: function () { return xflow_1.createKeybindingConfig; } });
Object.defineProperty(exports, "FormPanel", { enumerable: true, get: function () { return xflow_1.FlowchartFormPanel; } });
Object.defineProperty(exports, "FormWrapper", { enumerable: true, get: function () { return xflow_1.FlowchartFormWrapper; } });
Object.defineProperty(exports, "EditorPanels", { enumerable: true, get: function () { return xflow_1.EditorPanels; } });
// 临时方案
var graph_1 = __importDefault(require("./graph"));
exports.Flowchart = graph_1.default;
var toolbar_1 = require("./components/toolbar");
Object.defineProperty(exports, "ToolbarPanel", { enumerable: true, get: function () { return toolbar_1.ToolbarPanel; } });
var NodeService = xflow_1.FlowchartService.NodeService, EdgeService = xflow_1.FlowchartService.EdgeService, GroupService = xflow_1.FlowchartService.GroupService, CanvasService = xflow_1.FlowchartService.CanvasService;
exports.NodeService = NodeService;
exports.EdgeService = EdgeService;
exports.GroupService = GroupService;
exports.CanvasService = CanvasService;
__exportStar(require("./interface"), exports);
