// @ant-design/graphs/lib/utils/create-fetch-loading.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createFetchLoading = void 0;
var util_1 = require("@antv/util");
var constants_1 = require("../constants");
var create_node_1 = require("./create-node");
var set_styles_1 = require("./set-styles");
/** 开启加载动画， 不支持同时存在多个 */
var createFetchLoading = function (node, fetchLoading) {
    var loadingClassName = "".concat(constants_1.prefix, "-antd-loading");
    if (fetchLoading) {
        var loadingTemplate = (0, util_1.isFunction)(fetchLoading) ? fetchLoading(node) : fetchLoading;
        document.body.appendChild((0, create_node_1.createNode)(loadingTemplate, loadingClassName));
    }
    else {
        var container = document.createElement('div');
        container.className = loadingClassName;
        var styles = {
            position: 'fixed',
            left: '0',
            top: '0',
            width: '100vw',
            height: '100vh',
            display: 'flex',
            alignItems: 'center',
            justifyContent: 'center',
            background: 'rgba(0,0,0, 0.25)',
            color: '#fff',
            fontSize: '16px',
            zIndex: 999,
        };
        (0, set_styles_1.setStyles)(container, styles);
        var span = document.createElement('span');
        span.innerText = 'loading...';
        container.appendChild(span);
        document.body.appendChild(container);
    }
};
exports.createFetchLoading = createFetchLoading;
