// @antv/xflow-core/lib/xflow-main/graph/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createGraphConfig = exports.registerGraphConfig = exports.IGraphOptionProvider = exports.GraphConfig = exports.NsGraphConfig = void 0;
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importDefault(require("react"));
var mana_syringe_1 = require("mana-syringe");
var deferred_1 = require("../../common/deferred");
var constants_1 = require("../../constants");
var uuid_1 = require("../../common/uuid");
var react_group_node_1 = require("./react-group-node");
var react_node_1 = require("./react-node");
var NsGraphConfig;
(function (NsGraphConfig) {
    NsGraphConfig.CONFIG_TYPE = 'GraphConfig';
    /** 默认的Node解析函数 */
    NsGraphConfig.defaultNodeTypeParser = function (nodeConfig) { return nodeConfig.renderKey; };
    /** 构造的Edge解析函数 */
    NsGraphConfig.defaultEdgeTypeParser = function (edgeConfig) { return edgeConfig.renderKey; };
    /** XFlow默认的React组件 */
    NsGraphConfig.defaultNodeMapValue = [
        [constants_1.XFLOW_DEFAULT_NODE, react_node_1.XFlowDefaultNode],
        [constants_1.XFLOW_DEFAULT_GROUP_NODE, react_group_node_1.XFlowDefaultGroupNode],
    ];
})(NsGraphConfig = exports.NsGraphConfig || (exports.NsGraphConfig = {}));
var GraphConfig = /** @class */ (function () {
    /** 构造函数 */
    function GraphConfig() {
        var _this = this;
        /** 配置类型 */
        this.CONFIG_TYPE = NsGraphConfig.CONFIG_TYPE;
        /** 画布的属性 */
        this.x6Options = new deferred_1.Deferred();
        /** 画布的root节点 */
        this.rootContainer = new deferred_1.Deferred();
        /** 画布的dom节点 */
        this.graphContainer = new deferred_1.Deferred();
        /** 自定义节点 */
        this.nodeRender = new Map(NsGraphConfig.defaultNodeMapValue);
        /** 自定义边 */
        this.edgeRender = new Map();
        /** 解析node渲染类型的parser */
        this.nodeTypeParser = NsGraphConfig.defaultNodeTypeParser;
        /** 解析edge渲染类型的parser  */
        this.edgeTypeParser = NsGraphConfig.defaultEdgeTypeParser;
        /** 自定义事件 */
        this.events = [];
        this.mergeX6Config = function (config) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
            var current;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getX6Config()];
                    case 1:
                        current = _a.sent();
                        return [2 /*return*/, Object.assign(current, config)];
                }
            });
        }); };
        this.getX6Config = function () { return tslib_1.__awaiter(_this, void 0, void 0, function () {
            var config;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.x6Options.promise];
                    case 1:
                        config = _a.sent();
                        return [2 /*return*/, config];
                }
            });
        }); };
        this.setX6Config = function (options) {
            var _a;
            var defaultOptions = _this.getDefaultGraphOptions();
            (_a = _this.x6Options) === null || _a === void 0 ? void 0 : _a.resolve(tslib_1.__assign(tslib_1.__assign({}, defaultOptions), options));
        };
        this.setAppContainer = function (ele) {
            _this.appContainer = ele;
        };
        this.setRootContainer = function (ele) {
            ele && (_this === null || _this === void 0 ? void 0 : _this.rootContainer.resolve(ele));
        };
        this.setGraphContainer = function (ele) {
            var _a;
            ele && ((_a = _this.graphContainer) === null || _a === void 0 ? void 0 : _a.resolve(ele));
        };
        this.setEvents = function (events) {
            _this.events = events;
        };
        this.setDefaultNodeRender = function (component) {
            _this.nodeRender.set(constants_1.XFLOW_DEFAULT_NODE, component);
        };
        this.setDefaultEdgeRender = function (component) {
            _this.edgeRender.set(constants_1.XFLOW_DEFAULT_EDGE, component);
        };
        this.setNodeRender = function (renderKey, component) {
            _this.nodeRender.set(renderKey, component);
        };
        this.setEdgeRender = function (renderKey, component) {
            _this.edgeRender.set(renderKey, component);
        };
        /** 设置解析node render key的 parser */
        this.setNodeTypeParser = function (parser) {
            _this.nodeTypeParser = parser;
        };
        /** 设置解析edge render key的 parser */
        this.setEdgeTypeParser = function (parser) {
            _this.edgeTypeParser = parser;
        };
        /** 获取Graph参数 */
        this.getConfig = function () { return tslib_1.__awaiter(_this, void 0, void 0, function () {
            var _a, rootContainer, graphContainer, x6Options, graphConfig;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, Promise.all([
                            this.rootContainer.promise,
                            this.graphContainer.promise,
                            this.x6Options.promise,
                        ])];
                    case 1:
                        _a = _b.sent(), rootContainer = _a[0], graphContainer = _a[1], x6Options = _a[2];
                        graphConfig = {
                            xflowInstanceId: this.xflowInstanceId,
                            graphId: this.graphId,
                            nodeViewId: this.graphId,
                            appContainer: this.appContainer,
                            rootContainer: rootContainer,
                            graphContainer: graphContainer,
                            x6Options: x6Options,
                            events: this.events,
                            nodeRender: this.nodeRender,
                            edgeRender: this.edgeRender,
                            nodeTypeParser: this.nodeTypeParser,
                            edgeTypeParser: this.edgeTypeParser,
                        };
                        return [2 /*return*/, graphConfig];
                }
            });
        }); };
        this.getDefaultGraphOptions = function () {
            var defaultOptions = {
                // 节点是否可旋转
                rotating: false,
                // 节点是否可调整大小
                resizing: false,
                // 节点连线规则配置（详细文档：https://X6.antv.vision/zh/docs/api/graph/interaction#connecting）
                connecting: {
                    snap: true,
                    dangling: false,
                    highlight: false,
                    connectionPoint: 'rect',
                    router: { name: 'er' },
                    connector: {
                        name: 'rounded',
                        args: {
                            radius: 15,
                        },
                    },
                },
                // 画布背景，支持颜色/图片/水印等（详细文档：https://X6.antv.vision/zh/docs/tutorial/basic/background）
                background: {},
                // 网格配置（详细文档：https://X6.antv.vision/zh/docs/tutorial/basic/grid）
                grid: {
                    visible: true,
                },
                // 点选/框选配置（详细文档：https://X6.antv.vision/zh/docs/tutorial/basic/selection）
                selecting: {
                    strict: true,
                    enabled: true,
                    multiple: true,
                    selectCellOnMoved: true,
                    showNodeSelectionBox: false,
                    // 框选可以选中edge
                    rubberEdge: true,
                    // 框选可以选中node
                    rubberNode: true,
                    movable: true,
                },
                // 对齐线配置，辅助移动节点排版（详细文档：https://X6.antv.vision/zh/docs/tutorial/basic/snapline）
                snapline: {
                    enabled: false,
                },
                // 撤销/重做能力（详细文档：https://X6.antv.vision/zh/docs/tutorial/basic/history）
                history: {
                    enabled: false,
                },
                // 剪切板，支持跨画布的复制/粘贴（详细文档：https://X6.antv.vision/zh/docs/tutorial/basic/clipboard）
                clipboard: {
                    enabled: true,
                    useLocalStorage: true,
                },
                // 使画布具备滚动、平移、居中、缩放等能力（详细文档：https://X6.antv.vision/zh/docs/tutorial/basic/scroller）
                scroller: {
                    enabled: false,
                },
                // 滚轮缩放 （详细文档：https://X6.antv.vision/zh/docs/tutorial/basic/mousewheel）
                mousewheel: {
                    enabled: true,
                    minScale: constants_1.MIN_ZOOM,
                    maxScale: constants_1.MAX_ZOOM,
                    factor: 1.1,
                    modifiers: ['ctrl', 'meta'],
                },
                /** 普通画布, 支持拖拽平移 */
                panning: {
                    enabled: true,
                },
                /** 缩放参数 */
                scaling: {
                    min: constants_1.MIN_ZOOM,
                    max: constants_1.MAX_ZOOM,
                },
                /** keyboard */
                keyboard: {
                    enabled: true,
                },
                /** 定制节点和边的交互行为 */
                interacting: {
                    /** 节点默认可以被移动 */
                    nodeMovable: true,
                    /** 边上标签默认不可以被移动 */
                    edgeLabelMovable: false,
                },
                async: false,
            };
            return defaultOptions;
        };
        this.dispose = function () {
            // TODO: 统一看下defer的销毁
            _this.rootContainer = new deferred_1.Deferred();
            _this.graphContainer = new deferred_1.Deferred();
            _this.x6Options = new deferred_1.Deferred();
        };
        this.graphId = (0, uuid_1.uuidv4)();
    }
    return GraphConfig;
}());
exports.GraphConfig = GraphConfig;
exports.IGraphOptionProvider = Symbol('IGraphOptionProvider');
var registerGraphConfig = function (register, graphConfig) {
    register(exports.IGraphOptionProvider, {
        lifecycle: mana_syringe_1.Syringe.Lifecycle.singleton,
        useDynamic: function () {
            return {
                getOptions: function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
                    return tslib_1.__generator(this, function (_a) {
                        return [2 /*return*/, graphConfig.getConfig()];
                    });
                }); },
            };
        },
    });
};
exports.registerGraphConfig = registerGraphConfig;
var createGraphConfig = function (addOptions) {
    return function (props) {
        /** bridge config and props */
        var propsContainer = react_1.default.useMemo(function () { return ({ getValue: function () { return ({}); } }); }, []);
        propsContainer.getValue = function () { return props; };
        var graphConfig = react_1.default.useMemo(function () {
            var config = new GraphConfig();
            addOptions(config, propsContainer);
            return config;
        }, [propsContainer]);
        return graphConfig;
    };
};
exports.createGraphConfig = createGraphConfig;
