// @antv/xflow-core/lib/xflow-main/graph/graph-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerGraphModule = exports.GraphManager = exports.IModelServiceProvider = exports.ICommandServiceProvider = exports.IGraphProvider = void 0;
var tslib_1 = require("tslib");
var react_dom_1 = tslib_1.__importDefault(require("react-dom"));
var x6_1 = require("@antv/x6");
var disposable_1 = require("../../common/disposable");
var deferred_1 = require("../../common/deferred");
var config_1 = require("./config");
var mana_syringe_1 = require("mana-syringe");
var command_1 = require("../../command");
var interface_1 = require("../../model-service/interface");
var hooks_1 = require("../../hooks");
var context_1 = require("../../command-contributions/components/context");
var constants_1 = require("../../constants");
/** !!! 所有需要使用Graph相关信息的地方, 都统一使用IGraphProvider来获取 */
exports.IGraphProvider = Symbol('IGraphProvider');
exports.ICommandServiceProvider = Symbol('ICommandServiceProvider');
exports.IModelServiceProvider = Symbol('IModelServiceProvider');
var GraphManager = /** @class */ (function () {
    function GraphManager() {
        var _this = this;
        /** 处理画布实例的销毁 */
        this.toDispose = new disposable_1.DisposableCollection();
        /** 储存画布实例 */
        this.graphMap = new Map();
        /** 获取X6 Graph 实例 */
        this.getGraph = function (graphId) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
            var graphDefer, options_1, hooks_2, mergedOptions, graphContainer, edgeRender, nodeRender, edgeTypeParser_1, edgeRenderMap_1, clientHeight, clientWidth, commandService_1, modelService_1, graph_1;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        graphDefer = this.graphMap.get(graphId);
                        if (!!graphDefer) return [3 /*break*/, 8];
                        graphDefer = new deferred_1.Deferred();
                        this.graphMap.set(graphId, graphDefer);
                        return [4 /*yield*/, this.optionProvider.getOptions()
                            /** 获取hooks实例： hooks方便其他组件可以修改注册graphOpions/events */
                        ];
                    case 1:
                        options_1 = _a.sent();
                        hooks_2 = this.hookService.hookProvider();
                        return [4 /*yield*/, hooks_2.graphOptions.call(options_1.x6Options)];
                    case 2:
                        mergedOptions = _a.sent();
                        graphContainer = options_1.graphContainer, edgeRender = options_1.edgeRender, nodeRender = options_1.nodeRender, edgeTypeParser_1 = options_1.edgeTypeParser;
                        return [4 /*yield*/, hooks_2.reactEdgeLabelRender.call(edgeRender)
                            /** 执行hooks：更新nodeRender */
                        ];
                    case 3:
                        edgeRenderMap_1 = _a.sent();
                        /** 执行hooks：更新nodeRender */
                        return [4 /*yield*/, hooks_2.reactNodeRender.call(nodeRender)];
                    case 4:
                        /** 执行hooks：更新nodeRender */
                        _a.sent();
                        clientHeight = graphContainer.clientHeight, clientWidth = graphContainer.clientWidth;
                        return [4 /*yield*/, this.commandServiceProvider.getCommandService()];
                    case 5:
                        commandService_1 = _a.sent();
                        return [4 /*yield*/, this.modelServiceProvider.getModelService()
                            /** 实例化 X6 Graph */
                        ];
                    case 6:
                        modelService_1 = _a.sent();
                        graph_1 = new x6_1.Graph(tslib_1.__assign(tslib_1.__assign({ container: graphContainer, width: clientWidth, height: clientHeight }, mergedOptions), { 
                            /** X6提供了边渲染的钩子, 可以在这里设置边上需要渲染的React内容 */
                            onEdgeLabelRendered: function (args) {
                                var edge = args.edge, container = args.container;
                                _this.renderEdgeReactLabel({
                                    edge: edge,
                                    container: container,
                                    edgeRenderMap: edgeRenderMap_1,
                                    edgeTypeParser: edgeTypeParser_1,
                                    modelService: modelService_1,
                                    commandService: commandService_1,
                                });
                            } }));
                        /** 执行hooks：绑定事件执行api */
                        return [4 /*yield*/, hooks_2.afterGraphInit.call({
                                graph: graph_1,
                                commandService: commandService_1,
                                modelService: modelService_1,
                                options: options_1,
                            })];
                    case 7:
                        /** 执行hooks：绑定事件执行api */
                        _a.sent();
                        graphDefer.resolve(graph_1);
                        graph_1.on('node:moved', function (_a) {
                            var node = _a.node;
                            var nodeData = node.getData();
                            var position = node.position();
                            node.setData(tslib_1.__assign(tslib_1.__assign({}, nodeData), { x: position === null || position === void 0 ? void 0 : position.x, y: position === null || position === void 0 ? void 0 : position.y }));
                        });
                        graph_1.on('node:resized', function (_a) {
                            var node = _a.node;
                            var nodeData = node.getData();
                            var size = node.size();
                            node.setData(tslib_1.__assign(tslib_1.__assign({}, nodeData), { width: size === null || size === void 0 ? void 0 : size.width, height: size === null || size === void 0 ? void 0 : size.height }));
                        });
                        this.toDispose.push(disposable_1.Disposable.create(function () { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                            return tslib_1.__generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, hooks_2.beforeGraphDestroy.call({
                                            graph: graph_1,
                                            commandService: commandService_1,
                                            modelService: modelService_1,
                                            options: options_1,
                                        })];
                                    case 1:
                                        _a.sent();
                                        this.graphMap.delete(graphId);
                                        graph_1.dispose();
                                        return [2 /*return*/];
                                }
                            });
                        }); }));
                        _a.label = 8;
                    case 8: return [2 /*return*/, graphDefer.promise];
                }
            });
        }); };
        this.renderEdgeReactLabel = function (args) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
            var container, edgeTypeParser, edge, edgeRenderMap, commandService, modelService, renderKey, reactComponent, content, WrappedReactComponent;
            return tslib_1.__generator(this, function (_a) {
                container = args.container, edgeTypeParser = args.edgeTypeParser, edge = args.edge, edgeRenderMap = args.edgeRenderMap, commandService = args.commandService, modelService = args.modelService;
                renderKey = edgeTypeParser(edge === null || edge === void 0 ? void 0 : edge.data) || constants_1.XFLOW_DEFAULT_EDGE;
                reactComponent = edgeRenderMap.get(renderKey);
                if (!reactComponent) {
                    return [2 /*return*/];
                }
                if (edge && (edge === null || edge === void 0 ? void 0 : edge.data) && reactComponent) {
                    content = this.edgeAppendForeignObject(edge, container);
                    WrappedReactComponent = (0, context_1.getEdgeReactComponent)(reactComponent, commandService, modelService)(edge);
                    react_dom_1.default.render(WrappedReactComponent, content);
                }
                return [2 /*return*/];
            });
        }); };
    }
    /** 实现在连线上渲染React节点 */
    GraphManager.prototype.edgeAppendForeignObject = function (x6Edge, container) {
        var fo = x6_1.Dom.createSvgElement('foreignObject');
        var body = x6_1.Dom.createElementNS('body', x6_1.Dom.ns.xhtml);
        var content = x6_1.Dom.createElementNS('div', x6_1.Dom.ns.xhtml);
        var edgeData = x6Edge === null || x6Edge === void 0 ? void 0 : x6Edge.data;
        var foWidth = (edgeData === null || edgeData === void 0 ? void 0 : edgeData.edgeContentWidth) || 100;
        var foHeight = (edgeData === null || edgeData === void 0 ? void 0 : edgeData.edgeContentHeight) || 30;
        fo.setAttribute('width', "".concat(foWidth));
        fo.setAttribute('height', "".concat(foHeight));
        fo.setAttribute('x', "".concat((-1.0 * foWidth) / 2));
        fo.setAttribute('y', "".concat((-1.0 * foHeight) / 2));
        body.setAttribute('xhtmls', x6_1.Dom.ns.xhtml);
        body.style.width = '100%';
        body.style.height = '100%';
        body.style.padding = '0';
        body.style.margin = '0';
        body.style.background = 'unset';
        body.style.overflow = 'visible';
        body.className = 'xflow-edge-label-body';
        content.style.width = '100%';
        content.style.height = '100%';
        body.appendChild(content);
        fo.appendChild(body);
        container.appendChild(fo);
        return content;
    };
    GraphManager.prototype.dispose = function () {
        this.toDispose.dispose();
    };
    tslib_1.__decorate([
        (0, mana_syringe_1.inject)(config_1.IGraphOptionProvider),
        tslib_1.__metadata("design:type", Object)
    ], GraphManager.prototype, "optionProvider", void 0);
    tslib_1.__decorate([
        (0, mana_syringe_1.inject)(hooks_1.IHookService),
        tslib_1.__metadata("design:type", Object)
    ], GraphManager.prototype, "hookService", void 0);
    tslib_1.__decorate([
        (0, mana_syringe_1.inject)(exports.ICommandServiceProvider),
        tslib_1.__metadata("design:type", Object)
    ], GraphManager.prototype, "commandServiceProvider", void 0);
    tslib_1.__decorate([
        (0, mana_syringe_1.inject)(exports.IModelServiceProvider),
        tslib_1.__metadata("design:type", Object)
    ], GraphManager.prototype, "modelServiceProvider", void 0);
    GraphManager = tslib_1.__decorate([
        (0, mana_syringe_1.singleton)()
    ], GraphManager);
    return GraphManager;
}());
exports.GraphManager = GraphManager;
var registerGraphModule = function (register) {
    /** 注册GraphManager */
    register(GraphManager);
    register(exports.IGraphProvider, {
        lifecycle: mana_syringe_1.Syringe.Lifecycle.singleton,
        useDynamic: function (context) {
            return {
                getGraphInstance: function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
                    var graphManager, graphOptionProvider, graphConfig, x6Graph;
                    return tslib_1.__generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                graphManager = context.container.get(GraphManager);
                                graphOptionProvider = context.container.get(config_1.IGraphOptionProvider);
                                return [4 /*yield*/, (graphOptionProvider === null || graphOptionProvider === void 0 ? void 0 : graphOptionProvider.getOptions())];
                            case 1:
                                graphConfig = _a.sent();
                                return [4 /*yield*/, (graphManager === null || graphManager === void 0 ? void 0 : graphManager.getGraph(graphConfig.graphId))];
                            case 2:
                                x6Graph = _a.sent();
                                return [2 /*return*/, x6Graph];
                        }
                    });
                }); },
                getGraphOptions: function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
                    var graphOptionProvider;
                    return tslib_1.__generator(this, function (_a) {
                        graphOptionProvider = context.container.get(config_1.IGraphOptionProvider);
                        return [2 /*return*/, graphOptionProvider === null || graphOptionProvider === void 0 ? void 0 : graphOptionProvider.getOptions()];
                    });
                }); },
            };
        },
    });
    register(exports.ICommandServiceProvider, {
        lifecycle: mana_syringe_1.Syringe.Lifecycle.singleton,
        useDynamic: function (context) {
            return {
                getCommandService: function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
                    var commandService;
                    return tslib_1.__generator(this, function (_a) {
                        commandService = context.container.get(command_1.IGraphCommandService);
                        return [2 /*return*/, commandService];
                    });
                }); },
            };
        },
    });
    register(exports.IModelServiceProvider, {
        lifecycle: mana_syringe_1.Syringe.Lifecycle.singleton,
        useDynamic: function (context) {
            return {
                getModelService: function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
                    var modelService;
                    return tslib_1.__generator(this, function (_a) {
                        modelService = context.container.get(interface_1.IModelService);
                        return [2 /*return*/, modelService];
                    });
                }); },
            };
        },
    });
};
exports.registerGraphModule = registerGraphModule;
