// antd-token-previewer/es/component-demos/form/form.js
import React from 'react';
import { Form, Input, Button, Checkbox } from 'antd';

var Demo = function Demo() {
  var onFinish = function onFinish() {};

  var onFinishFailed = function onFinishFailed() {};

  return /*#__PURE__*/React.createElement(Form, {
    name: "basic",
    labelCol: {
      span: 8
    },
    wrapperCol: {
      span: 16
    },
    initialValues: {
      remember: true
    },
    onFinish: onFinish,
    onFinishFailed: onFinishFailed,
    autoComplete: "off"
  }, /*#__PURE__*/React.createElement(Form.Item, {
    label: "Username",
    name: "username",
    rules: [{
      required: true,
      message: 'Please input your username!'
    }]
  }, /*#__PURE__*/React.createElement(Input, null)), /*#__PURE__*/React.createElement(Form.Item, {
    label: "Password",
    name: "password",
    rules: [{
      required: true,
      message: 'Please input your password!'
    }]
  }, /*#__PURE__*/React.createElement(Input.Password, null)), /*#__PURE__*/React.createElement(Form.Item, {
    name: "remember",
    valuePropName: "checked",
    wrapperCol: {
      offset: 8,
      span: 16
    }
  }, /*#__PURE__*/React.createElement(Checkbox, null, "Remember me")), /*#__PURE__*/React.createElement(Form.Item, {
    wrapperCol: {
      offset: 8,
      span: 16
    }
  }, /*#__PURE__*/React.createElement(Button, {
    type: "primary",
    htmlType: "submit"
  }, "Submit")));
};

var componentDemo = {
  demo: /*#__PURE__*/React.createElement(Demo, null),
  tokens: ['colorError', 'controlOutline', 'colorErrorBorder', 'colorErrorHover'],
  key: 'default'
};
export default componentDemo;
