// antd-token-previewer/es/component-demos/index.js
import Alert from "./alert";
import Anchor from "./anchor";
import AutoComplete from "./autoComplete";
import Avatar from "./avatar";
import Badge from "./badge";
import Breadcrumb from "./breadcrumb";
import Button from "./button";
import Calendar from "./calendar";
import Card from "./card";
import Carousel from "./carousel";
import Cascader from "./cascader";
import Checkbox from "./checkbox";
import Collapse from "./collapse";
import DatePicker from "./datePicker";
import Descriptions from "./descriptions";
import Dropdown from "./dropdown";
import Empty from "./empty";
import Form from "./form";
import Grid from "./grid";
import Icon from "./icon";
import Image from "./image";
import InputNumber from "./inputNumber";
import Input from "./input";
import List from "./list";
import Mentions from "./mentions";
import Modal from "./modal";
import Notification from "./notification";
import Pagination from "./pagination";
import Popconfirm from "./popconfirm";
import Popover from "./popover";
import Radio from "./radio";
import Rate from "./rate";
import Select from "./select";
import Skeleton from "./skeleton";
import Slider from "./slider";
import Spin from "./spin";
import Statistic from "./statistic";
import Switch from "./switch";
import Table from "./table";
import Tabs from "./tabs";
import Tag from "./tag";
import TimePicker from "./timePicker";
import Timeline from "./timeline";
import Tooltip from "./tooltip";
import Transfer from "./transfer";
import TreeSelect from "./treeSelect";
import Tree from "./tree";
import Typography from "./typography";
import Upload from "./upload";
import Divider from "./divider";
import Space from "./space";
import Menu from "./menu";
import Steps from "./steps";
import Segmented from "./segmented";
import Drawer from "./drawer";
import Message from "./message";
import Progress from "./progress";
import Result from "./result";
var ComponentDemos = {
  Alert: Alert,
  Anchor: Anchor,
  AutoComplete: AutoComplete,
  Avatar: Avatar,
  Badge: Badge,
  Breadcrumb: Breadcrumb,
  Button: Button,
  Calendar: Calendar,
  Card: Card,
  Carousel: Carousel,
  Cascader: Cascader,
  Checkbox: Checkbox,
  Collapse: Collapse,
  DatePicker: DatePicker,
  Descriptions: Descriptions,
  Dropdown: Dropdown,
  Empty: Empty,
  Form: Form,
  Grid: Grid,
  Icon: Icon,
  Image: Image,
  InputNumber: InputNumber,
  Input: Input,
  List: List,
  Mentions: Mentions,
  Modal: Modal,
  Notification: Notification,
  Pagination: Pagination,
  Popconfirm: Popconfirm,
  Popover: Popover,
  Radio: Radio,
  Rate: Rate,
  Select: Select,
  Skeleton: Skeleton,
  Slider: Slider,
  Spin: Spin,
  Statistic: Statistic,
  Switch: Switch,
  Table: Table,
  Tabs: Tabs,
  Tag: Tag,
  TimePicker: TimePicker,
  Timeline: Timeline,
  Tooltip: Tooltip,
  Transfer: Transfer,
  TreeSelect: TreeSelect,
  Tree: Tree,
  Typography: Typography,
  Upload: Upload,
  Divider: Divider,
  Space: Space,
  Menu: Menu,
  Steps: Steps,
  Segmented: Segmented,
  Drawer: Drawer,
  Message: Message,
  Result: Result,
  Progress: Progress
};
export default ComponentDemos;
