// @ant-design/graphs/lib/constants/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultStatusBarWidth = exports.prefix = exports.defaultMargin = exports.defaultCardStyle = exports.defaultMinimapCfg = exports.defaultLineLabelStyle = exports.defaultLabelStyle = exports.defaultFlowGraphAnchorPoints = exports.defaultNodeStyle = exports.defaultNodeSize = exports.defaultStateStyles = void 0;
// 默认交互状态
exports.defaultStateStyles = {
    hover: {
        stroke: '#1890ff',
        lineWidth: 2,
    },
};
// card 默认节点大小
exports.defaultNodeSize = [120, 40];
// 默认节点样式
exports.defaultNodeStyle = {
    stroke: '#40a9ff',
};
// 默认 anchor 连接点
exports.defaultFlowGraphAnchorPoints = [
    [0, 0.5],
    [1, 0.5],
];
// card body|footer  默认样式
exports.defaultLabelStyle = {
    fill: '#000',
    fontSize: 12,
};
exports.defaultLineLabelStyle = {
    fill: '#000',
    fontSize: 12,
    textAlign: 'start',
    textBaseline: 'middle',
};
// 缩略图默认配置
exports.defaultMinimapCfg = {
    show: false,
    size: [150, 100],
    type: 'keyShape',
};
// card 默认样式
exports.defaultCardStyle = {
    fill: '#fff',
    stroke: '#40a9ff',
    radius: 2,
};
// card 内部 padding | margin | 行间距
exports.defaultMargin = 6;
// 打标前缀
exports.prefix = 'g';
// status bar 的默认宽度
exports.defaultStatusBarWidth = 4;
