// antd-token-previewer/lib/token-panel-pro/ComponentDemoPro.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _antd = require("antd");

var _react = _interopRequireDefault(require("react"));

var _ComponentDemoGroup = _interopRequireDefault(require("../component-panel/ComponentDemoGroup"));

var _locale = require("../locale");

var _overviews = require("../overviews");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var ComponentDemoPro = function ComponentDemoPro(_ref) {
  var selectedTokens = _ref.selectedTokens,
      theme = _ref.theme,
      components = _ref.components,
      activeComponents = _ref.activeComponents,
      componentDrawer = _ref.componentDrawer,
      showAll = _ref.showAll,
      style = _ref.style;

  var _React$useState = _react.default.useState('overview'),
      _React$useState2 = (0, _slicedToArray2.default)(_React$useState, 2),
      mode = _React$useState2[0],
      setMode = _React$useState2[1];

  var _antdTheme$useToken = _antd.theme.useToken(),
      colorBgLayout = _antdTheme$useToken.token.colorBgLayout;

  var locale = (0, _locale.useLocale)();

  var overviewDemo = _react.default.useMemo(function () {
    if (showAll) {
      return /*#__PURE__*/_react.default.createElement(_antd.Space, {
        direction: "vertical"
      }, /*#__PURE__*/_react.default.createElement(_overviews.Primary, null), /*#__PURE__*/_react.default.createElement(_overviews.Success, null), /*#__PURE__*/_react.default.createElement(_overviews.Error, null), /*#__PURE__*/_react.default.createElement(_overviews.Warning, null));
    }

    if (selectedTokens !== null && selectedTokens !== void 0 && selectedTokens.includes('colorError')) {
      return /*#__PURE__*/_react.default.createElement(_overviews.Error, null);
    }

    if (selectedTokens !== null && selectedTokens !== void 0 && selectedTokens.includes('colorSuccess')) {
      return /*#__PURE__*/_react.default.createElement(_overviews.Success, null);
    }

    if (selectedTokens !== null && selectedTokens !== void 0 && selectedTokens.includes('colorWarning')) {
      return /*#__PURE__*/_react.default.createElement(_overviews.Warning, null);
    }

    return /*#__PURE__*/_react.default.createElement(_overviews.Primary, null);
  }, [selectedTokens, showAll]);

  return /*#__PURE__*/_react.default.createElement("div", {
    style: _objectSpread(_objectSpread({}, style), {}, {
      background: colorBgLayout,
      paddingBottom: 24
    })
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      margin: 'auto',
      maxWidth: 960
    }
  }, /*#__PURE__*/_react.default.createElement(_antd.Segmented, {
    options: [{
      value: 'overview',
      label: locale.demo.overview
    }, {
      value: 'component',
      label: locale.demo.components
    }],
    value: mode,
    onChange: setMode,
    style: {
      margin: '12px 0 0 12px'
    }
  }), /*#__PURE__*/_react.default.createElement(_antd.ConfigProvider, {
    theme: {
      components: {
        Select: {
          zIndexPopup: 10
        },
        DatePicker: {
          zIndexPopup: 10
        },
        Dropdown: {
          zIndexPopup: 10
        },
        Mentions: {
          zIndexPopup: 10
        },
        Tooltip: {
          zIndexPopup: 10
        },
        Popover: {
          zIndexPopup: 10
        },
        Popconfirm: {
          zIndexPopup: 10
        }
      }
    }
  }, mode === 'overview' ? /*#__PURE__*/_react.default.createElement("div", {
    style: {
      margin: 12,
      maxWidth: 'fit-content'
    }
  }, overviewDemo) : /*#__PURE__*/_react.default.createElement(_ComponentDemoGroup.default, {
    selectedTokens: selectedTokens,
    themes: [theme],
    components: components,
    activeComponents: activeComponents,
    componentDrawer: componentDrawer,
    hideTokens: true
  }))));
};

var _default = function _default(props) {
  return /*#__PURE__*/_react.default.createElement(_antd.ConfigProvider, {
    theme: props.theme.config
  }, /*#__PURE__*/_react.default.createElement(ComponentDemoPro, props));
};

exports.default = _default;
