// antd-token-previewer/lib/token-panel-pro/TokenDetail.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _antd = require("antd");

var _tokenMeta = _interopRequireDefault(require("antd/lib/version/token-meta.json"));

var _classnames = _interopRequireDefault(require("classnames"));

var _react = _interopRequireWildcard(require("react"));

var _locale = require("../locale");

var _TokenRelation = require("../meta/TokenRelation");

var _TokenInput = _interopRequireDefault(require("../TokenInput"));

var _deepUpdateObj = _interopRequireDefault(require("../utils/deepUpdateObj"));

var _getDesignToken = _interopRequireDefault(require("../utils/getDesignToken"));

var _getValueByPath2 = _interopRequireDefault(require("../utils/getValueByPath"));

var _makeStyle = _interopRequireDefault(require("../utils/makeStyle"));

var _statistic = require("../utils/statistic");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var useStyle = (0, _makeStyle.default)('TokenDetail', function (token) {
  return {
    '.token-panel-token-detail': {
      '.token-panel-pro-token-collapse-map-collapse-token-description': {
        color: token.colorTextPlaceholder,
        marginBottom: 8,
        fontSize: 12
      },
      '.token-panel-pro-token-collapse-map-collapse-token-usage-tag-container': {
        textOverflow: 'ellipsis',
        whiteSpace: 'nowrap',
        overflow: 'hidden',
        color: token.colorTextSecondary
      },
      '.token-panel-pro-token-collapse-map-collapse-token-usage-tag': {
        display: 'inline-block',
        marginInlineEnd: 8,
        borderRadius: 4,
        height: 20,
        padding: '0 8px',
        fontSize: 12,
        lineHeight: '20px',
        backgroundColor: 'rgba(0,0,0,0.015)'
      },
      '.token-panel-pro-token-collapse-map-collapse-token-inputs': {
        padding: '8px 10px',
        backgroundColor: 'rgba(0,0,0,0.02)',
        marginTop: 12,
        '> *:not(:last-child)': {
          marginBottom: 8
        }
      }
    }
  };
});

var TokenDetail = function TokenDetail(_ref) {
  var _tokenName2;

  var themes = _ref.themes,
      path = _ref.path,
      tokenName = _ref.tokenName,
      className = _ref.className,
      style = _ref.style;

  var _useStyle = useStyle(),
      _useStyle2 = (0, _slicedToArray2.default)(_useStyle, 2),
      wrapSSR = _useStyle2[0],
      hashId = _useStyle2[1];

  var tokenPath = [].concat((0, _toConsumableArray2.default)(path), [tokenName]);
  var locale = (0, _locale.useLocale)();

  var handleTokenChange = function handleTokenChange(theme) {
    return function (value) {
      var _theme$onThemeChange;

      (_theme$onThemeChange = theme.onThemeChange) === null || _theme$onThemeChange === void 0 ? void 0 : _theme$onThemeChange.call(theme, (0, _deepUpdateObj.default)(theme.config, [].concat((0, _toConsumableArray2.default)(path), [tokenName]), value), [].concat((0, _toConsumableArray2.default)(path), [tokenName]));
    };
  };

  var relatedComponents = (0, _react.useMemo)(function () {
    var _tokenName;

    return (0, _statistic.getRelatedComponents)([tokenName].concat((0, _toConsumableArray2.default)((_tokenName = _TokenRelation.mapRelatedAlias[tokenName]) !== null && _tokenName !== void 0 ? _tokenName : [])));
  }, [tokenName]);
  return wrapSSR( /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)(className, hashId, 'token-panel-token-detail'),
    style: style
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "token-panel-pro-token-collapse-map-collapse-token-description"
  }, (_tokenName2 = _tokenMeta.default[tokenName]) === null || _tokenName2 === void 0 ? void 0 : _tokenName2[locale._lang === 'zh-CN' ? 'desc' : 'descEn']), relatedComponents.length > 0 && /*#__PURE__*/_react.default.createElement(_antd.Tooltip, {
    title: (0, _statistic.getRelatedComponents)(tokenName).join(', '),
    placement: "topLeft"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "token-panel-pro-token-collapse-map-collapse-token-usage-tag-container"
  }, relatedComponents.map(function (item) {
    return /*#__PURE__*/_react.default.createElement("span", {
      key: item,
      className: "token-panel-pro-token-collapse-map-collapse-token-usage-tag"
    }, item);
  }))), /*#__PURE__*/_react.default.createElement("div", {
    className: "token-panel-pro-token-collapse-map-collapse-token-inputs"
  }, themes.map(function (themeItem) {
    var _themeItem$getCanRese, _getValueByPath;

    return /*#__PURE__*/_react.default.createElement("div", {
      key: themeItem.key
    }, /*#__PURE__*/_react.default.createElement(_TokenInput.default, {
      hideTheme: themes.length === 1,
      theme: themeItem,
      canReset: (_themeItem$getCanRese = themeItem.getCanReset) === null || _themeItem$getCanRese === void 0 ? void 0 : _themeItem$getCanRese.call(themeItem, tokenPath),
      onReset: function onReset() {
        var _themeItem$onReset;

        return (_themeItem$onReset = themeItem.onReset) === null || _themeItem$onReset === void 0 ? void 0 : _themeItem$onReset.call(themeItem, tokenPath);
      },
      onChange: handleTokenChange(themeItem),
      value: (_getValueByPath = (0, _getValueByPath2.default)(themeItem.config, tokenPath)) !== null && _getValueByPath !== void 0 ? _getValueByPath : (0, _getDesignToken.default)(themeItem.config)[tokenName]
    }));
  }))));
};

var _default = TokenDetail;
exports.default = _default;
