// antd-token-previewer/lib/component-demos/datePicker/disabled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _antd = require("antd");

var Demo = function Demo() {
  return /*#__PURE__*/_react.default.createElement(_antd.Space, {
    direction: "vertical"
  }, /*#__PURE__*/_react.default.createElement(_antd.DatePicker, {
    disabled: true
  }), /*#__PURE__*/_react.default.createElement(_antd.DatePicker, {
    disabled: true,
    picker: "week"
  }), /*#__PURE__*/_react.default.createElement(_antd.DatePicker, {
    disabled: true,
    picker: "month"
  }), /*#__PURE__*/_react.default.createElement(_antd.DatePicker, {
    disabled: true,
    picker: "quarter"
  }), /*#__PURE__*/_react.default.createElement(_antd.DatePicker, {
    disabled: true,
    picker: "year"
  }));
};

var componentDemo = {
  demo: /*#__PURE__*/_react.default.createElement(Demo, null),
  tokens: ['colorBgContainerDisabled', 'colorTextDisabled'],
  key: 'disabled'
};
var _default = componentDemo;
exports.default = _default;
