// @antv/xflow-core/lib/model-service/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createModelServiceConfig = exports.ModelRegistry = exports.ModelServiceRegistry = exports.ModelServiceConfig = exports.IModelContribution = exports.IModelService = exports.MODELS = void 0;
var tslib_1 = require("tslib");
var interface_1 = require("./interface");
Object.defineProperty(exports, "IModelService", { enumerable: true, get: function () { return interface_1.IModelService; } });
Object.defineProperty(exports, "IModelContribution", { enumerable: true, get: function () { return interface_1.IModelContribution; } });
var config_1 = require("./config");
Object.defineProperty(exports, "ModelServiceConfig", { enumerable: true, get: function () { return config_1.ModelServiceConfig; } });
var model_registry_1 = require("./model-registry");
Object.defineProperty(exports, "ModelRegistry", { enumerable: true, get: function () { return model_registry_1.ModelRegistry; } });
var components_1 = require("./components");
Object.defineProperty(exports, "ModelServiceRegistry", { enumerable: true, get: function () { return components_1.ModelServiceRegistry; } });
Object.defineProperty(exports, "createModelServiceConfig", { enumerable: true, get: function () { return components_1.createModelServiceConfig; } });
var MODELS = tslib_1.__importStar(require("./constant"));
exports.MODELS = MODELS;
