// antd-token-previewer/lib/ColorPreview.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _makeStyle = _interopRequireDefault(require("./utils/makeStyle"));

var _getColorBgImg = _interopRequireDefault(require("./utils/getColorBgImg"));

var _excluded = ["color", "style", "className", "dark"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var useStyle = (0, _makeStyle.default)('ColorPreview', function () {
  return {
    '.previewer-color-preview': {
      width: 20,
      height: 20,
      position: 'relative',
      borderRadius: '50%',
      padding: 0,
      display: 'inline-block',
      '&::before': {
        content: '""',
        width: '100%',
        height: '100%',
        borderRadius: '50%',
        top: 0,
        insetInlineStart: 0,
        position: 'absolute',
        zIndex: 2,
        backgroundColor: 'var(--antd-token-previewer-color-preview)',
        boxShadow: '0 2px 3px -1px rgba(0,0,0,0.20), inset 0 0 0 1px rgba(0,0,0,0.09)'
      }
    }
  };
});

var ColorPreview = function ColorPreview(_ref) {
  var color = _ref.color,
      style = _ref.style,
      className = _ref.className,
      dark = _ref.dark,
      restProps = (0, _objectWithoutProperties2.default)(_ref, _excluded);

  var _useStyle = useStyle(),
      _useStyle2 = (0, _slicedToArray2.default)(_useStyle, 2),
      warpSSR = _useStyle2[0],
      hashId = _useStyle2[1];

  return warpSSR( /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({}, restProps, {
    className: (0, _classnames.default)('previewer-color-preview', className, hashId),
    style: _objectSpread((0, _defineProperty2.default)({}, '--antd-token-previewer-color-preview', color), style)
  }), /*#__PURE__*/_react.default.createElement("div", {
    style: {
      content: '""',
      width: 18,
      height: 18,
      borderRadius: '50%',
      top: 1,
      insetInlineStart: 1,
      position: 'absolute',
      zIndex: 1,
      background: "".concat((0, _getColorBgImg.default)(dark), " 0% 0% / 20px")
    }
  })));
};

var _default = ColorPreview;
exports.default = _default;
