// antd-token-previewer/lib/ThemeEditor.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _classnames = _interopRequireDefault(require("classnames"));

var _react = _interopRequireWildcard(require("react"));

var _componentPanel = require("./component-panel");

var _useControlledTheme2 = _interopRequireDefault(require("./hooks/useControlledTheme"));

var _locale = require("./locale");

var _TokenRelation = require("./meta/TokenRelation");

var _tokenPanelPro = _interopRequireDefault(require("./token-panel-pro"));

var _ComponentDemoPro = _interopRequireDefault(require("./token-panel-pro/ComponentDemoPro"));

var _makeStyle = _interopRequireDefault(require("./utils/makeStyle"));

var _statistic = require("./utils/statistic");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var useStyle = (0, _makeStyle.default)('ThemeEditor', function (token) {
  return {
    '.antd-theme-editor': {
      backgroundColor: token.colorBgLayout,
      display: 'flex'
    }
  };
});
var defaultTheme = {
  name: '默认主题',
  key: 'default',
  config: {}
};
var ThemeEditor = /*#__PURE__*/(0, _react.forwardRef)(function (_ref, ref) {
  var customTheme = _ref.theme,
      onThemeChange = _ref.onThemeChange,
      className = _ref.className,
      style = _ref.style,
      darkAlgorithm = _ref.darkAlgorithm,
      _ref$locale = _ref.locale,
      locale = _ref$locale === void 0 ? _locale.zhCN : _ref$locale;

  var _useStyle = useStyle(),
      _useStyle2 = (0, _slicedToArray2.default)(_useStyle, 2),
      wrapSSR = _useStyle2[0],
      hashId = _useStyle2[1];

  var _useState = (0, _react.useState)({
    seed: ['colorPrimary']
  }),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      selectedTokens = _useState2[0],
      setSelectedTokens = _useState2[1];

  var _useState3 = (0, _react.useState)(false),
      _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
      aliasOpen = _useState4[0],
      setAliasOpen = _useState4[1];

  var _useControlledTheme = (0, _useControlledTheme2.default)({
    theme: customTheme,
    defaultTheme: defaultTheme,
    onChange: onThemeChange,
    darkAlgorithm: darkAlgorithm
  }),
      theme = _useControlledTheme.theme,
      infoFollowPrimary = _useControlledTheme.infoFollowPrimary,
      onInfoFollowPrimaryChange = _useControlledTheme.onInfoFollowPrimaryChange,
      updateRef = _useControlledTheme.updateRef;

  (0, _react.useImperativeHandle)(ref, function () {
    return {
      updateRef: updateRef
    };
  });

  var handleTokenSelect = function handleTokenSelect(token, type) {
    setSelectedTokens(function (prev) {
      var tokens = typeof token === 'string' ? token ? [token] : [] : token;

      if (type === 'seed') {
        return {
          seed: tokens
        };
      }

      var newSelectedTokens = _objectSpread({}, prev);

      tokens.forEach(function (newToken) {
        var _prev$type, _prev$type2, _prev$type3;

        newSelectedTokens = _objectSpread(_objectSpread({}, prev), {}, (0, _defineProperty2.default)({}, type, (_prev$type = prev[type]) !== null && _prev$type !== void 0 && _prev$type.includes(newToken) ? (_prev$type2 = prev[type]) === null || _prev$type2 === void 0 ? void 0 : _prev$type2.filter(function (t) {
          return t !== newToken;
        }) : [].concat((0, _toConsumableArray2.default)((_prev$type3 = prev[type]) !== null && _prev$type3 !== void 0 ? _prev$type3 : []), [newToken])));
      });

      if (type === 'map') {
        delete newSelectedTokens.alias;
      }

      return newSelectedTokens;
    });
  };

  var computedSelectedTokens = (0, _react.useMemo)(function () {
    var _selectedTokens$seed, _selectedTokens$map, _selectedTokens$alias, _selectedTokens$map2, _selectedTokens$alias2, _selectedTokens$alias3;

    if ((_selectedTokens$seed = selectedTokens.seed) !== null && _selectedTokens$seed !== void 0 && _selectedTokens$seed.length && !((_selectedTokens$map = selectedTokens.map) !== null && _selectedTokens$map !== void 0 && _selectedTokens$map.length) && !((_selectedTokens$alias = selectedTokens.alias) !== null && _selectedTokens$alias !== void 0 && _selectedTokens$alias.length)) {
      var _selectedTokens$seed$, _selectedTokens$seed$2;

      return [].concat((0, _toConsumableArray2.default)(selectedTokens.seed), (0, _toConsumableArray2.default)((_selectedTokens$seed$ = _TokenRelation.seedRelatedMap[selectedTokens.seed[0]]) !== null && _selectedTokens$seed$ !== void 0 ? _selectedTokens$seed$ : []), (0, _toConsumableArray2.default)((_selectedTokens$seed$2 = _TokenRelation.seedRelatedAlias[selectedTokens.seed[0]]) !== null && _selectedTokens$seed$2 !== void 0 ? _selectedTokens$seed$2 : []));
    }

    if ((_selectedTokens$map2 = selectedTokens.map) !== null && _selectedTokens$map2 !== void 0 && _selectedTokens$map2.length && !((_selectedTokens$alias2 = selectedTokens.alias) !== null && _selectedTokens$alias2 !== void 0 && _selectedTokens$alias2.length)) {
      return [].concat((0, _toConsumableArray2.default)(selectedTokens.map), (0, _toConsumableArray2.default)(selectedTokens.map.reduce(function (result, item) {
        return result.concat(_TokenRelation.mapRelatedAlias[item]);
      }, [])));
    }

    if ((_selectedTokens$alias3 = selectedTokens.alias) !== null && _selectedTokens$alias3 !== void 0 && _selectedTokens$alias3.length) {
      return (0, _toConsumableArray2.default)(selectedTokens.alias);
    }

    return [];
  }, [selectedTokens]);
  var relatedComponents = (0, _react.useMemo)(function () {
    return computedSelectedTokens ? (0, _statistic.getRelatedComponents)(computedSelectedTokens) : [];
  }, [computedSelectedTokens]);
  return wrapSSR( /*#__PURE__*/_react.default.createElement(_locale.LocaleContext.Provider, {
    value: locale
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)(hashId, 'antd-theme-editor', className),
    style: style
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      flex: aliasOpen ? '0 0 860px' : "0 0 ".concat(860 - 320, "px"),
      height: '100%',
      backgroundColor: '#F7F8FA',
      backgroundImage: 'linear-gradient(180deg, #FFFFFF 0%, rgba(246,247,249,0.00) 100%)',
      display: 'flex',
      transition: 'all 0.3s'
    }
  }, /*#__PURE__*/_react.default.createElement(_tokenPanelPro.default, {
    aliasOpen: aliasOpen,
    onAliasOpenChange: function onAliasOpenChange(open) {
      return setAliasOpen(open);
    },
    theme: theme,
    style: {
      flex: 1
    },
    selectedTokens: selectedTokens,
    onTokenSelect: handleTokenSelect,
    infoFollowPrimary: infoFollowPrimary,
    onInfoFollowPrimaryChange: onInfoFollowPrimaryChange
  })), /*#__PURE__*/_react.default.createElement(_ComponentDemoPro.default, {
    theme: theme,
    components: _componentPanel.antdComponents,
    activeComponents: relatedComponents,
    selectedTokens: computedSelectedTokens,
    style: {
      flex: 1,
      overflow: 'auto',
      height: '100%'
    },
    componentDrawer: true
  }))));
});
var _default = ThemeEditor;
exports.default = _default;
