// antd-token-previewer/es/component-panel/ComponentTree.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import { SearchOutlined } from '@ant-design/icons';
import { Badge, Input, Tree } from 'antd';
import classNames from 'classnames';
import React, { useEffect, useMemo, useRef, useState } from 'react';
import makeStyle from "../utils/makeStyle";
import { getRelatedComponents } from "../utils/statistic";
var DirectoryTree = Tree.DirectoryTree;
var useStyle = makeStyle('ComponentTree', function (token) {
  var _$concat;

  return {
    '.component-tree-wrapper': _defineProperty({
      minWidth: 200,
      borderInlineEnd: "".concat(token.lineWidth, "px ").concat(token.lineType, " ").concat(token.colorSplit),
      height: '100%',
      overflow: 'hidden',
      display: 'flex',
      flexDirection: 'column',
      paddingBlock: token.paddingXS,
      '.component-tree-search': {
        margin: '0 8px 12px',
        width: 'calc(100% - 16px)',
        backgroundColor: 'rgba(0, 0, 0, 2%)',
        borderRadius: token.borderRadiusLG,
        height: 24,
        input: {
          fontSize: 12
        },
        '&:hover': {
          backgroundColor: 'rgba(0, 0, 0, 4%)'
        }
      }
    }, "".concat(token.rootCls, "-tree.component-tree"), (_$concat = {
      fontSize: token.fontSizeSM,
      '.component-tree-item.component-tree-item-highlight': {
        color: token.colorPrimary
      }
    }, _defineProperty(_$concat, "".concat(token.rootCls, "-tree-node-content-wrapper"), {
      transition: "background-color ".concat(token.motionDurationSlow),
      borderRadius: 4
    }), _defineProperty(_$concat, "".concat(token.rootCls, "-tree-treenode-selected ").concat(token.rootCls, "-tree-node-content-wrapper"), {
      color: token.colorTextLightSolid,
      '.component-tree-item.component-tree-item-highlight': {
        color: token.colorTextLightSolid
      }
    }), _defineProperty(_$concat, '.component-tree-item', {
      transition: "color ".concat(token.motionDurationMid),
      lineHeight: "24px",
      height: 24,
      display: 'inline-block'
    }), _$concat))
  };
});

var getTreeItemId = function getTreeItemId(component) {
  return "component-tree-item-".concat(component);
};

var ComponentTree = function ComponentTree(_ref) {
  var _onSelect = _ref.onSelect,
      components = _ref.components,
      selectedTokens = _ref.selectedTokens,
      _ref$filterMode = _ref.filterMode,
      filterMode = _ref$filterMode === void 0 ? 'filter' : _ref$filterMode,
      activeComponent = _ref.activeComponent;

  var _useStyle = useStyle(),
      _useStyle2 = _slicedToArray(_useStyle, 2),
      wrapSSR = _useStyle2[0],
      hashId = _useStyle2[1];

  var treeRef = useRef(null);

  var _useState = useState(''),
      _useState2 = _slicedToArray(_useState, 2),
      search = _useState2[0],
      setSearch = _useState2[1];

  var relatedComponents = useMemo(function () {
    return selectedTokens ? getRelatedComponents(selectedTokens) : []; // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [selectedTokens]);
  useEffect(function () {
    var _treeRef$current, _treeRef$current$quer;

    (_treeRef$current = treeRef.current) === null || _treeRef$current === void 0 ? void 0 : (_treeRef$current$quer = _treeRef$current.querySelector("#".concat(getTreeItemId(activeComponent || '')))) === null || _treeRef$current$quer === void 0 ? void 0 : _treeRef$current$quer.scrollIntoView({
      block: 'nearest',
      inline: 'nearest'
    });
  }, [activeComponent]);
  var treeData = useMemo(function () {
    return Object.entries(components).filter(function (_ref2) {
      var _ref3 = _slicedToArray(_ref2, 2),
          group = _ref3[1];

      return (filterMode === 'highlight' || !relatedComponents.length || group.some(function (item) {
        return relatedComponents.includes(item);
      })) && (!search || group.some(function (item) {
        return item.toLowerCase().includes(search.toLowerCase());
      }));
    }).map(function (_ref4) {
      var _ref5 = _slicedToArray(_ref4, 2),
          type = _ref5[0],
          group = _ref5[1];

      return {
        title: type,
        key: "type-".concat(type),
        children: group.filter(function (item) {
          return (filterMode === 'highlight' || !relatedComponents.length || relatedComponents.includes(item)) && (!search || item.toLowerCase().includes(search.toLowerCase()));
        }).map(function (item) {
          return {
            title: /*#__PURE__*/React.createElement("span", {
              id: getTreeItemId(item),
              className: classNames('component-tree-item', {
                'component-tree-item-highlight': filterMode === 'highlight' && relatedComponents.includes(item)
              })
            }, item),
            switcherIcon: function switcherIcon() {
              return /*#__PURE__*/React.createElement(Badge, {
                color: filterMode === 'highlight' && relatedComponents.includes(item) ? activeComponent === item ? 'white' : 'blue' : 'transparent'
              });
            },
            key: item
          };
        })
      };
    });
  }, [components, relatedComponents, filterMode, search, activeComponent]);
  useEffect(function () {
    if (filterMode === 'highlight') {
      setTimeout(function () {
        var _treeRef$current2, _treeRef$current2$get;

        (_treeRef$current2 = treeRef.current) === null || _treeRef$current2 === void 0 ? void 0 : (_treeRef$current2$get = _treeRef$current2.getElementsByClassName('component-tree-item-active')[0]) === null || _treeRef$current2$get === void 0 ? void 0 : _treeRef$current2$get.scrollIntoView({
          block: 'start',
          inline: 'nearest',
          behavior: 'smooth'
        });
      }, 100);
    }
  }, [selectedTokens, filterMode]);
  return wrapSSR( /*#__PURE__*/React.createElement("div", {
    className: classNames('component-tree-wrapper', hashId)
  }, /*#__PURE__*/React.createElement(Input, {
    allowClear: true,
    placeholder: "Type to search",
    value: search,
    onChange: function onChange(e) {
      return setSearch(e.target.value);
    },
    prefix: /*#__PURE__*/React.createElement(SearchOutlined, null),
    bordered: false,
    className: "component-tree-search"
  }), /*#__PURE__*/React.createElement("div", {
    ref: treeRef,
    style: {
      overflow: 'auto',
      flex: 1
    }
  }, /*#__PURE__*/React.createElement(DirectoryTree, {
    selectedKeys: [activeComponent !== null && activeComponent !== void 0 ? activeComponent : ''],
    showIcon: false,
    defaultExpandAll: true,
    treeData: treeData,
    className: "component-tree",
    onSelect: function onSelect(node) {
      return _onSelect === null || _onSelect === void 0 ? void 0 : _onSelect(node[0]);
    },
    expandAction: "doubleClick"
  }))));
};

export default ComponentTree;
