// @antv/x6-react-components/lib/toolbar/item.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ToolbarItem = void 0;
var react_1 = __importDefault(require("react"));
var classnames_1 = __importDefault(require("classnames"));
var antd_1 = require("antd");
var menu_1 = require("../menu");
var dropdown_1 = require("../dropdown");
var context_1 = require("./context");
var ToolbarItemInner = /** @class */ (function (_super) {
    __extends(ToolbarItemInner, _super);
    function ToolbarItemInner() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.handleClick = function () {
            _this.processClick();
        };
        _this.handleDropdownItemClick = function (name) {
            _this.processClick(name, false);
        };
        return _this;
    }
    ToolbarItemInner.prototype.processClick = function (name, dropdown) {
        if (name === void 0) { name = this.props.name; }
        if (dropdown === void 0) { dropdown = this.props.dropdown; }
        if (!this.props.disabled && !dropdown) {
            if (name) {
                this.props.context.onClick(name);
            }
            if (this.props.onClick) {
                this.props.onClick(name);
            }
        }
    };
    ToolbarItemInner.prototype.renderButton = function () {
        var _a;
        var _b = this.props, className = _b.className, hidden = _b.hidden, disabled = _b.disabled, active = _b.active, icon = _b.icon, text = _b.text, dropdown = _b.dropdown, dropdownArrow = _b.dropdownArrow, tooltip = _b.tooltip, tooltipProps = _b.tooltipProps, tooltipAsTitle = _b.tooltipAsTitle, children = _b.children;
        var prefixCls = this.props.context.prefixCls;
        var baseCls = prefixCls + "-item";
        var props = {
            onClick: this.handleClick,
            className: (0, classnames_1.default)(baseCls, (_a = {},
                _a[baseCls + "-hidden"] = hidden,
                _a[baseCls + "-active"] = active,
                _a[baseCls + "-disabled"] = disabled,
                _a[baseCls + "-dropdown"] = dropdown,
                _a), className),
        };
        if (tooltip && tooltipAsTitle) {
            props.title = tooltip;
        }
        var button = (react_1.default.createElement("button", __assign({ type: "button" }, props),
            icon && react_1.default.isValidElement(icon) && (react_1.default.createElement("span", { className: baseCls + "-icon" }, icon)),
            (text || children) && (react_1.default.createElement("span", { className: baseCls + "-text" }, text || children)),
            dropdown && dropdownArrow && (react_1.default.createElement("span", { className: baseCls + "-dropdown-arrow" }))));
        if (tooltip && !tooltipAsTitle && !disabled) {
            return (react_1.default.createElement(antd_1.Tooltip, __assign({ title: tooltip, placement: "bottom", mouseEnterDelay: 0, mouseLeaveDelay: 0 }, tooltipProps), button));
        }
        return button;
    };
    ToolbarItemInner.prototype.render = function () {
        var _a = this.props, dropdown = _a.dropdown, dropdownProps = _a.dropdownProps, disabled = _a.disabled;
        var content = this.renderButton();
        if (dropdown != null && !disabled) {
            var overlay = (react_1.default.createElement("div", null, dropdown.type === menu_1.Menu
                ? react_1.default.cloneElement(dropdown, {
                    onClick: this.handleDropdownItemClick,
                })
                : dropdown));
            var props = __assign(__assign({ trigger: ['click'] }, dropdownProps), { disabled: disabled, overlay: overlay });
            return react_1.default.createElement(dropdown_1.Dropdown, __assign({}, props), content);
        }
        return content;
    };
    return ToolbarItemInner;
}(react_1.default.PureComponent));
var ToolbarItem = function (props) { return (react_1.default.createElement(context_1.ToolbarContext.Consumer, null, function (context) { return react_1.default.createElement(ToolbarItemInner, __assign({ context: context }, props)); })); };
exports.ToolbarItem = ToolbarItem;
exports.ToolbarItem.defaultProps = {
    dropdownArrow: true,
};
