// antd-token-previewer/es/component-demos/modal/modalWithButton.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { Button, Modal } from 'antd';
import React, { useState } from 'react';

var Demo = function Demo() {
  var _useState = useState(false),
      _useState2 = _slicedToArray(_useState, 2),
      isModalVisible = _useState2[0],
      setIsModalVisible = _useState2[1];

  var showModal = function showModal() {
    setIsModalVisible(true);
  };

  var handleOk = function handleOk() {
    setIsModalVisible(false);
  };

  var handleCancel = function handleCancel() {
    setIsModalVisible(false);
  };

  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Button, {
    type: "primary",
    onClick: showModal
  }, "Open Modal"), /*#__PURE__*/React.createElement(Modal, {
    title: "Basic Modal",
    open: isModalVisible,
    onOk: handleOk,
    onCancel: handleCancel
  }, /*#__PURE__*/React.createElement("p", null, "Some contents..."), " ", /*#__PURE__*/React.createElement("p", null, "Some contents..."), " ", /*#__PURE__*/React.createElement("p", null, "Some contents...")));
};

var componentDemo = {
  demo: /*#__PURE__*/React.createElement(Demo, null),
  tokens: ['colorBgMask'],
  key: 'modalWithButton'
};
export default componentDemo;
