// @ant-design/graphs/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.G6 = exports.MindMapGraph = exports.RadialGraph = exports.OrganizationalGraph = exports.IndentedTreeGraph = exports.DagreFundFlowGraph = exports.IndentedTree = exports.DagreGraph = exports.OrganizationTreeGraph = exports.FundFlowGraph = exports.OrganizationGraph = exports.DecompositionTreeGraph = exports.RadialTreeGraph = exports.FlowAnalysisGraph = void 0;
var g6_1 = __importDefault(require("@antv/g6"));
exports.G6 = g6_1.default;
var organization_graph_1 = __importDefault(require("./components/organization-graph"));
exports.OrganizationGraph = organization_graph_1.default;
var radial_tree_graph_1 = __importDefault(require("./components/radial-tree-graph"));
exports.RadialTreeGraph = radial_tree_graph_1.default;
var flow_analysis_graph_1 = __importDefault(require("./components/flow-analysis-graph"));
exports.FlowAnalysisGraph = flow_analysis_graph_1.default;
var decomposition_tree_graph_1 = __importDefault(require("./components/decomposition-tree-graph"));
exports.DecompositionTreeGraph = decomposition_tree_graph_1.default;
var fund_flow_graph_1 = __importDefault(require("./components/fund-flow-graph"));
exports.FundFlowGraph = fund_flow_graph_1.default;
var radial_graph_1 = __importDefault(require("./components/radial-graph"));
exports.RadialGraph = radial_graph_1.default;
var mind_map_graph_1 = __importDefault(require("./components/mind-map-graph"));
exports.MindMapGraph = mind_map_graph_1.default;
/** @deprecated start */
var graph_1 = require("./obsolescent/graph");
Object.defineProperty(exports, "OrganizationTreeGraph", { enumerable: true, get: function () { return graph_1.OrganizationTreeGraph; } });
Object.defineProperty(exports, "DagreGraph", { enumerable: true, get: function () { return graph_1.DagreGraph; } });
Object.defineProperty(exports, "IndentedTree", { enumerable: true, get: function () { return graph_1.IndentedTree; } });
Object.defineProperty(exports, "DagreFundFlowGraph", { enumerable: true, get: function () { return graph_1.DagreFundFlowGraph; } });
var graphs_1 = require("./obsolescent/graphs");
Object.defineProperty(exports, "IndentedTreeGraph", { enumerable: true, get: function () { return graphs_1.IndentedTreeGraph; } });
Object.defineProperty(exports, "OrganizationalGraph", { enumerable: true, get: function () { return graphs_1.OrganizationalGraph; } });
__exportStar(require("./interface"), exports);
__exportStar(require("./layout"), exports);
