// antd-token-previewer/lib/component-demos/dropdown/dropdown.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _antd = require("antd");

var _icons = require("@ant-design/icons");

var _menu = _interopRequireDefault(require("./menu"));

var Demo = function Demo() {
  return /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("a", {
    className: "ant-dropdown-link",
    onClick: function onClick(e) {
      return e.preventDefault();
    }
  }, "Hover me ", /*#__PURE__*/_react.default.createElement(_icons.DownOutlined, null)), /*#__PURE__*/_react.default.createElement(_antd.Dropdown._InternalPanelDoNotUseOrYouWillBeFired, {
    overlay: _menu.default
  }));
};

var componentDemo = {
  demo: /*#__PURE__*/_react.default.createElement(Demo, null),
  tokens: ['colorPrimary', 'colorError', 'colorErrorHover', 'colorBgElevated'],
  key: 'default'
};
var _default = componentDemo;
exports.default = _default;
