// @antv/l7-core/lib/services/layer/ILayerService.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StencilType = exports.MaskOperation = exports.ILayerStage = exports.BlendType = void 0;
// @ts-ignore
var BlendType;
exports.BlendType = BlendType;
(function (BlendType) {
  BlendType["normal"] = "normal";
  BlendType["additive"] = "additive";
  BlendType["subtractive"] = "subtractive";
  BlendType["min"] = "min";
  BlendType["max"] = "max";
  BlendType["none"] = "none";
})(BlendType || (exports.BlendType = BlendType = {}));
var StencilType;
exports.StencilType = StencilType;
(function (StencilType) {
  StencilType["MULTIPLE"] = "MULTIPLE";
  StencilType["SINGLE"] = "SINGLE";
})(StencilType || (exports.StencilType = StencilType = {}));
var MaskOperation;
exports.MaskOperation = MaskOperation;
(function (MaskOperation) {
  MaskOperation["AND"] = "and";
  MaskOperation["OR"] = "or";
})(MaskOperation || (exports.MaskOperation = MaskOperation = {}));
var ILayerStage;
exports.ILayerStage = ILayerStage;
(function (ILayerStage) {
  ILayerStage["INIT"] = "init";
  ILayerStage["UPDATE"] = "update";
})(ILayerStage || (exports.ILayerStage = ILayerStage = {}));
