// @antv/l7-utils/lib/tileset-manager/const.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/tileset-manager/const.ts
var const_exports = {};
__export(const_exports, {
  BOUNDS_BUFFER_SCALE: () => BOUNDS_BUFFER_SCALE,
  DEFAULT_CACHE_SCALE: () => DEFAULT_CACHE_SCALE,
  DEFAULT_EXTENT: () => DEFAULT_EXTENT,
  NOOP: () => NOOP,
  TILE_SIZE: () => TILE_SIZE,
  UPDATE_TILE_STRATEGIES: () => UPDATE_TILE_STRATEGIES
});
module.exports = __toCommonJS(const_exports);
var import_types = require("./types");
var import_strategies = require("./utils/strategies");
var TILE_SIZE = 256;
var DEFAULT_EXTENT = [
  -Infinity,
  -Infinity,
  Infinity,
  Infinity
];
var BOUNDS_BUFFER_SCALE = 0.2;
var DEFAULT_CACHE_SCALE = 5;
var UPDATE_TILE_STRATEGIES = {
  [import_types.UpdateTileStrategy.Realtime]: import_strategies.updateTileStateRealtime,
  [import_types.UpdateTileStrategy.Overlap]: import_strategies.updateTileStateOverlap,
  [import_types.UpdateTileStrategy.Replace]: import_strategies.updateTileStateReplace
};
var NOOP = () => {
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  BOUNDS_BUFFER_SCALE,
  DEFAULT_CACHE_SCALE,
  DEFAULT_EXTENT,
  NOOP,
  TILE_SIZE,
  UPDATE_TILE_STRATEGIES
});
