// @antv/l7-utils/lib/tileset-manager/types.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/tileset-manager/types.ts
var types_exports = {};
__export(types_exports, {
  LoadTileDataStatus: () => LoadTileDataStatus,
  UpdateTileStrategy: () => UpdateTileStrategy
});
module.exports = __toCommonJS(types_exports);
var UpdateTileStrategy = /* @__PURE__ */ ((UpdateTileStrategy2) => {
  UpdateTileStrategy2["Realtime"] = "realtime";
  UpdateTileStrategy2["Overlap"] = "overlap";
  UpdateTileStrategy2["Replace"] = "replace";
  return UpdateTileStrategy2;
})(UpdateTileStrategy || {});
var LoadTileDataStatus = /* @__PURE__ */ ((LoadTileDataStatus2) => {
  LoadTileDataStatus2["Loading"] = "Loading";
  LoadTileDataStatus2["Loaded"] = "Loaded";
  LoadTileDataStatus2["Failure"] = "Failure";
  LoadTileDataStatus2["Cancelled"] = "Cancelled";
  return LoadTileDataStatus2;
})(LoadTileDataStatus || {});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  LoadTileDataStatus,
  UpdateTileStrategy
});
