// @antv/l7plot/dist/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Choropleth = exports.Area = exports.Flow = exports.Path = exports.Hexbin = exports.Grid = exports.Heatmap = exports.DotDensity = exports.Dot = exports.PrismLayer = exports.AreaLayer = exports.ArcLayer = exports.PathLayer = exports.HexbinLayer = exports.GridLayer = exports.HeatmapLayer = exports.TextLayer = exports.DotDensityLayer = exports.IconLayer = exports.DotLayer = exports.L7Plot = exports.Plot = exports.Map = exports.registerTheme = exports.getTheme = exports.unregisterIconFont = exports.registerIconFonts = exports.registerIconFont = exports.unregisterFontFace = exports.registerFontFace = exports.unregisterImage = exports.registerImages = exports.registerImage = exports.version = void 0;
var tslib_1 = require("tslib");
var version_1 = require("./version");
Object.defineProperty(exports, "version", { enumerable: true, get: function () { return tslib_1.__importDefault(version_1).default; } });
/** 资源静态注册 **/
var register_1 = require("./core/map/register");
Object.defineProperty(exports, "registerImage", { enumerable: true, get: function () { return register_1.registerImage; } });
Object.defineProperty(exports, "registerImages", { enumerable: true, get: function () { return register_1.registerImages; } });
Object.defineProperty(exports, "unregisterImage", { enumerable: true, get: function () { return register_1.unregisterImage; } });
Object.defineProperty(exports, "registerFontFace", { enumerable: true, get: function () { return register_1.registerFontFace; } });
Object.defineProperty(exports, "unregisterFontFace", { enumerable: true, get: function () { return register_1.unregisterFontFace; } });
Object.defineProperty(exports, "registerIconFont", { enumerable: true, get: function () { return register_1.registerIconFont; } });
Object.defineProperty(exports, "registerIconFonts", { enumerable: true, get: function () { return register_1.registerIconFonts; } });
Object.defineProperty(exports, "unregisterIconFont", { enumerable: true, get: function () { return register_1.unregisterIconFont; } });
/** 主题注册 **/
var theme_1 = require("./theme");
Object.defineProperty(exports, "getTheme", { enumerable: true, get: function () { return theme_1.getTheme; } });
Object.defineProperty(exports, "registerTheme", { enumerable: true, get: function () { return theme_1.registerTheme; } });
/** 类型定义导出 **/
tslib_1.__exportStar(require("./types"), exports);
/** L7Plot 的基类 **/
var map_1 = require("./core/map");
Object.defineProperty(exports, "Map", { enumerable: true, get: function () { return map_1.Map; } });
/** 各图表的基类 Plot **/
var plot_1 = require("./core/plot");
Object.defineProperty(exports, "Plot", { enumerable: true, get: function () { return plot_1.Plot; } });
/** 高级图表 **/
var plot_2 = require("./plot");
Object.defineProperty(exports, "L7Plot", { enumerable: true, get: function () { return plot_2.L7Plot; } });
/** 图层 **/
// 散点图层及类型定义 | author by [yunji]](https://github.com/lvisei)
var dot_layer_1 = require("./layers/dot-layer");
Object.defineProperty(exports, "DotLayer", { enumerable: true, get: function () { return dot_layer_1.DotLayer; } });
// 图标图层及类型定义 | author by [yunji]](https://github.com/lvisei)
var icon_layer_1 = require("./layers/icon-layer");
Object.defineProperty(exports, "IconLayer", { enumerable: true, get: function () { return icon_layer_1.IconLayer; } });
// 点密度图层及类型定义 | author by [yunji]](https://github.com/lvisei)
var dot_density_layer_1 = require("./layers/dot-density-layer");
Object.defineProperty(exports, "DotDensityLayer", { enumerable: true, get: function () { return dot_density_layer_1.DotDensityLayer; } });
// 文本图层及类型定义 | author by [yunji]](https://github.com/lvisei)
var text_layer_1 = require("./layers/text-layer");
Object.defineProperty(exports, "TextLayer", { enumerable: true, get: function () { return text_layer_1.TextLayer; } });
// 热力图层及类型定义 | author by [yunji]](https://github.com/lvisei)
var heatmap_layer_1 = require("./layers/heatmap-layer");
Object.defineProperty(exports, "HeatmapLayer", { enumerable: true, get: function () { return heatmap_layer_1.HeatmapLayer; } });
// 网格聚合图层及类型定义 | author by [yunji]](https://github.com/lvisei)
var grid_layer_1 = require("./layers/grid-layer");
Object.defineProperty(exports, "GridLayer", { enumerable: true, get: function () { return grid_layer_1.GridLayer; } });
// 蜂窝聚合图层及类型定义 | author by [yunji]](https://github.com/lvisei)
var hexbin_layer_1 = require("./layers/hexbin-layer");
Object.defineProperty(exports, "HexbinLayer", { enumerable: true, get: function () { return hexbin_layer_1.HexbinLayer; } });
// 路径图层及类型定义 | author by [yunji]](https://github.com/lvisei)
var path_layer_1 = require("./layers/path-layer");
Object.defineProperty(exports, "PathLayer", { enumerable: true, get: function () { return path_layer_1.PathLayer; } });
// 弧线图层及类型定义 | author by [yunji]](https://github.com/lvisei)
var arc_layer_1 = require("./layers/arc-layer");
Object.defineProperty(exports, "ArcLayer", { enumerable: true, get: function () { return arc_layer_1.ArcLayer; } });
// 区域图层及类型定义 | author by [yunji]](https://github.com/lvisei)
var area_layer_1 = require("./layers/area-layer");
Object.defineProperty(exports, "AreaLayer", { enumerable: true, get: function () { return area_layer_1.AreaLayer; } });
// 水晶体图层及类型定义 | author by [yunji]](https://github.com/lvisei)
var prism_layer_1 = require("./layers/prism-layer");
Object.defineProperty(exports, "PrismLayer", { enumerable: true, get: function () { return prism_layer_1.PrismLayer; } });
/** 图表 **/
// 散点图及类型定义 | author by [yunji]](https://github.com/lvisei)
var dot_1 = require("./plots/dot");
Object.defineProperty(exports, "Dot", { enumerable: true, get: function () { return dot_1.Dot; } });
// 点密度图及类型定义 | author by [yunji]](https://github.com/lvisei)
var dot_density_1 = require("./plots/dot-density");
Object.defineProperty(exports, "DotDensity", { enumerable: true, get: function () { return dot_density_1.DotDensity; } });
// 热力图及类型定义 | author by [yunji]](https://github.com/lvisei)
var heatmap_1 = require("./plots/heatmap");
Object.defineProperty(exports, "Heatmap", { enumerable: true, get: function () { return heatmap_1.Heatmap; } });
// 网格聚合图及类型定义 | author by [yunji]](https://github.com/lvisei)
var grid_1 = require("./plots/grid");
Object.defineProperty(exports, "Grid", { enumerable: true, get: function () { return grid_1.Grid; } });
// 蜂窝聚合图及类型定义 | author by [yunji]](https://github.com/lvisei)
var hexbin_1 = require("./plots/hexbin");
Object.defineProperty(exports, "Hexbin", { enumerable: true, get: function () { return hexbin_1.Hexbin; } });
// 路径图及类型定义 | author by [yunji]](https://github.com/lvisei)
var path_1 = require("./plots/path");
Object.defineProperty(exports, "Path", { enumerable: true, get: function () { return path_1.Path; } });
// 流向图及类型定义 | author by [yunji]](https://github.com/lvisei)
var flow_1 = require("./plots/flow");
Object.defineProperty(exports, "Flow", { enumerable: true, get: function () { return flow_1.Flow; } });
// 区域图及类型定义 | author by [yunji]](https://github.com/lvisei)
var area_1 = require("./plots/area");
Object.defineProperty(exports, "Area", { enumerable: true, get: function () { return area_1.Area; } });
// 行政区域图及类型定义 | author by [yunji]](https://github.com/lvisei)
var choropleth_1 = require("./plots/choropleth");
Object.defineProperty(exports, "Choropleth", { enumerable: true, get: function () { return choropleth_1.Choropleth; } });
