// @antv/l7-component/lib/utils/popper.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Popper = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _l7Utils = require("@antv/l7-utils");
var _eventemitter = require("eventemitter3");
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }
function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
var Popper = /*#__PURE__*/function (_EventEmitter) {
  (0, _inherits2.default)(Popper, _EventEmitter);
  var _super = _createSuper(Popper);
  function Popper(button, option) {
    var _this;
    (0, _classCallCheck2.default)(this, Popper);
    _this = _super.call(this);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "isShow", false);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "timeout", null);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "show", function () {
      if (_this.isShow || !_this.contentDOM.innerHTML) {
        return (0, _assertThisInitialized2.default)(_this);
      }
      _this.resetPopperPosition();
      _l7Utils.DOM.removeClass(_this.popperDOM, 'l7-popper-hide');
      _this.isShow = true;
      if (_this.option.unique) {
        // console.log(Popper.conflictPopperList.length);
        Popper.conflictPopperList.forEach(function (popper) {
          if (popper !== (0, _assertThisInitialized2.default)(_this) && popper.isShow) {
            popper.hide();
          }
        });
      }
      _this.emit('show');
      return (0, _assertThisInitialized2.default)(_this);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "hide", function () {
      if (!_this.isShow) {
        return (0, _assertThisInitialized2.default)(_this);
      }
      _l7Utils.DOM.addClass(_this.popperDOM, 'l7-popper-hide');
      _this.isShow = false;
      _this.emit('hide');
      return (0, _assertThisInitialized2.default)(_this);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "setHideTimeout", function () {
      if (_this.timeout) {
        return;
      }
      _this.timeout = window.setTimeout(function () {
        if (!_this.isShow) {
          return;
        }
        _this.hide();
        _this.timeout = null;
      }, 300);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "clearHideTimeout", function () {
      if (_this.timeout) {
        window.clearTimeout(_this.timeout);
        _this.timeout = null;
      }
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onBtnClick", function () {
      if (_this.isShow) {
        _this.hide();
      } else {
        _this.show();
      }
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onBtnMouseLeave", function () {
      _this.setHideTimeout();
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onBtnMouseMove", function () {
      _this.clearHideTimeout();
      if (_this.isShow) {
        return;
      }
      _this.show();
    });
    _this.button = button;
    _this.option = option;
    _this.init();
    if (option.unique) {
      Popper.conflictPopperList.push((0, _assertThisInitialized2.default)(_this));
    }
    return _this;
  }
  (0, _createClass2.default)(Popper, [{
    key: "buttonRect",
    get: function get() {
      return this.button.getBoundingClientRect();
    }
  }, {
    key: "getPopperDOM",
    value: function getPopperDOM() {
      return this.popperDOM;
    }
  }, {
    key: "getIsShow",
    value: function getIsShow() {
      return this.isShow;
    }
  }, {
    key: "getContent",
    value: function getContent() {
      return this.content;
    }
  }, {
    key: "setContent",
    value: function setContent(content) {
      if (typeof content === 'string') {
        this.contentDOM.innerHTML = content;
      } else if (content instanceof HTMLElement) {
        _l7Utils.DOM.clearChildren(this.contentDOM);
        this.contentDOM.appendChild(content);
      }
      this.content = content;
    }
  }, {
    key: "init",
    value: function init() {
      var trigger = this.option.trigger;
      this.popperDOM = this.createPopper();
      if (trigger === 'click') {
        this.button.addEventListener('click', this.onBtnClick);
      } else {
        this.button.addEventListener('mousemove', this.onBtnMouseMove);
        this.button.addEventListener('mouseleave', this.onBtnMouseLeave);
        this.popperDOM.addEventListener('mousemove', this.onBtnMouseMove);
        this.popperDOM.addEventListener('mouseleave', this.onBtnMouseLeave);
      }
    }
  }, {
    key: "destroy",
    value: function destroy() {
      this.button.removeEventListener('click', this.onBtnClick);
      this.button.removeEventListener('mousemove', this.onBtnMouseMove);
      this.button.removeEventListener('mousemove', this.onBtnMouseLeave);
      this.popperDOM.removeEventListener('mousemove', this.onBtnMouseMove);
      this.popperDOM.removeEventListener('mouseleave', this.onBtnMouseLeave);
      _l7Utils.DOM.remove(this.popperDOM);
    }
  }, {
    key: "resetPopperPosition",
    value: function resetPopperPosition() {
      var popperStyleObj = {};
      var _this$option = this.option,
        container = _this$option.container,
        _this$option$offset = _this$option.offset,
        offset = _this$option$offset === void 0 ? [0, 0] : _this$option$offset,
        placement = _this$option.placement;
      var _offset = (0, _slicedToArray2.default)(offset, 2),
        offsetX = _offset[0],
        offsetY = _offset[1];
      var buttonRect = this.button.getBoundingClientRect();
      var containerRect = container.getBoundingClientRect();
      var _DOM$getDiffRect = _l7Utils.DOM.getDiffRect(buttonRect, containerRect),
        left = _DOM$getDiffRect.left,
        right = _DOM$getDiffRect.right,
        top = _DOM$getDiffRect.top,
        bottom = _DOM$getDiffRect.bottom;
      var isTransformX = false;
      var isTransformY = false;
      if (/^(left|right)/.test(placement)) {
        if (placement.includes('left')) {
          popperStyleObj.right = "".concat(buttonRect.width + right, "px");
        } else if (placement.includes('right')) {
          popperStyleObj.left = "".concat(buttonRect.width + left, "px");
        }
        if (placement.includes('start')) {
          popperStyleObj.top = "".concat(top, "px");
        } else if (placement.includes('end')) {
          popperStyleObj.bottom = "".concat(bottom, "px");
        } else {
          popperStyleObj.top = "".concat(top + buttonRect.height / 2, "px");
          isTransformY = true;
          popperStyleObj.transform = "translate(".concat(offsetX, "px, calc(").concat(offsetY, "px - 50%))");
        }
      } else if (/^(top|bottom)/.test(placement)) {
        if (placement.includes('top')) {
          popperStyleObj.bottom = "".concat(buttonRect.height + bottom, "px");
        } else if (placement.includes('bottom')) {
          popperStyleObj.top = "".concat(buttonRect.height + top, "px");
        }
        if (placement.includes('start')) {
          popperStyleObj.left = "".concat(left, "px");
        } else if (placement.includes('end')) {
          popperStyleObj.right = "".concat(right, "px");
        } else {
          popperStyleObj.left = "".concat(left + buttonRect.width / 2, "px");
          isTransformX = true;
          popperStyleObj.transform = "translate(calc(".concat(offsetX, "px - 50%), ").concat(offsetY, "px)");
        }
      }
      popperStyleObj.transform = "translate(calc(".concat(offsetX, "px - ").concat(isTransformX ? '50%' : '0%', "), calc(").concat(offsetY, "px - ").concat(isTransformY ? '50%' : '0%', ")");
      var posList = placement.split('-');
      if (posList.length) {
        _l7Utils.DOM.addClass(this.popperDOM, posList.map(function (pos) {
          return "l7-popper-".concat(pos);
        }).join(' '));
      }
      _l7Utils.DOM.addStyle(this.popperDOM, _l7Utils.DOM.css2Style(popperStyleObj));
    }
  }, {
    key: "createPopper",
    value: function createPopper() {
      var _this$option2 = this.option,
        container = _this$option2.container,
        _this$option2$classNa = _this$option2.className,
        className = _this$option2$classNa === void 0 ? '' : _this$option2$classNa,
        content = _this$option2.content;
      var popper = _l7Utils.DOM.create('div', "l7-popper l7-popper-hide ".concat(className));
      var popperContent = _l7Utils.DOM.create('div', 'l7-popper-content');
      var popperArrow = _l7Utils.DOM.create('div', 'l7-popper-arrow');
      popper.appendChild(popperContent);
      popper.appendChild(popperArrow);
      container.appendChild(popper);
      this.popperDOM = popper;
      this.contentDOM = popperContent;
      if (content) {
        this.setContent(content);
      }
      return popper;
    }
  }]);
  return Popper;
}(_eventemitter.EventEmitter);
exports.Popper = Popper;
(0, _defineProperty2.default)(Popper, "conflictPopperList", []);
