// antd-token-previewer/es/component-demos/drawer/drawer.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { Button, Drawer } from 'antd';
import React, { useState } from 'react';

var Demo = function Demo() {
  var _useState = useState(false),
      _useState2 = _slicedToArray(_useState, 2),
      visible = _useState2[0],
      setVisible = _useState2[1];

  var showDrawer = function showDrawer() {
    setVisible(true);
  };

  var onClose = function onClose() {
    setVisible(false);
  };

  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Button, {
    type: "primary",
    onClick: showDrawer
  }, "Open"), /*#__PURE__*/React.createElement(Drawer, {
    title: "Basic Drawer",
    placement: "right",
    onClose: onClose,
    open: visible
  }, /*#__PURE__*/React.createElement("p", null, "Some contents..."), /*#__PURE__*/React.createElement("p", null, "Some contents..."), /*#__PURE__*/React.createElement("p", null, "Some contents...")));
};

var componentDemo = {
  demo: /*#__PURE__*/React.createElement(Demo, null),
  tokens: ['colorBgMask', 'colorBgElevated'],
  key: 'default'
};
export default componentDemo;
