// @antv/l7plot-component/dist/lib/tooltip/theme.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var theme_1 = tslib_1.__importDefault(require("../configs/theme"));
var constants_1 = require("./constants");
exports.default = (_a = {},
    // css style for tooltip
    _a[constants_1.CONTAINER_CLASS] = {
        visibility: 'visible',
        zIndex: 999,
        transition: 'visibility 0.2s cubic-bezier(0.23, 1, 0.32, 1), ' +
            'left 0.4s cubic-bezier(0.23, 1, 0.32, 1), ' +
            'top 0.4s cubic-bezier(0.23, 1, 0.32, 1)',
        backgroundColor: 'rgba(255, 255, 255, 0.9)',
        boxShadow: 'rgb(0 0 0 / 16%) 0px 6px 12px 0px',
        borderRadius: '2px',
        color: 'rgb(87, 87, 87)',
        fontSize: '12px',
        fontFamily: theme_1.default.fontFamily,
        lineHeight: '20px',
        padding: '10px 10px 6px 10px',
    },
    _a[constants_1.TITLE_CLASS] = {
        marginBottom: '4px',
    },
    _a[constants_1.LIST_CLASS] = {
        margin: '0px',
        listStyleType: 'none',
        padding: '0px',
    },
    _a[constants_1.LIST_ITEM_CLASS] = {
        listStyleType: 'none',
        marginBottom: '4px',
        whiteSpace: 'nowrap',
        display: 'flex',
        justifyContent: 'space-between',
    },
    _a[constants_1.NAME_CLASS] = {},
    _a[constants_1.VALUE_CLASS] = {
        marginLeft: '30px',
    },
    _a);
