// antd-token-previewer/lib/component-demos/popover/popover.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _antd = require("antd");

var content = /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("p", null, "Content"), " ", /*#__PURE__*/_react.default.createElement("p", null, "Content"));

var Demo = function Demo() {
  return /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(_antd.Popover._InternalPanelDoNotUseOrYouWillBeFired, {
    content: content,
    title: "Title"
  }), /*#__PURE__*/_react.default.createElement(_antd.Button, {
    type: "primary"
  }, "Hover me"));
};

var componentDemo = {
  demo: /*#__PURE__*/_react.default.createElement(Demo, null),
  tokens: ['colorBgElevated'],
  key: 'default'
};
var _default = componentDemo;
exports.default = _default;
