// antd/es/steps/style/progress.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
var genStepsProgressStyle = function genStepsProgressStyle(token) {
  var _ref6;
  var antCls = token.antCls,
    componentCls = token.componentCls;
  return _defineProperty({}, "&" + componentCls + "-with-progress", (_ref6 = {}, _defineProperty(_ref6, componentCls + "-item", _defineProperty({
    paddingTop: token.paddingXXS
  }, "&-process " + componentCls + "-item-container " + componentCls + "-item-icon " + componentCls + "-icon", {
    color: token.processIconColor
  })), _defineProperty(_ref6, "&" + componentCls + "-vertical > " + componentCls + "-item ", _defineProperty({
    paddingInlineStart: token.paddingXXS
  }, "> " + componentCls + "-item-container > " + componentCls + "-item-tail", {
    top: token.marginXXS,
    insetInlineStart: token.stepsIconSize / 2 - token.lineWidth + token.paddingXXS
  })), _defineProperty(_ref6, "&, &" + componentCls + "-small", _defineProperty({}, "&" + componentCls + "-horizontal " + componentCls + "-item:first-child", {
    paddingBottom: token.paddingXXS,
    paddingInlineStart: token.paddingXXS
  })), _defineProperty(_ref6, "&" + componentCls + "-small" + componentCls + "-vertical > " + componentCls + "-item > " + componentCls + "-item-container > " + componentCls + "-item-tail", {
    insetInlineStart: token.stepsSmallIconSize / 2 - token.lineWidth + token.paddingXXS
  }), _defineProperty(_ref6, "&" + componentCls + "-label-vertical", _defineProperty({}, componentCls + "-item " + componentCls + "-item-tail", {
    top: token.margin - 2 * token.lineWidth
  })), _defineProperty(_ref6, componentCls + "-item-icon", _defineProperty({
    position: 'relative'
  }, antCls + "-progress", {
    position: 'absolute',
    insetBlockStart: (token.stepsIconSize - token.stepsProgressSize - token.lineWidth * 2) / 2,
    insetInlineStart: (token.stepsIconSize - token.stepsProgressSize - token.lineWidth * 2) / 2
  })), _ref6));
};
export default genStepsProgressStyle;
