// @antv/l7plot/dist/esm/index.js
export { default as version } from './version';
/** 资源静态注册 **/
export { registerImage, registerImages, unregisterImage, registerFontFace, unregisterFontFace, registerIconFont, registerIconFonts, unregisterIconFont, } from './core/map/register';
/** 主题注册 **/
export { getTheme, registerTheme } from './theme';
/** 类型定义导出 **/
export * from './types';
/** L7Plot 的基类 **/
export { Map } from './core/map';
/** 各图表的基类 Plot **/
export { Plot } from './core/plot';
/** 高级图表 **/
export { L7Plot } from './plot';
/** 图层 **/
// 散点图层及类型定义 | author by [yunji]](https://github.com/lvisei)
export { DotLayer } from './layers/dot-layer';
// 图标图层及类型定义 | author by [yunji]](https://github.com/lvisei)
export { IconLayer } from './layers/icon-layer';
// 点密度图层及类型定义 | author by [yunji]](https://github.com/lvisei)
export { DotDensityLayer } from './layers/dot-density-layer';
// 文本图层及类型定义 | author by [yunji]](https://github.com/lvisei)
export { TextLayer } from './layers/text-layer';
// 热力图层及类型定义 | author by [yunji]](https://github.com/lvisei)
export { HeatmapLayer } from './layers/heatmap-layer';
// 网格聚合图层及类型定义 | author by [yunji]](https://github.com/lvisei)
export { GridLayer } from './layers/grid-layer';
// 蜂窝聚合图层及类型定义 | author by [yunji]](https://github.com/lvisei)
export { HexbinLayer } from './layers/hexbin-layer';
// 路径图层及类型定义 | author by [yunji]](https://github.com/lvisei)
export { PathLayer } from './layers/path-layer';
// 弧线图层及类型定义 | author by [yunji]](https://github.com/lvisei)
export { ArcLayer } from './layers/arc-layer';
// 区域图层及类型定义 | author by [yunji]](https://github.com/lvisei)
export { AreaLayer } from './layers/area-layer';
// 水晶体图层及类型定义 | author by [yunji]](https://github.com/lvisei)
export { PrismLayer } from './layers/prism-layer';
/** 图表 **/
// 散点图及类型定义 | author by [yunji]](https://github.com/lvisei)
export { Dot } from './plots/dot';
// 点密度图及类型定义 | author by [yunji]](https://github.com/lvisei)
export { DotDensity } from './plots/dot-density';
// 热力图及类型定义 | author by [yunji]](https://github.com/lvisei)
export { Heatmap } from './plots/heatmap';
// 网格聚合图及类型定义 | author by [yunji]](https://github.com/lvisei)
export { Grid } from './plots/grid';
// 蜂窝聚合图及类型定义 | author by [yunji]](https://github.com/lvisei)
export { Hexbin } from './plots/hexbin';
// 路径图及类型定义 | author by [yunji]](https://github.com/lvisei)
export { Path } from './plots/path';
// 流向图及类型定义 | author by [yunji]](https://github.com/lvisei)
export { Flow } from './plots/flow';
// 区域图及类型定义 | author by [yunji]](https://github.com/lvisei)
export { Area } from './plots/area';
// 行政区域图及类型定义 | author by [yunji]](https://github.com/lvisei)
export { Choropleth } from './plots/choropleth';
