// antd-token-previewer/lib/component-demos/tabs/tabs.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _antd = require("antd");

var TabPane = _antd.Tabs.TabPane;

function callback() {}

var Demo = function Demo() {
  return /*#__PURE__*/_react.default.createElement(_antd.Tabs, {
    defaultActiveKey: "1",
    onChange: callback
  }, /*#__PURE__*/_react.default.createElement(TabPane, {
    tab: "Tab 1",
    key: "1"
  }, "Content of Tab Pane 1"), /*#__PURE__*/_react.default.createElement(TabPane, {
    tab: "Tab 2",
    key: "2"
  }, "Content of Tab Pane 2"), /*#__PURE__*/_react.default.createElement(TabPane, {
    tab: "Tab 3",
    key: "3"
  }, "Content of Tab Pane 3"));
};

var componentDemo = {
  demo: /*#__PURE__*/_react.default.createElement(Demo, null),
  tokens: ['colorPrimary', 'colorPrimaryHover', 'colorPrimaryActive', 'colorBgContainer'],
  key: 'default'
};
var _default = componentDemo;
exports.default = _default;
