// antd-token-previewer/es/component-demos/tree/tree.js
import React from 'react';
import { Tree } from 'antd';
var treeData = [{
  title: 'parent 1',
  key: '0-0',
  children: [{
    title: 'parent 1-0',
    key: '0-0-0',
    disabled: true,
    children: [{
      title: 'leaf',
      key: '0-0-0-0',
      disableCheckbox: true
    }, {
      title: 'leaf',
      key: '0-0-0-1'
    }]
  }, {
    title: 'parent 1-1',
    key: '0-0-1',
    children: [{
      title: /*#__PURE__*/React.createElement("span", {
        style: {
          color: '#1890ff'
        }
      }, "sss"),
      key: '0-0-1-0'
    }]
  }]
}];

var Demo = function Demo() {
  return /*#__PURE__*/React.createElement(Tree, {
    checkable: true,
    defaultExpandedKeys: ['0-0-0', '0-0-1'],
    defaultSelectedKeys: ['0-0-0', '0-0-1'],
    defaultCheckedKeys: ['0-0-0', '0-0-1'],
    treeData: treeData
  });
};

var componentDemo = {
  demo: /*#__PURE__*/React.createElement(Demo, null),
  tokens: ['colorPrimary', 'controlOutline', 'colorBgContainer'],
  key: 'default'
};
export default componentDemo;
