// @antv/xflow-extension/lib/canvas-mini-map/components/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CanvasMiniMap = exports.getDefaultMinimapOptions = exports.CONFIG_TYPE = void 0;
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importDefault(require("react"));
var module_1 = require("../module");
var xflow_core_1 = require("@antv/xflow-core");
var xflow_core_2 = require("@antv/xflow-core");
var x6_node_view_1 = require("../x6-node-view");
/** 配置类型 */
exports.CONFIG_TYPE = 'CanvasMiniMap';
/** 获取默认配置 */
var getDefaultMinimapOptions = function () {
    return {
        grid: false,
        enabled: true,
        width: 200,
        height: 120,
        padding: 8,
        graphOptions: {
            async: true,
            getCellView: function (cell) {
                if (cell.isNode()) {
                    return x6_node_view_1.SimpleNodeView;
                }
            },
            createCellView: function (cell) {
                if (cell.isEdge()) {
                    return null;
                }
            },
        },
    };
};
exports.getDefaultMinimapOptions = getDefaultMinimapOptions;
var CanvasMiniMap = function (props) {
    var nodeFillColor = props.nodeFillColor, _a = props.borderColor, borderColor = _a === void 0 ? '#ced4de' : _a, _b = props.handlerColor, handlerColor = _b === void 0 ? 'rgba(0,0,0,.3)' : _b, _c = props.miniMapClz, miniMapClz = _c === void 0 ? '' : _c, _d = props.position, position = _d === void 0 ? { bottom: 12, right: 12 } : _d, _e = props.minimapOptions, minimapOptions = _e === void 0 ? {} : _e;
    /** x6画布的dom节点 */
    var minimapRef = react_1.default.useRef(null);
    var minimapConfigRef = react_1.default.useRef(props.minimapOptions);
    var classNameSet = react_1.default.useMemo(function () {
        return new Set([
            miniMapClz,
            'xflow-minimap',
            'minimap-instacne-' + Date.now().toString(),
        ]);
    }, [miniMapClz]);
    var getConfig = react_1.default.useMemo(function () { return function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            return [2 /*return*/, tslib_1.__assign(tslib_1.__assign(tslib_1.__assign({}, (0, exports.getDefaultMinimapOptions)()), minimapConfigRef.current), { container: minimapRef.current })];
        });
    }); }; }, [minimapRef]);
    var style = (0, xflow_core_2.usePositionStyle)(position);
    react_1.default.useEffect(function () {
        /** 设置minimap容器 */
        minimapConfigRef.current = minimapOptions;
    }, [minimapConfigRef, minimapOptions]);
    react_1.default.useEffect(function () {
        /** 设置Node样式 */
        if (nodeFillColor) {
            x6_node_view_1.SimpleNodeView.setNodeFillColor(nodeFillColor);
        }
    }, [nodeFillColor]);
    react_1.default.useEffect(function () {
        var containerSelector = Array.from(classNameSet).reduce(function (acc, clz) { return (clz ? "".concat(acc, ".").concat(clz) : acc); }, '');
        var cssString = "\n      ".concat(containerSelector, " .x6-widget-minimap-viewport {\n        border: 1px solid ").concat(borderColor, ";\n        margin: 0;\n      }\n      ").concat(containerSelector, " .x6-widget-minimap-viewport-zoom {\n        border: 1px solid ").concat(handlerColor, ";\n      }\n    ");
        var $style = (0, xflow_core_1.insertCss)(cssString);
        return function () {
            document.head.removeChild($style);
        };
    }, [borderColor, classNameSet, handlerColor]);
    var clzName = Array.from(classNameSet).join(' ');
    return (react_1.default.createElement(xflow_core_2.XFlowAppExtensionModule, { config: {
            getConfig: getConfig,
            CONFIG_TYPE: exports.CONFIG_TYPE,
        }, createModule: module_1.createModule },
        react_1.default.createElement("div", { ref: minimapRef, className: clzName, style: style })));
};
exports.CanvasMiniMap = CanvasMiniMap;
