// @antv/l7-layers/lib/utils/stencil.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getStencil = getStencil;
exports.getStencilMask = getStencilMask;
var _l7Core = require("@antv/l7-core");
// 掩膜配置
function getStencil(mask, maskInside) {
  return {
    enable: mask,
    mask: 0xff,
    func: {
      cmp: _l7Core.gl.EQUAL,
      // gl.EQUAL,
      ref: maskInside ? 1 : 0,
      mask: 1
    }
  };
}
// 模版配置
function getStencilMask(option) {
  if (option.maskOperation === _l7Core.MaskOperation.OR) {
    return {
      enable: true,
      mask: 0xff,
      func: {
        cmp: _l7Core.gl.ALWAYS,
        ref: 1,
        mask: 0xff
      },
      opFront: {
        fail: _l7Core.gl.KEEP,
        zfail: _l7Core.gl.REPLACE,
        zpass: _l7Core.gl.REPLACE
      }
    };
  }
  return {
    enable: true,
    mask: 0xff,
    func: {
      cmp: option.stencilType === _l7Core.StencilType.SINGLE // 单层
      ? _l7Core.gl.ALWAYS : option.stencilIndex === 0 // 多层
      ? _l7Core.gl.ALWAYS : _l7Core.gl.LESS,
      ref: option.stencilType === _l7Core.StencilType.SINGLE ? 1 : option.stencilIndex === 0 ? 2 : 1,
      mask: 0xff
    },
    opFront: {
      fail: _l7Core.gl.KEEP,
      zfail: _l7Core.gl.REPLACE,
      zpass: _l7Core.gl.REPLACE
    }
  };
}
