// antd-token-previewer/lib/component-demos/card/cardGrid.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _antd = require("antd");

var gridStyle = {
  width: '25%',
  textAlign: 'center'
};

var Demo = function Demo() {
  return /*#__PURE__*/_react.default.createElement(_antd.Card, {
    title: "Card Title"
  }, /*#__PURE__*/_react.default.createElement(_antd.Card.Grid, {
    style: gridStyle
  }, "Content"), /*#__PURE__*/_react.default.createElement(_antd.Card.Grid, {
    hoverable: false,
    style: gridStyle
  }, "Content"), /*#__PURE__*/_react.default.createElement(_antd.Card.Grid, {
    style: gridStyle
  }, "Content"), /*#__PURE__*/_react.default.createElement(_antd.Card.Grid, {
    style: gridStyle
  }, "Content"), /*#__PURE__*/_react.default.createElement(_antd.Card.Grid, {
    style: gridStyle
  }, "Content"), /*#__PURE__*/_react.default.createElement(_antd.Card.Grid, {
    style: gridStyle
  }, "Content"), /*#__PURE__*/_react.default.createElement(_antd.Card.Grid, {
    style: gridStyle
  }, "Content"));
};

var componentDemo = {
  demo: /*#__PURE__*/_react.default.createElement(Demo, null),
  tokens: ['colorBorderSecondary'],
  key: 'cardGrid'
};
var _default = componentDemo;
exports.default = _default;
