// @antv/l7-component/lib/control/baseControl/popperControl.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.PopperControl = void 0;
var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _get2 = _interopRequireDefault(require("@babel/runtime/helpers/get"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _popper = require("../../utils/popper");
var _buttonControl = _interopRequireDefault(require("./buttonControl"));
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }
function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
var PopperPlacementMap = {
  topleft: 'right-start',
  topcenter: 'bottom',
  topright: 'left-start',
  bottomleft: 'right-end',
  bottomcenter: 'top',
  bottomright: 'left-end',
  lefttop: 'bottom-start',
  leftcenter: 'right',
  leftbottom: 'top-start',
  righttop: 'bottom-end',
  rightcenter: 'left',
  rightbottom: 'top-end'
};
var PopperControl = /*#__PURE__*/function (_ButtonControl) {
  (0, _inherits2.default)(PopperControl, _ButtonControl);
  var _super = _createSuper(PopperControl);
  function PopperControl() {
    (0, _classCallCheck2.default)(this, PopperControl);
    return _super.apply(this, arguments);
  }
  (0, _createClass2.default)(PopperControl, [{
    key: "getPopper",
    value:
    /**
     * 气泡实例
     * @protected
     */

    function getPopper() {
      return this.popper;
    }
  }, {
    key: "hide",
    value: function hide() {
      this.popper.hide();
      (0, _get2.default)((0, _getPrototypeOf2.default)(PopperControl.prototype), "hide", this).call(this);
    }

    /**
     * 获取默认配置
     * @param option
     */
  }, {
    key: "getDefault",
    value: function getDefault(option) {
      var _option$position;
      var defaultOption = (0, _get2.default)((0, _getPrototypeOf2.default)(PopperControl.prototype), "getDefault", this).call(this, option);
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      var position = (_option$position = option === null || option === void 0 ? void 0 : option.position) !== null && _option$position !== void 0 ? _option$position : defaultOption.position;
      return (0, _objectSpread2.default)((0, _objectSpread2.default)({}, (0, _get2.default)((0, _getPrototypeOf2.default)(PopperControl.prototype), "getDefault", this).call(this, option)), {}, {
        popperPlacement: position instanceof Element ? 'bottom' : PopperPlacementMap[position],
        popperTrigger: 'click'
      });
    }
  }, {
    key: "onAdd",
    value: function onAdd() {
      var button = (0, _get2.default)((0, _getPrototypeOf2.default)(PopperControl.prototype), "onAdd", this).call(this);
      this.initPopper();
      return button;
    }
  }, {
    key: "onRemove",
    value: function onRemove() {
      this.popper.destroy();
    }
  }, {
    key: "initPopper",
    value: function initPopper() {
      var _this = this;
      var _this$controlOption = this.controlOption,
        popperClassName = _this$controlOption.popperClassName,
        popperPlacement = _this$controlOption.popperPlacement,
        popperTrigger = _this$controlOption.popperTrigger;
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      var popperContainer = this.mapsService.getMapContainer();

      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      this.popper = new _popper.Popper(this.button, {
        className: popperClassName,
        placement: popperPlacement,
        trigger: popperTrigger,
        container: popperContainer,
        unique: true
      });
      this.popper.on('show', function () {
        _this.emit('popperShow', _this);
      }).on('hide', function () {
        _this.emit('popperHide', _this);
      });
      return this.popper;
    }
  }, {
    key: "setOptions",
    value: function setOptions(option) {
      (0, _get2.default)((0, _getPrototypeOf2.default)(PopperControl.prototype), "setOptions", this).call(this, option);
      if (this.checkUpdateOption(option, ['popperPlacement', 'popperTrigger', 'popperClassName'])) {
        var content = this.popper.getContent();
        this.popper.destroy();
        this.initPopper();
        this.popper.setContent(content);
      }
    }
  }]);
  return PopperControl;
}(_buttonControl.default);
exports.default = exports.PopperControl = PopperControl;
