// @antv/l7-component/lib/control/baseControl/selectControl.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.SelectControl = void 0;
var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));
var _get2 = _interopRequireDefault(require("@babel/runtime/helpers/get"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _l7Utils = require("@antv/l7-utils");
var _popperControl = require("./popperControl");
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }
function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
var SelectControlConstant = /*#__PURE__*/function (SelectControlConstant) {
  SelectControlConstant["ActiveOptionClassName"] = "l7-select-control-item-active";
  SelectControlConstant["OptionValueAttrKey"] = "data-option-value";
  SelectControlConstant["OptionIndexAttrKey"] = "data-option-index";
  return SelectControlConstant;
}(SelectControlConstant || {});
var SelectControl = /*#__PURE__*/function (_PopperControl) {
  (0, _inherits2.default)(SelectControl, _PopperControl);
  var _super = _createSuper(SelectControl);
  function SelectControl() {
    var _this;
    (0, _classCallCheck2.default)(this, SelectControl);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "selectValue", []);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "createNormalOption", function (option) {
      var isSelect = _this.selectValue.includes(option.value);
      var optionDOM = _l7Utils.DOM.create('div', "l7-select-control-item ".concat(isSelect ? SelectControlConstant.ActiveOptionClassName : ''));
      if (_this.getIsMultiple()) {
        optionDOM.appendChild(_this.createCheckbox(isSelect));
      }
      if (option.icon) {
        optionDOM.appendChild(option.icon);
      }
      var textDOM = _l7Utils.DOM.create('span');
      textDOM.innerText = option.text;
      optionDOM.appendChild(textDOM);
      return optionDOM;
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onItemClick", function (item) {
      if (_this.getIsMultiple()) {
        var targetIndex = _this.selectValue.findIndex(function (value) {
          return value === item.value;
        });
        if (targetIndex > -1) {
          _this.selectValue.splice(targetIndex, 1);
        } else {
          _this.selectValue = [].concat((0, _toConsumableArray2.default)(_this.selectValue), [item.value]);
        }
      } else {
        _this.selectValue = [item.value];
      }
      _this.setSelectValue(_this.selectValue);
    });
    return _this;
  }
  (0, _createClass2.default)(SelectControl, [{
    key: "setOptions",
    value: function setOptions(option) {
      (0, _get2.default)((0, _getPrototypeOf2.default)(SelectControl.prototype), "setOptions", this).call(this, option);
      var options = option.options;
      if (options) {
        this.popper.setContent(this.getPopperContent(options));
      }
    }
  }, {
    key: "onAdd",
    value: function onAdd() {
      var button = (0, _get2.default)((0, _getPrototypeOf2.default)(SelectControl.prototype), "onAdd", this).call(this);
      var defaultValue = this.controlOption.defaultValue;
      if (defaultValue) {
        this.selectValue = this.transSelectValue(defaultValue);
      }
      this.popper.setContent(this.getPopperContent(this.controlOption.options));
      return button;
    }
  }, {
    key: "getSelectValue",
    value: function getSelectValue() {
      return this.getIsMultiple() ? this.selectValue : this.selectValue[0];
    }
  }, {
    key: "setSelectValue",
    value: function setSelectValue(value) {
      var _this2 = this;
      var emitEvent = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;
      var finalValue = this.transSelectValue(value);
      this.optionDOMList.forEach(function (optionDOM) {
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        var optionValue = optionDOM.getAttribute(SelectControlConstant.OptionValueAttrKey);
        var checkboxDOM = _this2.getIsMultiple() ? optionDOM.querySelector('input[type=checkbox]') : undefined;
        if (finalValue.includes(optionValue)) {
          _l7Utils.DOM.addClass(optionDOM, SelectControlConstant.ActiveOptionClassName);
          if (checkboxDOM) {
            // @ts-ignore
            _l7Utils.DOM.setChecked(checkboxDOM, true);
          }
        } else {
          _l7Utils.DOM.removeClass(optionDOM, SelectControlConstant.ActiveOptionClassName);
          if (checkboxDOM) {
            // @ts-ignore
            _l7Utils.DOM.setChecked(checkboxDOM, false);
          }
        }
      });
      this.selectValue = finalValue;
      if (emitEvent) {
        this.emit('selectChange', this.getIsMultiple() ? finalValue : finalValue[0]);
      }
    }

    /**
     * 是否为多选
     * @protected
     */
  }, {
    key: "getIsMultiple",
    value: function getIsMultiple() {
      return false;
    }
  }, {
    key: "getPopperContent",
    value: function getPopperContent(options) {
      var _this3 = this;
      var isImageOptions = this.isImageOptions();
      var content = _l7Utils.DOM.create('div', isImageOptions ? 'l7-select-control--image' : 'l7-select-control--normal');
      if (this.getIsMultiple()) {
        _l7Utils.DOM.addClass(content, 'l7-select-control--multiple');
      }
      var optionsDOMList = options.map(function (option, optionIndex) {
        var optionDOM = isImageOptions ?
        // @ts-ignore
        _this3.createImageOption(option) : _this3.createNormalOption(option);
        optionDOM.setAttribute(SelectControlConstant.OptionValueAttrKey, option.value);
        optionDOM.setAttribute(SelectControlConstant.OptionIndexAttrKey, window.String(optionIndex));
        optionDOM.addEventListener('click', _this3.onItemClick.bind(_this3, option));
        return optionDOM;
      });
      content.append.apply(content, (0, _toConsumableArray2.default)(optionsDOMList));
      this.optionDOMList = optionsDOMList;
      return content;
    }
  }, {
    key: "createImageOption",
    value: function createImageOption(option) {
      var isSelect = this.selectValue.includes(option.value);
      var optionDOM = _l7Utils.DOM.create('div', "l7-select-control-item ".concat(isSelect ? SelectControlConstant.ActiveOptionClassName : ''));
      var imgDOM = _l7Utils.DOM.create('img');
      imgDOM.setAttribute('src', option.img);
      _l7Utils.DOM.setUnDraggable(imgDOM);
      optionDOM.appendChild(imgDOM);
      var rowDOM = _l7Utils.DOM.create('div', 'l7-select-control-item-row');
      if (this.getIsMultiple()) {
        optionDOM.appendChild(this.createCheckbox(isSelect));
      }
      var textDOM = _l7Utils.DOM.create('span');
      textDOM.innerText = option.text;
      rowDOM.appendChild(textDOM);
      optionDOM.appendChild(rowDOM);
      return optionDOM;
    }
  }, {
    key: "createCheckbox",
    value: function createCheckbox(isSelect) {
      var checkboxDOM = _l7Utils.DOM.create('input');
      checkboxDOM.setAttribute('type', 'checkbox');
      if (isSelect) {
        _l7Utils.DOM.setChecked(checkboxDOM, true);
      }
      return checkboxDOM;
    }
  }, {
    key: "isImageOptions",
    value: function isImageOptions() {
      // @ts-ignore
      return !!this.controlOption.options.find(function (item) {
        return item.img;
      });
    }
  }, {
    key: "transSelectValue",
    value: function transSelectValue(value) {
      return Array.isArray(value) ? value : [value];
    }
  }]);
  return SelectControl;
}(_popperControl.PopperControl);
exports.default = exports.SelectControl = SelectControl;
