// antd-token-previewer/es/overviews/Error.js
import { Card } from 'antd';
import React from 'react';
import { Flexbox } from '@arvinxu/layout-kit';
import Alert from "../component-demos/alert/error";
import Button from "../component-demos/button/dangerButton";
import Message from "../component-demos/message/error";
import Progress from "../component-demos/progress/danger";
import Tag from "../component-demos/tag/error";
import Badge from "../component-demos/badge/badge";
import Menu from "../component-demos/menu/menuDanger";
import Upload from "../component-demos/upload/danger";
import Dropdown from "../component-demos/dropdown/dropdownError";
import Notification from "../component-demos/notification/error";
import Timeline from "../component-demos/timeline/danger";
export var Error = function Error() {
  return /*#__PURE__*/React.createElement(Card, {
    size: 'small'
  }, /*#__PURE__*/React.createElement(Flexbox, {
    horizontal: true,
    align: 'start',
    gap: 24
  }, /*#__PURE__*/React.createElement(Flexbox, {
    gap: 24
  }, /*#__PURE__*/React.createElement(Flexbox, {
    horizontal: true,
    align: 'center',
    gap: 12,
    style: {
      marginTop: 8
    }
  }, Button.demo, /*#__PURE__*/React.createElement("div", null, Tag.demo), Badge.demo), Alert.demo), /*#__PURE__*/React.createElement(Flexbox, {
    align: 'center',
    gap: 28
  }, Message.demo, Progress.demo)), /*#__PURE__*/React.createElement(Flexbox, {
    horizontal: true,
    gap: 40,
    style: {
      marginTop: 32
    }
  }, /*#__PURE__*/React.createElement("div", null, Notification.demo), /*#__PURE__*/React.createElement("div", null, Timeline.demo)), /*#__PURE__*/React.createElement(Flexbox, {
    horizontal: true,
    gap: 40
  }, Menu.demo, /*#__PURE__*/React.createElement("div", {
    style: {
      width: 300
    }
  }, Upload.demo), Dropdown.demo));
};
