// antd-token-previewer/es/overviews/Primary.js
import { Card, Space } from 'antd';
import React from 'react';
import Menu from "../component-demos/menu/menu";
import SelectTag from "../component-demos/select/selectTag";
import Button from "../component-demos/button/button-icon";
import Switch from "../component-demos/switch/switch";
import Radio from "../component-demos/radio/radio";
import RadioButton from "../component-demos/radio/button";
import Checkbox from "../component-demos/checkbox/checkbox";
import Tabs from "../component-demos/tabs/tabs";
import Pagination from "../component-demos/pagination/outline";
import Steps from "../component-demos/steps/steps";
import Popconfirm from "../component-demos/popconfirm/popconfirm";
import Timeline from "../component-demos/timeline/timeline";
import Table from "../component-demos/table/table";
export var Primary = function Primary() {
  return /*#__PURE__*/React.createElement(Card, {
    size: 'small'
  }, /*#__PURE__*/React.createElement(Space, {
    direction: 'vertical'
  }, /*#__PURE__*/React.createElement(Space, {
    align: 'start',
    size: 'large'
  }, Menu.demo, /*#__PURE__*/React.createElement(Space, {
    direction: 'vertical',
    size: 'large'
  }, /*#__PURE__*/React.createElement(Space, {
    size: 'large',
    align: 'start'
  }, /*#__PURE__*/React.createElement(Space, {
    direction: 'vertical',
    size: 'large'
  }, /*#__PURE__*/React.createElement("div", null, Button.demo), /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("span", null, Radio.demo), Checkbox.demo, Switch.demo), /*#__PURE__*/React.createElement("div", null, RadioButton.demo), Tabs.demo), SelectTag.demo), Pagination.demo, /*#__PURE__*/React.createElement("div", {
    style: {
      padding: 12
    }
  }, Steps.demo), /*#__PURE__*/React.createElement(Space, {
    size: 'large',
    align: 'start'
  }, Popconfirm.demo, Timeline.demo))), Table.demo));
};
