// antd-token-previewer/es/overviews/Success.js
import { Card } from 'antd';
import React from 'react';
import Alert from "../component-demos/alert/success";
import Message from "../component-demos/message/success";
import Progress from "../component-demos/progress/success";
import Tag from "../component-demos/tag/success";
import Input from "../component-demos/input/success";
import Result from "../component-demos/result/success";
import Notification from "../component-demos/notification/success";
import Timeline from "../component-demos/timeline/success";
import { Flexbox } from '@arvinxu/layout-kit';
export var Success = function Success() {
  return /*#__PURE__*/React.createElement(Card, {
    size: 'small'
  }, /*#__PURE__*/React.createElement(Flexbox, {
    horizontal: true,
    align: 'start',
    gap: 24
  }, /*#__PURE__*/React.createElement(Flexbox, {
    gap: 40
  }, /*#__PURE__*/React.createElement(Flexbox, {
    horizontal: true,
    align: 'center',
    gap: 12
  }, /*#__PURE__*/React.createElement("div", null, Tag.demo), Input.demo), Alert.demo), /*#__PURE__*/React.createElement(Flexbox, {
    align: 'center',
    gap: 28
  }, Message.demo, Progress.demo)), /*#__PURE__*/React.createElement(Flexbox, {
    horizontal: true,
    gap: 40,
    style: {
      marginTop: 32
    }
  }, /*#__PURE__*/React.createElement("div", null, Notification.demo), /*#__PURE__*/React.createElement("div", null, Timeline.demo)), Result.demo);
};
