// antd-token-previewer/es/overviews/Warning.js
import { Card } from 'antd';
import React from 'react';
import Alert from "../component-demos/alert/warning";
import Message from "../component-demos/message/warning";
import Popconfirm from "../component-demos/popconfirm/popconfirm";
import Modal from "../component-demos/modal/warning";
import Badge from "../component-demos/badge/warning";
import Text from "../component-demos/typography/warningText";
import Title from "../component-demos/typography/warningTitle";
import Tag from "../component-demos/tag/warning";
import Input from "../component-demos/input/warning";
import Result from "../component-demos/result/warning";
import Notification from "../component-demos/notification/warning";
import { Flexbox } from '@arvinxu/layout-kit';
export var Warning = function Warning() {
  return /*#__PURE__*/React.createElement(Card, {
    size: 'small'
  }, /*#__PURE__*/React.createElement(Flexbox, {
    horizontal: true,
    align: 'start',
    gap: 24
  }, /*#__PURE__*/React.createElement(Flexbox, {
    gap: 24
  }, /*#__PURE__*/React.createElement(Flexbox, {
    horizontal: true,
    gap: 12
  }, /*#__PURE__*/React.createElement("div", {
    style: {
      width: 200
    }
  }, Title.demo), /*#__PURE__*/React.createElement("div", {
    style: {
      width: '100%'
    }
  }, Input.demo)), Alert.demo), /*#__PURE__*/React.createElement(Flexbox, {
    align: 'center',
    gap: 28
  }, Message.demo, Popconfirm.demo, /*#__PURE__*/React.createElement(Flexbox, {
    horizontal: true,
    gap: 16
  }, Badge.demo, Tag.demo, Text.demo))), /*#__PURE__*/React.createElement(Flexbox, {
    horizontal: true,
    gap: 24,
    style: {
      marginTop: 32
    }
  }, /*#__PURE__*/React.createElement("div", null, Notification.demo), /*#__PURE__*/React.createElement("div", null, Modal.demo)), Result.demo);
};
