// @antv/l7-layers/lib/core/interface.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.lineStyleType = exports.TextureBlend = exports.SizeUnitType = exports.LinearDir = exports.CanvasUpdateType = void 0;
var lineStyleType = /*#__PURE__*/function (lineStyleType) {
  lineStyleType[lineStyleType["solid"] = 0] = "solid";
  lineStyleType[lineStyleType["dash"] = 1] = "dash";
  return lineStyleType;
}({});
exports.lineStyleType = lineStyleType;
var LinearDir = /*#__PURE__*/function (LinearDir) {
  LinearDir["VERTICAL"] = "vertical";
  LinearDir["HORIZONTAL"] = "horizontal";
  return LinearDir;
}({});
exports.LinearDir = LinearDir;
var TextureBlend = /*#__PURE__*/function (TextureBlend) {
  TextureBlend["NORMAL"] = "normal";
  TextureBlend["REPLACE"] = "replace";
  return TextureBlend;
}({});
/**
 * 基础图层类型定义
 */
exports.TextureBlend = TextureBlend;
var SizeUnitType = /*#__PURE__*/function (SizeUnitType) {
  SizeUnitType[SizeUnitType["pixel"] = 0] = "pixel";
  SizeUnitType[SizeUnitType["meter"] = 1] = "meter";
  return SizeUnitType;
}({});
exports.SizeUnitType = SizeUnitType;
var CanvasUpdateType = /*#__PURE__*/function (CanvasUpdateType) {
  CanvasUpdateType["ALWAYS"] = "always";
  CanvasUpdateType["DRAGEND"] = "dragend";
  return CanvasUpdateType;
}({});
exports.CanvasUpdateType = CanvasUpdateType;
