// @antv/xflow-extension/lib/flowchart-node-panel/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FlowchartNodePanel = exports.NodePanelMain = exports.setGroupRender = void 0;
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importStar(require("react"));
var xflow_core_1 = require("@antv/xflow-core");
var icons_1 = require("@ant-design/icons");
var lodash_1 = require("lodash");
var base_panel_1 = require("../base-panel");
var utils_1 = require("./utils");
var panel_body_1 = require("./panel-body");
// import { NodePanelHeader } from '../canvas-node-tree-panel/panel-header'
var panel_header_1 = require("./panel-header");
var utils_2 = require("../canvas-node-tree-panel/utils");
var service_1 = require("./service");
var constants_1 = require("./constants");
var group_panel_1 = require("./group-panel");
Object.defineProperty(exports, "setGroupRender", { enumerable: true, get: function () { return group_panel_1.setGroupRender; } });
tslib_1.__exportStar(require("./constants"), exports);
tslib_1.__exportStar(require("./utils"), exports);
tslib_1.__exportStar(require("./interface"), exports);
var NodePanelMain = function (props) {
    var prefixClz = props.prefixClz, _a = props.position, position = _a === void 0 ? { width: 240, top: 0, bottom: 0, left: 0 } : _a, _b = props.showHeader, showHeader = _b === void 0 ? true : _b, rest = tslib_1.__rest(props, ["prefixClz", "position", "showHeader"]);
    var _c = position.width, width = _c === void 0 ? 200 : _c;
    var _d = (0, utils_2.usePanelLyaoutStyle)(props), headerStyle = _d.headerStyle, bodyStyle = _d.bodyStyle;
    var _e = (0, service_1.usePanelData)(props), state = _e.state, onKeywordChange = _e.onKeywordChange;
    return (react_1.default.createElement(react_1.default.Fragment, null,
        showHeader && (react_1.default.createElement(base_panel_1.WorkspacePanel, tslib_1.__assign({}, rest, { position: {
                top: 0,
                left: 0,
                height: constants_1.PANEL_HEADER_HEIGHT,
                width: width,
            } }),
            react_1.default.createElement(panel_header_1.NodePanelHeader, tslib_1.__assign({}, props, { state: state, style: headerStyle, onKeywordChange: onKeywordChange })))),
        react_1.default.createElement(base_panel_1.WorkspacePanel, tslib_1.__assign({ className: "".concat(constants_1.CONTAINER_CLASS, "-nodes") }, rest, { position: tslib_1.__assign(tslib_1.__assign({}, position), { top: showHeader ? constants_1.PANEL_HEADER_HEIGHT : 0 }) }),
            react_1.default.createElement(panel_body_1.NodePanelBody, tslib_1.__assign({}, props, { state: state, style: bodyStyle })))));
};
exports.NodePanelMain = NodePanelMain;
var FlowchartNodePanel = function (props) {
    (0, utils_1.registerCustomNode)((0, lodash_1.get)(props, 'registerNode'));
    var prefixClz = (0, xflow_core_1.useXflowPrefixCls)('node-panel');
    var _a = (0, react_1.useState)(false), collpased = _a[0], setCollpased = _a[1];
    var _b = props.show, show = _b === void 0 ? true : _b, _c = props.position, position = _c === void 0 ? { width: 240, top: 40, bottom: 0, left: 0 } : _c, rest = tslib_1.__rest(props, ["show", "position"]);
    if (!show) {
        return null;
    }
    var _d = position.width, width = _d === void 0 ? 200 : _d, left = position.left;
    return (react_1.default.createElement(base_panel_1.WorkspacePanel, { className: constants_1.CONTAINER_CLASS, position: tslib_1.__assign(tslib_1.__assign({}, position), { left: !collpased ? left : -width }) },
        react_1.default.createElement("div", { className: "".concat(constants_1.CONTAINER_CLASS, "-wrapper") },
            react_1.default.createElement(base_panel_1.WorkspacePanel, tslib_1.__assign({ className: prefixClz }, rest, { position: {
                    top: 0,
                    bottom: 0,
                    left: 0,
                    right: 0,
                } }),
                react_1.default.createElement(exports.NodePanelMain, tslib_1.__assign({}, props, { prefixClz: prefixClz, position: position }))),
            react_1.default.createElement("div", { className: "".concat(constants_1.CONTAINER_CLASS, "-icon"), style: {
                    top: 21,
                    right: !collpased ? -10 : -20,
                    borderRadius: !collpased ? '50%' : '0 50% 50% 0',
                    borderLeft: !collpased ? '' : 'none',
                }, onClick: function () {
                    setCollpased(!collpased);
                } }, collpased ? react_1.default.createElement(icons_1.DoubleRightOutlined, null) : react_1.default.createElement(icons_1.DoubleLeftOutlined, null)))));
};
exports.FlowchartNodePanel = FlowchartNodePanel;
