// @antv/xflow-extension/lib/flowchart-node-panel/panel-body.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodePanelBody = void 0;
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importStar(require("react"));
var antd_1 = require("antd");
var x6_1 = require("@antv/x6");
var xflow_core_1 = require("@antv/xflow-core");
var utils_1 = require("../flowchart-canvas/utils");
var panel_body_1 = require("../canvas-node-tree-panel/panel-body");
var lodash_1 = require("lodash");
var Panel = antd_1.Collapse.Panel;
var NodePanelBody = function (props) {
    var x6NodeFactory = props.x6NodeFactory, dndOptions = props.dndOptions, state = props.state, prefixClz = props.prefixClz, _a = props.defaultActiveKey, defaultActiveKey = _a === void 0 ? ['official'] : _a, _b = props.showOfficial, showOfficial = _b === void 0 ? true : _b;
    var registerNode = props.registerNode
        ? (0, lodash_1.isArray)(props.registerNode)
            ? props.registerNode
            : [props.registerNode]
        : [];
    var _c = (0, xflow_core_1.useXFlowApp)(), graphProvider = _c.graphProvider, modelService = _c.modelService, commandService = _c.commandService;
    var _d = react_1.default.useState(), dnd = _d[0], setDnd = _d[1];
    /** 获取graph实例 */
    var _e = react_1.default.useState(), graph = _e[0], setGraph = _e[1];
    graphProvider.getGraphInstance().then(function (x6Graph) {
        setGraph(x6Graph);
    });
    var graphConfig = undefined;
    graphProvider.getGraphOptions().then(function (x6GraphConfig) {
        graphConfig = x6GraphConfig;
    });
    var onNodeDrop = (0, react_1.useCallback)(function (node) { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
        var ports, nodeConfig, args, onAddNode;
        var _a;
        return tslib_1.__generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    ports = node.ports;
                    nodeConfig = tslib_1.__assign(tslib_1.__assign({}, node), { id: "node-".concat((0, xflow_core_1.uuidv4)()), zIndex: 10, ports: tslib_1.__assign(tslib_1.__assign({}, ports), { items: (_a = ports.items) === null || _a === void 0 ? void 0 : _a.map(function (item) { return (tslib_1.__assign(tslib_1.__assign({}, item), { id: (0, xflow_core_1.uuidv4)() })); }) }) });
                    args = {
                        nodeConfig: nodeConfig,
                    };
                    return [4 /*yield*/, commandService.executeCommand(xflow_core_1.XFlowNodeCommands.ADD_NODE.id, args)];
                case 1:
                    _b.sent();
                    onAddNode = (0, utils_1.getProps)('onAddNode');
                    if (typeof onAddNode === 'function') {
                        onAddNode(nodeConfig);
                    }
                    return [2 /*return*/];
            }
        });
    }); }, [commandService]);
    react_1.default.useEffect(function () {
        if (!graph) {
            return;
        }
        var dndInstance = new x6_1.Addon.Dnd(tslib_1.__assign(tslib_1.__assign({ scaled: false, animation: false }, dndOptions), { target: graph, 
            /** 这里考虑到需要新增群组的需求，不使用x6的getDropNod方法
             * 在validateNode时调用command添加
             */
            validateNode: function (droppingNode) { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
                var nodeConfig;
                return tslib_1.__generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            nodeConfig = tslib_1.__assign(tslib_1.__assign({}, droppingNode.getData()), droppingNode.getPosition());
                            return [4 /*yield*/, onNodeDrop(nodeConfig)];
                        case 1:
                            _a.sent();
                            return [2 /*return*/, false];
                    }
                });
            }); } }));
        setDnd(dndInstance);
    }, [commandService, dndOptions, graph, modelService, onNodeDrop]);
    var onMouseDown = react_1.default.useCallback(function (nodeConfig) { return function (e) {
        if (!graph || !dnd || !graphConfig) {
            return;
        }
        var renderKey = graphConfig.nodeTypeParser(nodeConfig);
        var _a = nodeConfig.width, width = _a === void 0 ? 180 : _a, _b = nodeConfig.height, height = _b === void 0 ? 40 : _b;
        var reactComponent = graphConfig.nodeRender.get(renderKey);
        var wrappedComponent = (0, xflow_core_1.getNodeReactComponent)(reactComponent, commandService, modelService);
        var nodeData = {
            data: nodeConfig,
            width: width,
            height: height,
            view: graphConfig.graphId,
            component: wrappedComponent,
        };
        var x6Node = x6NodeFactory ? x6NodeFactory(nodeData) : (0, panel_body_1.defaultNodeFactory)(nodeData);
        dnd.start(x6Node, e.nativeEvent);
    }; }, [commandService, dnd, graph, graphConfig, modelService, x6NodeFactory]);
    var renderTree = react_1.default.useCallback(function (list) {
        if (list === void 0) { list = []; }
        return list.map(function (item) {
            var popoverContent = item.popoverContent;
            return (react_1.default.createElement(panel_body_1.NodeTitle, { item: item, key: item.id, onMouseDown: onMouseDown(item), popoverContent: popoverContent, prefixClz: prefixClz, modelService: modelService, commandService: commandService, graphConfig: graphConfig }));
        });
    }, [commandService, graphConfig, modelService, onMouseDown, prefixClz]);
    var officialNode = state.nodeList.filter(function (item) { return !item.isCustom; });
    var searchOfficialNode = state.searchList.filter(function (item) { return !item.isCustom; });
    var customNode = function (key) {
        return state.nodeList.filter(function (item) { return item.isCustom && item.parentKey === key; });
    };
    var searchCustomNode = function (key) {
        return state.searchList.filter(function (item) { return item.isCustom && item.parentKey === key; });
    };
    return (react_1.default.createElement(react_1.default.Fragment, null,
        react_1.default.createElement("div", { className: "".concat(prefixClz, "-body") },
            react_1.default.createElement(antd_1.Collapse, { defaultActiveKey: defaultActiveKey, style: { border: 'none' } },
                showOfficial && (react_1.default.createElement(Panel, { header: "\u901A\u7528\u8282\u70B9", key: "official", style: { border: 'none' } },
                    !state.keyword && (react_1.default.createElement("div", { className: "".concat(prefixClz, "-official") }, renderTree(officialNode))),
                    state.searchList.length > 0 && (react_1.default.createElement("div", { className: "".concat(prefixClz, "-official") }, renderTree(searchOfficialNode))))),
                (registerNode === null || registerNode === void 0 ? void 0 : registerNode.length) > 0 &&
                    registerNode.map(function (item) {
                        return !item.hidden &&
                            item.nodes.length > 0 && (react_1.default.createElement(Panel, { header: item.title, key: item.key, style: { border: 'none' } },
                            !state.keyword && (react_1.default.createElement("div", { className: "".concat(prefixClz, "-custom") }, renderTree(customNode(item.key)))),
                            state.searchList.length > 0 && (react_1.default.createElement("div", { className: "".concat(prefixClz, "-custom") }, renderTree(searchCustomNode(item.key))))));
                    })),
            state.keyword && state.searchList.length === 0 && react_1.default.createElement(antd_1.Empty, { style: { marginTop: '48px' } }))));
};
exports.NodePanelBody = NodePanelBody;
