// @antv/xflow-extension/lib/flowchart-node-panel/service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.usePanelData = exports.NsPanelData = void 0;
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importDefault(require("react"));
var xflow_core_1 = require("@antv/xflow-core");
var utils_1 = require("./utils");
var lodash_1 = require("lodash");
var NsPanelData;
(function (NsPanelData) {
    NsPanelData.id = 'NODE_PANEL_DATA';
})(NsPanelData = exports.NsPanelData || (exports.NsPanelData = {}));
/** 节点查找 */
var DefaultsearchService = function (nodeList, keyword) {
    if (nodeList === void 0) { nodeList = []; }
    return tslib_1.__awaiter(void 0, void 0, void 0, function () {
        var list;
        return tslib_1.__generator(this, function (_a) {
            list = nodeList.filter(function (i) { var _a, _b; return i.isDirectory || ((_a = i.label) === null || _a === void 0 ? void 0 : _a.includes(keyword)) || ((_b = i.name) === null || _b === void 0 ? void 0 : _b.includes(keyword)); });
            return [2 /*return*/, list];
        });
    });
};
var usePanelData = function (props) {
    var _a = props.searchService, searchService = _a === void 0 ? DefaultsearchService : _a;
    var registerNode = (props.registerNode
        ? (0, lodash_1.isArray)(props.registerNode)
            ? props.registerNode
            : [props.registerNode]
        : []);
    var nodes = [];
    registerNode.forEach(function (item) {
        nodes = nodes.concat(item.nodes.map(function (node) { return (tslib_1.__assign(tslib_1.__assign({}, node), { parentKey: item.key })); }));
    });
    var modelService = (0, xflow_core_1.useXFlowApp)().modelService;
    /** 使用model */
    var _b = (0, xflow_core_1.createComponentModel)({
        searchList: [],
        nodeList: [],
        defaultExpandAll: false,
        keyword: '',
    }), state = _b[0], setState = _b[1], panelModel = _b[2];
    /** 注册成为全局状态，方便其他组件联动 */
    react_1.default.useEffect(function () {
        if (modelService.findDeferredModel(NsPanelData.id)) {
            return;
        }
        modelService.registerModel({
            id: NsPanelData.id,
            modelFactory: function () { return panelModel; },
            watchChange: function (self) { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
                var graphMetaModel, fetch, graphMetaDisposable;
                return tslib_1.__generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, xflow_core_1.MODELS.GRAPH_META.getModel(modelService)]; //useContext(MODELS.GRAPH_META.id)
                        case 1:
                            graphMetaModel = _a.sent() //useContext(MODELS.GRAPH_META.id)
                            ;
                            fetch = function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
                                var listData;
                                return tslib_1.__generator(this, function (_a) {
                                    switch (_a.label) {
                                        case 0: return [4 /*yield*/, (0, utils_1.nodeService)(nodes)];
                                        case 1:
                                            listData = _a.sent();
                                            return [2 /*return*/, { listData: listData }];
                                    }
                                });
                            }); };
                            graphMetaDisposable = graphMetaModel.watch(function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
                                var data;
                                return tslib_1.__generator(this, function (_a) {
                                    switch (_a.label) {
                                        case 0: return [4 /*yield*/, fetch()];
                                        case 1:
                                            data = _a.sent();
                                            self.setValue({
                                                nodeList: data.listData,
                                                defaultExpandAll: false,
                                                keyword: '',
                                                searchList: [],
                                            });
                                            return [2 /*return*/];
                                    }
                                });
                            }); });
                            return [2 /*return*/, xflow_core_1.Disposable.create(function () {
                                    graphMetaDisposable.dispose();
                                })];
                    }
                });
            }); },
        });
        /* eslint-disable-next-line  */
    }, []);
    /** 搜索 */
    var onKeywordChange = react_1.default.useCallback(function (keyword) { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
        var list_1;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!searchService) {
                        return [2 /*return*/];
                    }
                    if (!keyword) return [3 /*break*/, 2];
                    return [4 /*yield*/, searchService(state.nodeList, keyword)];
                case 1:
                    list_1 = _a.sent();
                    setState(function (modelState) {
                        modelState.keyword = keyword;
                        modelState.searchList = list_1;
                    });
                    return [3 /*break*/, 3];
                case 2:
                    setState(function (modelState) {
                        modelState.keyword = '';
                        modelState.searchList = [];
                    });
                    _a.label = 3;
                case 3: return [2 /*return*/];
            }
        });
    }); }, [searchService, state.nodeList, setState]);
    return {
        state: state,
        setState: setState,
        onKeywordChange: onKeywordChange,
    };
};
exports.usePanelData = usePanelData;
