// @antv/xflow-extension/lib/flowchart-node-panel/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createPath = exports.setNodeRender = exports.registerCustomNode = exports.nodeService = exports.getRegisterNode = void 0;
var tslib_1 = require("tslib");
var xflow_core_1 = require("@antv/xflow-core");
var lodash_1 = require("lodash");
var utils_1 = require("../flowchart-canvas/utils");
var NodesComponent = tslib_1.__importStar(require("./nodes"));
var constants_1 = require("./constants");
var lodash_2 = require("lodash");
/** 和 graph config 注册的节点保持一致 */
var getAnchorStyle = function (position) {
    return {
        position: { name: position },
        attrs: {
            circle: {
                r: 4,
                magnet: true,
                stroke: '#31d0c6',
                strokeWidth: 2,
                fill: '#fff',
                style: {
                    visibility: 'hidden',
                },
            },
        },
        zIndex: 10,
    };
};
var getPorts = function (position) {
    if (position === void 0) { position = ['top', 'right', 'bottom', 'left']; }
    return {
        items: position.map(function (name) {
            return { group: name, id: (0, xflow_core_1.uuidv4)() };
        }),
        groups: {
            top: getAnchorStyle('top'),
            right: getAnchorStyle('right'),
            bottom: getAnchorStyle('bottom'),
            left: getAnchorStyle('left'),
        },
    };
};
var getRegisterNode = function (nodes) {
    return (nodes || []).map(function (item) {
        var name = item.name, popover = item.popover, _a = item.label, label = _a === void 0 ? '' : _a, _b = item.width, width = _b === void 0 ? constants_1.NODE_HEIGHT : _b, _c = item.height, height = _c === void 0 ? constants_1.NODE_HEIGHT : _c, ports = item.ports, parentKey = item.parentKey;
        var id = (0, xflow_core_1.uuidv4)(); // 暂不使用上层数据
        return {
            id: id,
            renderKey: name,
            name: name,
            label: label,
            popoverContent: popover,
            width: width,
            height: height,
            ports: ports || getPorts(),
            originData: tslib_1.__assign({}, item),
            isCustom: true,
            parentKey: parentKey,
        };
    });
};
exports.getRegisterNode = getRegisterNode;
var nodeService = function (nodes) { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
    var customNodes;
    return tslib_1.__generator(this, function (_a) {
        customNodes = (0, exports.getRegisterNode)(nodes);
        return [2 /*return*/, tslib_1.__spreadArray(tslib_1.__spreadArray([], customNodes, true), constants_1.NODEPOOL.map(function (_a) {
                var name = _a.name, ports = _a.ports, _b = _a.width, width = _b === void 0 ? constants_1.NODE_WIDTH : _b, _c = _a.height, height = _c === void 0 ? constants_1.NODE_HEIGHT : _c, _d = _a.label, label = _d === void 0 ? '' : _d;
                return {
                    id: (0, xflow_core_1.uuidv4)(),
                    renderKey: name,
                    name: name,
                    label: label,
                    popoverContent: function () { return name; },
                    width: width,
                    height: height,
                    ports: getPorts(ports),
                };
            }), true)];
    });
}); };
exports.nodeService = nodeService;
var registerCustomNode = function (panelConfigs) {
    var registerNodes = (panelConfigs ? ((0, lodash_2.isArray)(panelConfigs) ? panelConfigs : [panelConfigs]) : []);
    var nodes = [];
    registerNodes.forEach(function (item) {
        nodes = nodes.concat(item.nodes.map(function (node) { return (tslib_1.__assign(tslib_1.__assign({}, node), { parentKey: item.key })); }));
    });
    if (nodes.length) {
        (0, utils_1.setProps)({
            registerNode: nodes,
        });
    }
    var graphConfig = (0, utils_1.getProps)('graphConfig');
    var registerNode = (0, utils_1.getProps)('registerNode');
    if (!graphConfig || !(registerNode === null || registerNode === void 0 ? void 0 : registerNode.length)) {
        return;
    }
    registerNode.forEach(function (item) {
        var name = item.name, component = item.component;
        graphConfig.setNodeRender(name, component);
    });
};
exports.registerCustomNode = registerCustomNode;
var setNodeRender = function (graphConfig) {
    /** props 共享 */
    (0, exports.registerCustomNode)();
    /** 默认节点，通过 Terminal 标识，避免多次调用*/
    if (!graphConfig.nodeRender.get('Terminal')) {
        constants_1.NODEPOOL.forEach(function (item) {
            ;
            graphConfig.setNodeRender(item.name, NodesComponent["".concat(item.name.replace(/\s+/g, ''), "Node")]);
        });
    }
};
exports.setNodeRender = setNodeRender;
// 创建节点路径
var createPath = function (paths, offsetX, offsetY) {
    if (offsetX === void 0) { offsetX = 0; }
    if (offsetY === void 0) { offsetY = 0; }
    if (!paths.length) {
        return null;
    }
    var path = '';
    // @ts-ignore
    paths.forEach(function (item) {
        var c = item[0], x = item[1], y = item[2], c2x = item[3], c2y = item[4];
        path += (0, lodash_1.isNumber)(y) ? " ".concat(c, " ").concat(x + offsetX, " ").concat(y + offsetY) : " ".concat(c);
        if (c2y) {
            path += " ".concat(c2x + offsetX, " ").concat(c2y + offsetY);
        }
    });
    return path;
};
exports.createPath = createPath;
