// @antv/xflow-core/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.XFLOW_GROUP_DEFAULT_COLLAPSED_SIZE = exports.DEFAULT_NODE_WIDTH = exports.DEFAULT_NODE_HEIGHT = exports.LOCAL_STORAGE_KEY = exports.XFLOW_DEFAULT_EDGE = exports.XFLOW_DEFAULT_GROUP_NODE = exports.XFLOW_DEFAULT_NODE = exports.MAX_ZOOM = exports.MIN_ZOOM = exports.XFLOW_PREFIX_CLS = exports.X6_NODE_PORTAL_NODE_VIEW = void 0;
// node portal id
exports.X6_NODE_PORTAL_NODE_VIEW = 'X6_NODE_PORTAL_NODE_VIEW';
// style classname prefix
exports.XFLOW_PREFIX_CLS = 'xflow';
/** 画布缩放最小倍数 */
exports.MIN_ZOOM = 0.01;
/** 画布缩放最大倍数 */
exports.MAX_ZOOM = 1.5;
/** node render 的默认key值 */
exports.XFLOW_DEFAULT_NODE = 'XFLOW_DEFAULT_NODE';
/** node render 的默认key值 */
exports.XFLOW_DEFAULT_GROUP_NODE = 'XFLOW_DEFAULT_GROUP_NODE';
/** edge render 的默认key值 */
exports.XFLOW_DEFAULT_EDGE = 'XFLOW_DEFAULT_EDGE';
/** localStorage key */
exports.LOCAL_STORAGE_KEY = 'XFLOW_COPY_ITEMS';
/** node height */
exports.DEFAULT_NODE_HEIGHT = 36;
/** node width */
exports.DEFAULT_NODE_WIDTH = 180;
/** default group size */
exports.XFLOW_GROUP_DEFAULT_COLLAPSED_SIZE = {
    width: exports.DEFAULT_NODE_WIDTH,
    height: exports.DEFAULT_NODE_HEIGHT,
};
