// @antv/xflow-core/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IMenuService = exports.IToolbarContribution = exports.IToolbarService = exports.ToolbarRegistry = exports.useIsMountedRef = exports.useModelAsync = exports.createComponentModel = exports.useModel = exports.NsModel = exports.RxModel = exports.createModelServiceConfig = exports.IModelContribution = exports.IModelService = exports.MODELS = exports.createHookConfig = exports.IHookContribution = exports.IHookService = exports.getNodeReactComponent = exports.useAppContext = exports.AppContext = exports.createCmdConfig = exports.XFlowModelCommands = exports.XFlowGraphCommands = exports.XFlowGroupCommands = exports.XFlowEdgeCommands = exports.XFlowNodeCommands = exports.commandRegistryModule = exports.GraphCommandRegistry = exports.IGraphCommandFactory = exports.IGraphCommandContribution = exports.ICommandContextProvider = exports.IGraphCommandService = exports.ICommandHandler = exports.IApplicationContribution = exports.IApplication = exports.GraphConfig = exports.createGraphConfig = exports.IGraphProvider = exports.useExtensionRegistry = exports.ExtensionRegistryContext = exports.XFlowConfigProviderContext = exports.useXflowPrefixCls = exports.useXFlowApp = exports.XFlowAppContext = exports.XFlowAppProvider = exports.XFlowAppExtensionModule = exports.XFlowCanvas = exports.XFlow = exports.NsGraph = exports.ManaSyringe = void 0;
exports.XFlowConstants = exports.IconStore = exports.isReactComponent = exports.insertCss = exports.usePositionStyle = exports.uuidv4 = exports.delay = exports.Deferred = exports.DisposableCollection = exports.Disposable = exports.IKeyBindingContribution = exports.KeybindingConfig = exports.createKeybindingConfig = exports.KeyBindings = exports.MenuRegistry = exports.MenuItemType = exports.IMenuContribution = void 0;
var tslib_1 = require("tslib");
// IOC
var ManaSyringe = tslib_1.__importStar(require("mana-syringe"));
exports.ManaSyringe = ManaSyringe;
/** XFlow 基础 interface */
var interface_1 = require("./interface");
Object.defineProperty(exports, "NsGraph", { enumerable: true, get: function () { return interface_1.NsGraph; } });
/*******************************************************
 * 核心组件：
 * 1. Application：XFlow
 * 2. Application Extension：XFlowAppExtensionModule
 * 3. Graph：XFlowCanvas
 *****************************************************/
var xflow_main_1 = require("./xflow-main");
// 组件
Object.defineProperty(exports, "XFlow", { enumerable: true, get: function () { return xflow_main_1.XFlow; } });
Object.defineProperty(exports, "XFlowCanvas", { enumerable: true, get: function () { return xflow_main_1.XFlowCanvas; } });
Object.defineProperty(exports, "XFlowAppExtensionModule", { enumerable: true, get: function () { return xflow_main_1.XFlowAppExtensionModule; } });
// app context
Object.defineProperty(exports, "XFlowAppProvider", { enumerable: true, get: function () { return xflow_main_1.XFlowAppProvider; } });
Object.defineProperty(exports, "XFlowAppContext", { enumerable: true, get: function () { return xflow_main_1.XFlowAppContext; } });
Object.defineProperty(exports, "useXFlowApp", { enumerable: true, get: function () { return xflow_main_1.useXFlowApp; } });
// config provider context：使用全局Config
Object.defineProperty(exports, "useXflowPrefixCls", { enumerable: true, get: function () { return xflow_main_1.useXflowPrefixCls; } });
Object.defineProperty(exports, "XFlowConfigProviderContext", { enumerable: true, get: function () { return xflow_main_1.XFlowConfigProviderContext; } });
// extension context: 注册扩展
Object.defineProperty(exports, "ExtensionRegistryContext", { enumerable: true, get: function () { return xflow_main_1.ExtensionRegistryContext; } });
Object.defineProperty(exports, "useExtensionRegistry", { enumerable: true, get: function () { return xflow_main_1.useExtensionRegistry; } });
/** widget：extension  */
var xflow_main_2 = require("./xflow-main");
/** graphProvider：注入Graph时 需要 */
var xflow_main_3 = require("./xflow-main");
Object.defineProperty(exports, "IGraphProvider", { enumerable: true, get: function () { return xflow_main_3.IGraphProvider; } });
Object.defineProperty(exports, "createGraphConfig", { enumerable: true, get: function () { return xflow_main_3.createGraphConfig; } });
Object.defineProperty(exports, "GraphConfig", { enumerable: true, get: function () { return xflow_main_3.GraphConfig; } });
/** app：用于extension扩展*/
var xflow_main_4 = require("./xflow-main");
Object.defineProperty(exports, "IApplication", { enumerable: true, get: function () { return xflow_main_4.IApplication; } });
Object.defineProperty(exports, "IApplicationContribution", { enumerable: true, get: function () { return xflow_main_4.IApplicationContribution; } });
/** Command Service */
var command_1 = require("./command");
Object.defineProperty(exports, "ICommandHandler", { enumerable: true, get: function () { return command_1.ICommandHandler; } });
Object.defineProperty(exports, "IGraphCommandService", { enumerable: true, get: function () { return command_1.IGraphCommandService; } });
Object.defineProperty(exports, "ICommandContextProvider", { enumerable: true, get: function () { return command_1.ICommandContextProvider; } });
Object.defineProperty(exports, "IGraphCommandContribution", { enumerable: true, get: function () { return command_1.IGraphCommandContribution; } });
Object.defineProperty(exports, "IGraphCommandFactory", { enumerable: true, get: function () { return command_1.IGraphCommandFactory; } });
Object.defineProperty(exports, "GraphCommandRegistry", { enumerable: true, get: function () { return command_1.GraphCommandRegistry; } });
Object.defineProperty(exports, "commandRegistryModule", { enumerable: true, get: function () { return command_1.commandRegistryModule; } });
/** Command 常量 */
var command_contributions_1 = require("./command-contributions");
Object.defineProperty(exports, "XFlowNodeCommands", { enumerable: true, get: function () { return command_contributions_1.XFlowNodeCommands; } });
Object.defineProperty(exports, "XFlowEdgeCommands", { enumerable: true, get: function () { return command_contributions_1.XFlowEdgeCommands; } });
Object.defineProperty(exports, "XFlowGroupCommands", { enumerable: true, get: function () { return command_contributions_1.XFlowGroupCommands; } });
Object.defineProperty(exports, "XFlowGraphCommands", { enumerable: true, get: function () { return command_contributions_1.XFlowGraphCommands; } });
Object.defineProperty(exports, "XFlowModelCommands", { enumerable: true, get: function () { return command_contributions_1.XFlowModelCommands; } });
/** 创建 Command hook config */
Object.defineProperty(exports, "createCmdConfig", { enumerable: true, get: function () { return command_contributions_1.createCmdConfig; } });
/** React Node Context */
var command_contributions_2 = require("./command-contributions");
Object.defineProperty(exports, "AppContext", { enumerable: true, get: function () { return command_contributions_2.AppContext; } });
Object.defineProperty(exports, "useAppContext", { enumerable: true, get: function () { return command_contributions_2.useAppContext; } });
Object.defineProperty(exports, "getNodeReactComponent", { enumerable: true, get: function () { return command_contributions_2.getNodeReactComponent; } });
/*******************************************************
 *  XFlow Hooks： 钩子函数
 *****************************************************/
var hooks_1 = require("./hooks");
// 扩展Hooks
Object.defineProperty(exports, "IHookService", { enumerable: true, get: function () { return hooks_1.IHookService; } });
Object.defineProperty(exports, "IHookContribution", { enumerable: true, get: function () { return hooks_1.IHookContribution; } });
// 创建 React config hook
Object.defineProperty(exports, "createHookConfig", { enumerable: true, get: function () { return hooks_1.createHookConfig; } });
/*******************************************************
 *  Model Service：全局状态
 *****************************************************/
var model_service_1 = require("./model-service");
Object.defineProperty(exports, "MODELS", { enumerable: true, get: function () { return model_service_1.MODELS; } });
Object.defineProperty(exports, "IModelService", { enumerable: true, get: function () { return model_service_1.IModelService; } });
Object.defineProperty(exports, "IModelContribution", { enumerable: true, get: function () { return model_service_1.IModelContribution; } });
Object.defineProperty(exports, "createModelServiceConfig", { enumerable: true, get: function () { return model_service_1.createModelServiceConfig; } });
var rx_model_1 = require("./common/rx-model");
Object.defineProperty(exports, "RxModel", { enumerable: true, get: function () { return rx_model_1.RxModel; } });
Object.defineProperty(exports, "NsModel", { enumerable: true, get: function () { return rx_model_1.NsModel; } });
var rx_model_hook_1 = require("./common/rx-model-hook");
Object.defineProperty(exports, "useModel", { enumerable: true, get: function () { return rx_model_hook_1.useModel; } });
Object.defineProperty(exports, "createComponentModel", { enumerable: true, get: function () { return rx_model_hook_1.createComponentModel; } });
Object.defineProperty(exports, "useModelAsync", { enumerable: true, get: function () { return rx_model_hook_1.useModelAsync; } });
Object.defineProperty(exports, "useIsMountedRef", { enumerable: true, get: function () { return rx_model_hook_1.useIsMountedRef; } });
/*******************************************************
 *  Toolbar：工具栏
 *****************************************************/
/** Toolbar 配置 */
var toolbar_1 = require("./toolbar");
// component
Object.defineProperty(exports, "ToolbarRegistry", { enumerable: true, get: function () { return toolbar_1.ToolbarRegistry; } });
// ioc 扩展
Object.defineProperty(exports, "IToolbarService", { enumerable: true, get: function () { return toolbar_1.IToolbarService; } });
Object.defineProperty(exports, "IToolbarContribution", { enumerable: true, get: function () { return toolbar_1.IToolbarContribution; } });
/*******************************************************
 *  Menu：菜单
 *****************************************************/
/** Menu 配置 */
var menu_1 = require("./menu");
// 扩展 Service
Object.defineProperty(exports, "IMenuService", { enumerable: true, get: function () { return menu_1.IMenuService; } });
Object.defineProperty(exports, "IMenuContribution", { enumerable: true, get: function () { return menu_1.IMenuContribution; } });
// menu item
Object.defineProperty(exports, "MenuItemType", { enumerable: true, get: function () { return menu_1.MenuItemType; } });
// Component
Object.defineProperty(exports, "MenuRegistry", { enumerable: true, get: function () { return menu_1.MenuRegistry; } });
/*******************************************************
 *  KeyBindings：快捷键
 *****************************************************/
/** KeyBindings 配置 */
var keybinding_1 = require("./keybinding");
Object.defineProperty(exports, "KeyBindings", { enumerable: true, get: function () { return keybinding_1.KeyBindings; } });
Object.defineProperty(exports, "createKeybindingConfig", { enumerable: true, get: function () { return keybinding_1.createKeybindingConfig; } });
Object.defineProperty(exports, "KeybindingConfig", { enumerable: true, get: function () { return keybinding_1.KeybindingConfig; } });
Object.defineProperty(exports, "IKeyBindingContribution", { enumerable: true, get: function () { return keybinding_1.IKeyBindingContribution; } });
/*******************************************************
 *  UTILS：工具方法
 *****************************************************/
/** utils：dispose */
var disposable_1 = require("./common/disposable");
Object.defineProperty(exports, "Disposable", { enumerable: true, get: function () { return disposable_1.Disposable; } });
Object.defineProperty(exports, "DisposableCollection", { enumerable: true, get: function () { return disposable_1.DisposableCollection; } });
/** utils：defer 延迟 */
var deferred_1 = require("./common/deferred");
Object.defineProperty(exports, "Deferred", { enumerable: true, get: function () { return deferred_1.Deferred; } });
/** utils：延迟 */
var delay_1 = require("./common/delay");
Object.defineProperty(exports, "delay", { enumerable: true, get: function () { return delay_1.delay; } });
/** utils：uuid */
var uuid_1 = require("./common/uuid");
Object.defineProperty(exports, "uuidv4", { enumerable: true, get: function () { return uuid_1.uuidv4; } });
/** utils：绝对定位 */
var position_1 = require("./common/position");
Object.defineProperty(exports, "usePositionStyle", { enumerable: true, get: function () { return position_1.usePositionStyle; } });
var types_1 = require("./common/types");
/** utils：insertCss */
var utils_1 = require("./common/utils");
Object.defineProperty(exports, "insertCss", { enumerable: true, get: function () { return utils_1.insertCss; } });
Object.defineProperty(exports, "isReactComponent", { enumerable: true, get: function () { return utils_1.isReactComponent; } });
/*******************************************************
 * Icon：Antd Icon
 *****************************************************/
/** ICON */
var antd_icons_1 = require("./antd-icons");
Object.defineProperty(exports, "IconStore", { enumerable: true, get: function () { return antd_icons_1.IconStore; } });
/** 全局常量 */
exports.XFlowConstants = tslib_1.__importStar(require("./constants"));
