// @antv/l7-source/lib/utils/tile/getRasterTile.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/utils/tile/getRasterTile.ts
var getRasterTile_exports = {};
__export(getRasterTile_exports, {
  defaultFormat: () => defaultFormat,
  getTileBuffer: () => getTileBuffer,
  getTileImage: () => getTileImage
});
module.exports = __toCommonJS(getRasterTile_exports);
var import_l7_utils = require("@antv/l7-utils");
var import_getRasterData = require("./getRasterData");
var import_request = require("./request");
var getTileBuffer = async (url, tileParams, tile, rasterFormat, operation) => {
  const requestParameters = {
    // getTileUrl 将原始的 url 路径进行转化（多服务器）
    url: (0, import_request.getTileUrl)(url, tileParams)
  };
  return new Promise((resolve, reject) => {
    (0, import_getRasterData.getRasterFile)(
      tile,
      requestParameters,
      (err, img) => {
        if (err) {
          reject(err);
        } else if (img) {
          resolve(img);
        }
      },
      rasterFormat,
      operation
    );
  });
};
var getTileImage = async (url, tileParams, tile, cfg) => {
  let imageUrl;
  const templateUrl = Array.isArray(url) ? url[0] : url;
  if (cfg.wmtsOptions) {
    const getWMTSURLFromTemplateNew = (cfg == null ? void 0 : cfg.getURLFromTemplate) || import_l7_utils.getWMTSURLFromTemplate;
    imageUrl = getWMTSURLFromTemplateNew(templateUrl, {
      ...tileParams,
      ...cfg.wmtsOptions
    });
  } else {
    const getURLFromTemplateNew = (cfg == null ? void 0 : cfg.getURLFromTemplate) || import_l7_utils.getURLFromTemplate;
    imageUrl = getURLFromTemplateNew(templateUrl, tileParams);
  }
  return new Promise((resolve, reject) => {
    var _a;
    const xhr = (0, import_l7_utils.getImage)(
      {
        url: imageUrl,
        type: ((_a = cfg == null ? void 0 : cfg.requestParameters) == null ? void 0 : _a.type) || "arrayBuffer"
      },
      (err, img) => {
        if (err) {
          reject(err);
        } else if (img) {
          resolve(img);
        }
      },
      cfg.transformResponse
    );
    tile.xhrCancel = () => xhr.abort();
  });
};
var defaultFormat = () => {
  return {
    rasterData: new Uint8Array([0]),
    width: 1,
    height: 1
  };
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  defaultFormat,
  getTileBuffer,
  getTileImage
});
