// antd-token-previewer/lib/component-demos/progress/info.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _antd = require("antd");

var _react = _interopRequireDefault(require("react"));

var Demo = function Demo() {
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_antd.Progress, {
    percent: 30
  }), /*#__PURE__*/_react.default.createElement(_antd.Progress, {
    percent: 50,
    status: "active"
  }), /*#__PURE__*/_react.default.createElement(_antd.Progress, {
    percent: 70,
    type: 'dashboard'
  }), /*#__PURE__*/_react.default.createElement(_antd.Progress, {
    percent: 80,
    type: 'circle'
  }), /*#__PURE__*/_react.default.createElement(_antd.Progress, {
    steps: 8,
    percent: 30
  }));
};

var componentDemo = {
  demo: /*#__PURE__*/_react.default.createElement(Demo, null),
  tokens: ['colorInfo'],
  key: 'info'
};
var _default = componentDemo;
exports.default = _default;
