// antd-token-previewer/lib/component-demos/progress/progress.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _antd = require("antd");

var _react = _interopRequireDefault(require("react"));

var Demo = function Demo() {
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_antd.Progress, {
    percent: 30
  }), /*#__PURE__*/_react.default.createElement(_antd.Progress, {
    percent: 50,
    status: "active"
  }), /*#__PURE__*/_react.default.createElement(_antd.Progress, {
    percent: 70,
    status: "exception"
  }), /*#__PURE__*/_react.default.createElement(_antd.Progress, {
    percent: 100
  }), /*#__PURE__*/_react.default.createElement(_antd.Progress, {
    percent: 50,
    showInfo: false
  }), /*#__PURE__*/_react.default.createElement(_antd.Progress, {
    steps: 8
  }));
};

var componentDemo = {
  demo: /*#__PURE__*/_react.default.createElement(Demo, null),
  tokens: ['colorFillSecondary', 'colorText', 'colorBgContainer'],
  key: 'default'
};
var _default = componentDemo;
exports.default = _default;
