// antd-token-previewer/es/component-demos/card/card.js
import React from 'react';
import { Card, Space } from 'antd';

var Demo = function Demo() {
  return /*#__PURE__*/React.createElement(Space, null, /*#__PURE__*/React.createElement(Card, {
    title: "Default size card",
    extra: /*#__PURE__*/React.createElement("a", {
      href: "#"
    }, "More"),
    style: {
      width: 300
    }
  }, /*#__PURE__*/React.createElement("p", null, "Card content"), " ", /*#__PURE__*/React.createElement("p", null, "Card content"), " ", /*#__PURE__*/React.createElement("p", null, "Card content")), /*#__PURE__*/React.createElement(Card, {
    loading: true,
    size: "small",
    title: "Small size card",
    extra: /*#__PURE__*/React.createElement("a", {
      href: "#"
    }, "More"),
    style: {
      width: 300
    }
  }, /*#__PURE__*/React.createElement("p", null, "Card content"), " ", /*#__PURE__*/React.createElement("p", null, "Card content"), " ", /*#__PURE__*/React.createElement("p", null, "Card content")));
};

var componentDemo = {
  demo: /*#__PURE__*/React.createElement(Demo, null),
  tokens: ['colorText', 'colorTextHeading', 'colorTextSecondary', 'colorBgContainer', 'colorBorderSecondary', 'colorPrimary', 'colorBgContainer'],
  key: 'card'
};
export default componentDemo;
