// antd-token-previewer/es/component-demos/card/cardGrid.js
import React from 'react';
import { Card } from 'antd';
var gridStyle = {
  width: '25%',
  textAlign: 'center'
};

var Demo = function Demo() {
  return /*#__PURE__*/React.createElement(Card, {
    title: "Card Title"
  }, /*#__PURE__*/React.createElement(Card.Grid, {
    style: gridStyle
  }, "Content"), /*#__PURE__*/React.createElement(Card.Grid, {
    hoverable: false,
    style: gridStyle
  }, "Content"), /*#__PURE__*/React.createElement(Card.Grid, {
    style: gridStyle
  }, "Content"), /*#__PURE__*/React.createElement(Card.Grid, {
    style: gridStyle
  }, "Content"), /*#__PURE__*/React.createElement(Card.Grid, {
    style: gridStyle
  }, "Content"), /*#__PURE__*/React.createElement(Card.Grid, {
    style: gridStyle
  }, "Content"), /*#__PURE__*/React.createElement(Card.Grid, {
    style: gridStyle
  }, "Content"));
};

var componentDemo = {
  demo: /*#__PURE__*/React.createElement(Demo, null),
  tokens: ['colorBorderSecondary'],
  key: 'cardGrid'
};
export default componentDemo;
