// antd-token-previewer/lib/overviews/Error.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Error = void 0;

var _antd = require("antd");

var _react = _interopRequireDefault(require("react"));

var _layoutKit = require("@arvinxu/layout-kit");

var _error = _interopRequireDefault(require("../component-demos/alert/error"));

var _dangerButton = _interopRequireDefault(require("../component-demos/button/dangerButton"));

var _error2 = _interopRequireDefault(require("../component-demos/message/error"));

var _danger = _interopRequireDefault(require("../component-demos/progress/danger"));

var _error3 = _interopRequireDefault(require("../component-demos/tag/error"));

var _badge = _interopRequireDefault(require("../component-demos/badge/badge"));

var _menuDanger = _interopRequireDefault(require("../component-demos/menu/menuDanger"));

var _danger2 = _interopRequireDefault(require("../component-demos/upload/danger"));

var _dropdownError = _interopRequireDefault(require("../component-demos/dropdown/dropdownError"));

var _error4 = _interopRequireDefault(require("../component-demos/notification/error"));

var _danger3 = _interopRequireDefault(require("../component-demos/timeline/danger"));

var Error = function Error() {
  return /*#__PURE__*/_react.default.createElement(_antd.Card, {
    size: 'small'
  }, /*#__PURE__*/_react.default.createElement(_layoutKit.Flexbox, {
    horizontal: true,
    align: 'start',
    gap: 24
  }, /*#__PURE__*/_react.default.createElement(_layoutKit.Flexbox, {
    gap: 24
  }, /*#__PURE__*/_react.default.createElement(_layoutKit.Flexbox, {
    horizontal: true,
    align: 'center',
    gap: 12,
    style: {
      marginTop: 8
    }
  }, _dangerButton.default.demo, /*#__PURE__*/_react.default.createElement("div", null, _error3.default.demo), _badge.default.demo), _error.default.demo), /*#__PURE__*/_react.default.createElement(_layoutKit.Flexbox, {
    align: 'center',
    gap: 28
  }, _error2.default.demo, _danger.default.demo)), /*#__PURE__*/_react.default.createElement(_layoutKit.Flexbox, {
    horizontal: true,
    gap: 40,
    style: {
      marginTop: 32
    }
  }, /*#__PURE__*/_react.default.createElement("div", null, _error4.default.demo), /*#__PURE__*/_react.default.createElement("div", null, _danger3.default.demo)), /*#__PURE__*/_react.default.createElement(_layoutKit.Flexbox, {
    horizontal: true,
    gap: 40
  }, _menuDanger.default.demo, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      width: 300
    }
  }, _danger2.default.demo), _dropdownError.default.demo));
};

exports.Error = Error;
