// antd-token-previewer/lib/overviews/Primary.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Primary = void 0;

var _antd = require("antd");

var _react = _interopRequireDefault(require("react"));

var _menu = _interopRequireDefault(require("../component-demos/menu/menu"));

var _selectTag = _interopRequireDefault(require("../component-demos/select/selectTag"));

var _buttonIcon = _interopRequireDefault(require("../component-demos/button/button-icon"));

var _switch = _interopRequireDefault(require("../component-demos/switch/switch"));

var _radio = _interopRequireDefault(require("../component-demos/radio/radio"));

var _button = _interopRequireDefault(require("../component-demos/radio/button"));

var _checkbox = _interopRequireDefault(require("../component-demos/checkbox/checkbox"));

var _tabs = _interopRequireDefault(require("../component-demos/tabs/tabs"));

var _outline = _interopRequireDefault(require("../component-demos/pagination/outline"));

var _steps = _interopRequireDefault(require("../component-demos/steps/steps"));

var _popconfirm = _interopRequireDefault(require("../component-demos/popconfirm/popconfirm"));

var _timeline = _interopRequireDefault(require("../component-demos/timeline/timeline"));

var _table = _interopRequireDefault(require("../component-demos/table/table"));

var Primary = function Primary() {
  return /*#__PURE__*/_react.default.createElement(_antd.Card, {
    size: 'small'
  }, /*#__PURE__*/_react.default.createElement(_antd.Space, {
    direction: 'vertical'
  }, /*#__PURE__*/_react.default.createElement(_antd.Space, {
    align: 'start',
    size: 'large'
  }, _menu.default.demo, /*#__PURE__*/_react.default.createElement(_antd.Space, {
    direction: 'vertical',
    size: 'large'
  }, /*#__PURE__*/_react.default.createElement(_antd.Space, {
    size: 'large',
    align: 'start'
  }, /*#__PURE__*/_react.default.createElement(_antd.Space, {
    direction: 'vertical',
    size: 'large'
  }, /*#__PURE__*/_react.default.createElement("div", null, _buttonIcon.default.demo), /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("span", null, _radio.default.demo), _checkbox.default.demo, _switch.default.demo), /*#__PURE__*/_react.default.createElement("div", null, _button.default.demo), _tabs.default.demo), _selectTag.default.demo), _outline.default.demo, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      padding: 12
    }
  }, _steps.default.demo), /*#__PURE__*/_react.default.createElement(_antd.Space, {
    size: 'large',
    align: 'start'
  }, _popconfirm.default.demo, _timeline.default.demo))), _table.default.demo));
};

exports.Primary = Primary;
