// antd-token-previewer/lib/overviews/Success.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Success = void 0;

var _antd = require("antd");

var _react = _interopRequireDefault(require("react"));

var _success = _interopRequireDefault(require("../component-demos/alert/success"));

var _success2 = _interopRequireDefault(require("../component-demos/message/success"));

var _success3 = _interopRequireDefault(require("../component-demos/progress/success"));

var _success4 = _interopRequireDefault(require("../component-demos/tag/success"));

var _success5 = _interopRequireDefault(require("../component-demos/input/success"));

var _success6 = _interopRequireDefault(require("../component-demos/result/success"));

var _success7 = _interopRequireDefault(require("../component-demos/notification/success"));

var _success8 = _interopRequireDefault(require("../component-demos/timeline/success"));

var _layoutKit = require("@arvinxu/layout-kit");

var Success = function Success() {
  return /*#__PURE__*/_react.default.createElement(_antd.Card, {
    size: 'small'
  }, /*#__PURE__*/_react.default.createElement(_layoutKit.Flexbox, {
    horizontal: true,
    align: 'start',
    gap: 24
  }, /*#__PURE__*/_react.default.createElement(_layoutKit.Flexbox, {
    gap: 40
  }, /*#__PURE__*/_react.default.createElement(_layoutKit.Flexbox, {
    horizontal: true,
    align: 'center',
    gap: 12
  }, /*#__PURE__*/_react.default.createElement("div", null, _success4.default.demo), _success5.default.demo), _success.default.demo), /*#__PURE__*/_react.default.createElement(_layoutKit.Flexbox, {
    align: 'center',
    gap: 28
  }, _success2.default.demo, _success3.default.demo)), /*#__PURE__*/_react.default.createElement(_layoutKit.Flexbox, {
    horizontal: true,
    gap: 40,
    style: {
      marginTop: 32
    }
  }, /*#__PURE__*/_react.default.createElement("div", null, _success7.default.demo), /*#__PURE__*/_react.default.createElement("div", null, _success8.default.demo)), _success6.default.demo);
};

exports.Success = Success;
