// antd-token-previewer/lib/overviews/Warning.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Warning = void 0;

var _antd = require("antd");

var _react = _interopRequireDefault(require("react"));

var _warning = _interopRequireDefault(require("../component-demos/alert/warning"));

var _warning2 = _interopRequireDefault(require("../component-demos/message/warning"));

var _popconfirm = _interopRequireDefault(require("../component-demos/popconfirm/popconfirm"));

var _warning3 = _interopRequireDefault(require("../component-demos/modal/warning"));

var _warning4 = _interopRequireDefault(require("../component-demos/badge/warning"));

var _warningText = _interopRequireDefault(require("../component-demos/typography/warningText"));

var _warningTitle = _interopRequireDefault(require("../component-demos/typography/warningTitle"));

var _warning5 = _interopRequireDefault(require("../component-demos/tag/warning"));

var _warning6 = _interopRequireDefault(require("../component-demos/input/warning"));

var _warning7 = _interopRequireDefault(require("../component-demos/result/warning"));

var _warning8 = _interopRequireDefault(require("../component-demos/notification/warning"));

var _layoutKit = require("@arvinxu/layout-kit");

var Warning = function Warning() {
  return /*#__PURE__*/_react.default.createElement(_antd.Card, {
    size: 'small'
  }, /*#__PURE__*/_react.default.createElement(_layoutKit.Flexbox, {
    horizontal: true,
    align: 'start',
    gap: 24
  }, /*#__PURE__*/_react.default.createElement(_layoutKit.Flexbox, {
    gap: 24
  }, /*#__PURE__*/_react.default.createElement(_layoutKit.Flexbox, {
    horizontal: true,
    gap: 12
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      width: 200
    }
  }, _warningTitle.default.demo), /*#__PURE__*/_react.default.createElement("div", {
    style: {
      width: '100%'
    }
  }, _warning6.default.demo)), _warning.default.demo), /*#__PURE__*/_react.default.createElement(_layoutKit.Flexbox, {
    align: 'center',
    gap: 28
  }, _warning2.default.demo, _popconfirm.default.demo, /*#__PURE__*/_react.default.createElement(_layoutKit.Flexbox, {
    horizontal: true,
    gap: 16
  }, _warning4.default.demo, _warning5.default.demo, _warningText.default.demo))), /*#__PURE__*/_react.default.createElement(_layoutKit.Flexbox, {
    horizontal: true,
    gap: 24,
    style: {
      marginTop: 32
    }
  }, /*#__PURE__*/_react.default.createElement("div", null, _warning8.default.demo), /*#__PURE__*/_react.default.createElement("div", null, _warning3.default.demo)), _warning7.default.demo);
};

exports.Warning = Warning;
