// antd-token-previewer/es/component-demos/anchor/anchorInLayout.js
import React from 'react';
import { Anchor, theme } from 'antd';
var Link = Anchor.Link;

var Demo = function Demo() {
  var _theme$useToken = theme.useToken(),
      token = _theme$useToken.token;

  return /*#__PURE__*/React.createElement("div", {
    style: {
      background: token.colorBorderSecondary,
      padding: 12
    }
  }, /*#__PURE__*/React.createElement(Anchor, null, /*#__PURE__*/React.createElement(Link, {
    href: "#components-anchor-demo-basic",
    title: "Basic demo"
  }), /*#__PURE__*/React.createElement(Link, {
    href: "#components-anchor-demo-static",
    title: "Static demo"
  }), /*#__PURE__*/React.createElement(Link, {
    href: "#API",
    title: "API"
  }, /*#__PURE__*/React.createElement(Link, {
    href: "#Anchor-Props",
    title: "Anchor Props"
  }), /*#__PURE__*/React.createElement(Link, {
    href: "#Link-Props",
    title: "Link Props"
  }))));
};

var componentDemo = {
  demo: /*#__PURE__*/React.createElement(Demo, null),
  tokens: ['colorSplit'],
  key: 'anchorInLayout'
};
export default componentDemo;
