// antd-token-previewer/lib/component-demos/menu/menuInLayout.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _antd = require("antd");

var _react = _interopRequireDefault(require("react"));

var _data = _interopRequireDefault(require("./data"));

var Demo = function Demo() {
  var _theme$useToken = _antd.theme.useToken(),
      token = _theme$useToken.token;

  return /*#__PURE__*/_react.default.createElement("div", {
    style: {
      background: token.colorBorderSecondary,
      padding: 12
    }
  }, /*#__PURE__*/_react.default.createElement(_antd.Menu, {
    style: {
      width: 256
    },
    defaultSelectedKeys: ['1'],
    defaultOpenKeys: ['sub1'],
    mode: "inline",
    items: _data.default
  }));
};

var componentDemo = {
  demo: /*#__PURE__*/_react.default.createElement(Demo, null),
  tokens: ['colorSplit'],
  key: 'menuInLayout'
};
var _default = componentDemo;
exports.default = _default;
