// antd-token-previewer/lib/component-demos/menu/menu.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _antd = require("antd");

var _react = _interopRequireDefault(require("react"));

var _data = _interopRequireDefault(require("./data"));

var Demo = function Demo() {
  var onClick = function onClick(e) {
    console.log('click ', e);
  };

  return /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(_antd.Menu, {
    onClick: onClick,
    style: {
      width: 256
    },
    defaultSelectedKeys: ['1'],
    defaultOpenKeys: ['sub1', 'sub2'],
    mode: "inline",
    items: _data.default
  }));
};

var componentDemo = {
  demo: /*#__PURE__*/_react.default.createElement(Demo, null),
  tokens: ['colorPrimary', 'colorBgContainer', 'colorFillAlter', 'colorSplit', 'colorPrimaryHover'],
  key: 'default'
};
var _default = componentDemo;
exports.default = _default;
