// antd-token-previewer/es/component-demos/badge/badge.js
import React from 'react';
import { Badge, Avatar, Space, theme } from 'antd';
import { ClockCircleFilled } from '@ant-design/icons';

var Demo = function Demo() {
  var _theme$useToken = theme.useToken(),
      token = _theme$useToken.token;

  return /*#__PURE__*/React.createElement(Space, {
    size: "large"
  }, /*#__PURE__*/React.createElement(Badge, {
    count: 5
  }, /*#__PURE__*/React.createElement(Avatar, {
    shape: "square",
    size: "large"
  })), /*#__PURE__*/React.createElement(Badge, {
    count: 0,
    showZero: true
  }, /*#__PURE__*/React.createElement(Avatar, {
    shape: "square",
    size: "large"
  })), /*#__PURE__*/React.createElement(Badge, {
    count: /*#__PURE__*/React.createElement(ClockCircleFilled, {
      style: {
        color: token.colorError
      }
    })
  }, /*#__PURE__*/React.createElement(Avatar, {
    shape: "square",
    size: "large"
  })));
};

var componentDemo = {
  demo: /*#__PURE__*/React.createElement(Demo, null),
  tokens: ['colorError', 'colorBorderBg', 'colorBgContainer'],
  key: 'badge'
};
export default componentDemo;
