// @antv/l7-layers/es/earth/models/base.js
import _asyncToGenerator from "@babel/runtime/helpers/esm/asyncToGenerator";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/esm/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _regeneratorRuntime from "@babel/runtime/regenerator";
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }
function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
import { AttributeType, gl } from '@antv/l7-core';
import BaseModel from "../../core/BaseModel";
import { earthTriangulation } from "../../core/triangulation";
/* babel-plugin-inline-import '../shaders/base_frag.glsl' */
var baseFrag = "\nuniform sampler2D u_texture;\n\nvarying vec2 v_texCoord;\nvarying float v_lightWeight;\n\n\nvoid main() {\n\n    vec4 color = texture2D(u_texture,vec2(v_texCoord.x,v_texCoord.y));\n    color.xyz = color.xyz * v_lightWeight;\n    gl_FragColor = color;\n}\n";
/* babel-plugin-inline-import '../shaders/base_vert.glsl' */
var baseVert = "// attribute vec4 a_Color;\nattribute vec3 a_Position;\nattribute vec3 a_Normal;\nattribute vec2 a_Uv;\nvarying vec2 v_texCoord;\n\n// attribute vec2 a_Extrude;\n// attribute float a_Size;\n// attribute float a_Shape;\n\nuniform vec3 u_CameraPosition;\nuniform mat4 u_ViewProjectionMatrix;\nuniform mat4 u_ModelMatrix;\nuniform float u_ambientRatio : 0.5;\nuniform float u_diffuseRatio : 0.3;\nuniform float u_specularRatio : 0.2;\nuniform vec3 u_sunLight: [1.0, -10.5, 12.0];\n\n\n\nfloat calc_lighting(vec4 pos) {\n\n    vec3 worldPos = vec3(pos * u_ModelMatrix);\n\n    vec3 worldNormal = a_Normal;\n\n    // cal light weight\n    vec3 viewDir = normalize(u_CameraPosition - worldPos);\n\n    vec3 lightDir = normalize(u_sunLight);\n\n    vec3 halfDir = normalize(viewDir+lightDir);\n    // lambert\n    float lambert = dot(worldNormal, lightDir);\n    // specular\n    float specular = pow(max(0.0, dot(worldNormal, halfDir)), 32.0);\n    //sum to light weight\n    float lightWeight = u_ambientRatio + u_diffuseRatio * lambert + u_specularRatio * specular;\n\n    return lightWeight;\n}\n\nvarying float v_lightWeight;\nvoid main() {\n\n    v_texCoord = a_Uv;\n\n    float lightWeight = calc_lighting(vec4(a_Position, 1.0));\n    v_lightWeight = lightWeight;\n\n    gl_Position = u_ViewProjectionMatrix * u_ModelMatrix * vec4(a_Position, 1.0);\n}\n";
var BaseEarthModel = /*#__PURE__*/function (_BaseModel) {
  _inherits(BaseEarthModel, _BaseModel);
  var _super = _createSuper(BaseEarthModel);
  function BaseEarthModel() {
    var _this;
    _classCallCheck(this, BaseEarthModel);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _super.call.apply(_super, [this].concat(args));
    _defineProperty(_assertThisInitialized(_this), "earthTime", 3.4);
    _defineProperty(_assertThisInitialized(_this), "sunX", 1000);
    _defineProperty(_assertThisInitialized(_this), "sunY", 1000);
    _defineProperty(_assertThisInitialized(_this), "sunZ", 1000);
    _defineProperty(_assertThisInitialized(_this), "sunRadius", Math.sqrt(_this.sunX * _this.sunX + _this.sunY * _this.sunY + _this.sunZ * _this.sunZ));
    return _this;
  }
  _createClass(BaseEarthModel, [{
    key: "getUninforms",
    value: function getUninforms() {
      var _this$layer$getLayerC = this.layer.getLayerConfig(),
        animateOption = _this$layer$getLayerC.animateOption,
        globalOptions = _this$layer$getLayerC.globalOptions;
      if (animateOption !== null && animateOption !== void 0 && animateOption.enable) {
        // @ts-ignore
        // T: rotateY 方法只有在地球模式下存在
        this.mapService.rotateY({
          reg: 0.002
        });
        this.earthTime += 0.02;
        this.sunY = 10;
        this.sunX = Math.cos(this.earthTime) * (this.sunRadius - this.sunY);
        this.sunZ = Math.sin(this.earthTime) * (this.sunRadius - this.sunY);
      }
      return {
        u_ambientRatio: (globalOptions === null || globalOptions === void 0 ? void 0 : globalOptions.ambientRatio) || 0.6,
        // 环境光
        u_diffuseRatio: (globalOptions === null || globalOptions === void 0 ? void 0 : globalOptions.diffuseRatio) || 0.4,
        // 漫反射
        u_specularRatio: (globalOptions === null || globalOptions === void 0 ? void 0 : globalOptions.specularRatio) || 0.1,
        // 高光反射
        u_sunLight: [this.sunX, this.sunY, this.sunZ],
        u_texture: this.texture
      };
    }
  }, {
    key: "setEarthTime",
    value: function setEarthTime(time) {
      this.earthTime = time;
      this.sunY = 10;
      this.sunX = Math.cos(this.earthTime) * (this.sunRadius - this.sunY);
      this.sunZ = Math.sin(this.earthTime) * (this.sunRadius - this.sunY);
      this.layerService.throttleRenderLayers();
    }
  }, {
    key: "initModels",
    value: function () {
      var _initModels = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee() {
        var _this2 = this;
        var _this$layer$getLayerC2, globalOptions, source, createTexture2D;
        return _regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              _this$layer$getLayerC2 = this.layer.getLayerConfig(), globalOptions = _this$layer$getLayerC2.globalOptions;
              if ((globalOptions === null || globalOptions === void 0 ? void 0 : globalOptions.earthTime) !== undefined) {
                this.setEarthTime(globalOptions.earthTime);
              }
              source = this.layer.getSource();
              createTexture2D = this.rendererService.createTexture2D;
              this.texture = createTexture2D({
                height: 0,
                width: 0
              });
              source.data.images.then(function (imageData) {
                _this2.texture = createTexture2D({
                  data: imageData[0],
                  width: imageData[0].width,
                  height: imageData[0].height
                });
                _this2.layerService.reRender();
              });
              return _context.abrupt("return", this.buildModels());
            case 7:
            case "end":
              return _context.stop();
          }
        }, _callee, this);
      }));
      function initModels() {
        return _initModels.apply(this, arguments);
      }
      return initModels;
    }()
  }, {
    key: "clearModels",
    value: function clearModels() {
      return '';
    }
  }, {
    key: "buildModels",
    value: function () {
      var _buildModels = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee2() {
        var model;
        return _regeneratorRuntime.wrap(function _callee2$(_context2) {
          while (1) switch (_context2.prev = _context2.next) {
            case 0:
              // Tip: 调整图层的绘制顺序 地球大气层
              this.layer.zIndex = -998;
              _context2.next = 3;
              return this.layer.buildLayerModel({
                moduleName: 'earthBase',
                vertexShader: baseVert,
                fragmentShader: baseFrag,
                triangulation: earthTriangulation,
                depth: {
                  enable: true
                },
                blend: this.getBlend()
              });
            case 3:
              model = _context2.sent;
              return _context2.abrupt("return", [model]);
            case 5:
            case "end":
              return _context2.stop();
          }
        }, _callee2, this);
      }));
      function buildModels() {
        return _buildModels.apply(this, arguments);
      }
      return buildModels;
    }()
  }, {
    key: "registerBuiltinAttributes",
    value: function registerBuiltinAttributes() {
      this.styleAttributeService.registerStyleAttribute({
        name: 'size',
        type: AttributeType.Attribute,
        descriptor: {
          name: 'a_Size',
          buffer: {
            usage: gl.DYNAMIC_DRAW,
            data: [],
            type: gl.FLOAT
          },
          size: 1,
          update: function update(feature) {
            var _feature$size = feature.size,
              size = _feature$size === void 0 ? 1 : _feature$size;
            return Array.isArray(size) ? [size[0]] : [size];
          }
        }
      });
      this.styleAttributeService.registerStyleAttribute({
        name: 'normal',
        type: AttributeType.Attribute,
        descriptor: {
          name: 'a_Normal',
          buffer: {
            usage: gl.STATIC_DRAW,
            data: [],
            type: gl.FLOAT
          },
          size: 3,
          update: function update(feature, featureIdx, vertex, attributeIdx, normal) {
            return normal;
          }
        }
      });
      this.styleAttributeService.registerStyleAttribute({
        name: 'uv',
        type: AttributeType.Attribute,
        descriptor: {
          name: 'a_Uv',
          buffer: {
            // give the WebGL driver a hint that this buffer may change
            usage: gl.DYNAMIC_DRAW,
            data: [],
            type: gl.FLOAT
          },
          size: 2,
          update: function update(feature, featureIdx, vertex) {
            return [vertex[3], vertex[4]];
          }
        }
      });
    }
  }]);
  return BaseEarthModel;
}(BaseModel);
export { BaseEarthModel as default };
