// @antv/xflow-core/lib/xflow-main/components/extension-registry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createExtensionRegistry = exports.ExtensionRegistry = void 0;
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importDefault(require("react"));
var lodash_1 = require("lodash");
var disposable_1 = require("../../common/disposable");
var uuid_1 = require("../../common/uuid");
/** 保存组件上的配置 */
var ExtensionRegistry = /** @class */ (function () {
    function ExtensionRegistry() {
        var _this = this;
        this.extensions = [];
        this.containerClassNames = new Set();
        this.addCoreModule = function (extension) {
            // core module 添加在前面
            _this.extensions.unshift(extension);
            return disposable_1.Disposable.create(function () {
                _this.removeExtension(extension);
            });
        };
        this.addExtension = function (extension) {
            _this.extensions.push(extension);
            return disposable_1.Disposable.create(function () {
                _this.removeExtension(extension);
            });
        };
        this.getExtension = function (config_type) {
            return (0, lodash_1.find)(_this.extensions, function (extension) { return (0, lodash_1.get)(extension, 'config.CONFIG_TYPE') === config_type; });
        };
        this.removeExtension = function (extension) {
            var idx = _this.extensions.indexOf(extension);
            if (idx > -1) {
                _this.extensions.splice(idx, 1);
            }
        };
        this.getInstanceId = function () {
            return _this.instanceId;
        };
        this.addContainerClassNames = function (clz) {
            return _this.containerClassNames.add(clz);
        };
        this.getContainerClassNames = function () {
            return Array.from(_this.containerClassNames);
        };
        this.getAllExtensions = function () {
            return _this.extensions;
        };
        this.getAllExtensionConfigs = function () {
            return _this.extensions.map(function (i) { return i.config; });
        };
        this.instanceId = (0, uuid_1.uuidv4)();
    }
    ExtensionRegistry.prototype.has = function (name) {
        return this.extensions.some(function (ext) { return ext.config.CONFIG_TYPE === name; });
    };
    return ExtensionRegistry;
}());
exports.ExtensionRegistry = ExtensionRegistry;
/** 获取Xflow extension，用于收集React组件的配置 */
var createExtensionRegistry = function () {
    /* eslint-disable-next-line react-hooks/rules-of-hooks */
    var registry = react_1.default.useMemo(function () {
        return new ExtensionRegistry();
    }, []);
    return registry;
};
exports.createExtensionRegistry = createExtensionRegistry;
