// @antv/xflow-core/lib/xflow-main/components/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExtensionRegistry = exports.XFlowAppExtension = exports.XFlowCanvas = exports.useXflowPrefixCls = exports.useXFlowApp = exports.XFlow = void 0;
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importDefault(require("react"));
var classnames_1 = tslib_1.__importDefault(require("classnames"));
/** app */
var application_module_1 = require("../application-module");
var app_context_1 = require("./app-context");
Object.defineProperty(exports, "useXFlowApp", { enumerable: true, get: function () { return app_context_1.useXFlowApp; } });
/** app-extension */
var extension_context_1 = require("./extension-context");
var global_config_context_1 = require("./global-config-context");
Object.defineProperty(exports, "useXflowPrefixCls", { enumerable: true, get: function () { return global_config_context_1.useXflowPrefixCls; } });
var app_extension_module_1 = require("./app-extension-module");
Object.defineProperty(exports, "XFlowAppExtension", { enumerable: true, get: function () { return app_extension_module_1.XFlowAppExtensionModule; } });
var extension_registry_1 = require("./extension-registry");
Object.defineProperty(exports, "ExtensionRegistry", { enumerable: true, get: function () { return extension_registry_1.ExtensionRegistry; } });
/** graph */
var canvas_1 = require("./canvas");
Object.defineProperty(exports, "XFlowCanvas", { enumerable: true, get: function () { return canvas_1.XFlowCanvas; } });
/** command */
var model_service_1 = require("../../model-service");
var command_contributions_1 = require("../../command-contributions");
/** hook */
var hooks_1 = require("../../hooks");
/** UI model */
var toolbar_1 = require("../../toolbar");
var menu_1 = require("../../menu");
var XFlow = function (props) {
    var meta = props.meta, graphConfig = props.graphConfig, graphData = props.graphData, graphLayout = props.graphLayout, onLoad = props.onLoad, isAutoCenter = props.isAutoCenter, hookConfig = props.hookConfig, modelServiceConfig = props.modelServiceConfig, commandConfig = props.commandConfig, onAppConfigReady = props.onAppConfigReady, onAppDestroy = props.onAppDestroy, _a = props.children, children = _a === void 0 ? [] : _a, className = props.className, style = props.style;
    var _b = react_1.default.useState(), appRef = _b[0], setAppRef = _b[1];
    /** 所有组件父容器 */
    var appContainerRef = react_1.default.useRef(null);
    /** XFlow App 配置中心 */
    var extensionRegistry = (0, extension_registry_1.createExtensionRegistry)();
    /** didMount */
    react_1.default.useEffect(function () {
        /** before app start */
        if (onAppConfigReady) {
            onAppConfigReady(extensionRegistry);
        }
        /** 初始化应用 */
        var app = (0, application_module_1.initApp)(extensionRegistry);
        app.start().then(function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
            var _a;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        /** 保留引用 */
                        setAppRef(app);
                        (_a = extensionRegistry.getExtension('GraphConfig')) === null || _a === void 0 ? void 0 : _a.config.setAppContainer(appContainerRef.current);
                        if (!meta) return [3 /*break*/, 2];
                        return [4 /*yield*/, app.commandService.executeCommand(command_contributions_1.XFlowGraphCommands.LOAD_META.id, { meta: meta })];
                    case 1:
                        _b.sent();
                        _b.label = 2;
                    case 2:
                        /** after app start */
                        if (onLoad) {
                            onLoad(app, extensionRegistry);
                        }
                        if (!graphData) return [3 /*break*/, 6];
                        if (!graphLayout) return [3 /*break*/, 4];
                        return [4 /*yield*/, app.commandService.executeCommand(command_contributions_1.XFlowGraphCommands.GRAPH_LAYOUT.id, tslib_1.__assign({ graphData: graphData }, graphLayout))];
                    case 3:
                        _b.sent();
                        _b.label = 4;
                    case 4: return [4 /*yield*/, app.commandService.executeCommand(command_contributions_1.XFlowGraphCommands.GRAPH_RENDER.id, { graphData: graphData })];
                    case 5:
                        _b.sent();
                        _b.label = 6;
                    case 6: return [2 /*return*/];
                }
            });
        }); });
        /** unmount */
        var destroy = function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                onAppDestroy && onAppDestroy(app);
                return [2 /*return*/];
            });
        }); };
        return function () {
            destroy();
        };
        /** 不要删 保证只生成一次 */
        // eslint-disable-next-line
    }, []);
    /** 自动更新meta */
    react_1.default.useEffect(function () {
        if (appRef) {
            appRef.commandService.executeCommand(command_contributions_1.XFlowGraphCommands.LOAD_META.id, { meta: meta });
        }
        /** 不要删 只和meta联动 */
        // eslint-disable-next-line
    }, [meta]);
    /** 自动渲染画布内容 */
    react_1.default.useEffect(function () {
        var fn = function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
            var x6Graph;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!appRef) return [3 /*break*/, 5];
                        if (!(graphData && graphLayout)) return [3 /*break*/, 2];
                        return [4 /*yield*/, appRef.commandService.executeCommand(command_contributions_1.XFlowGraphCommands.GRAPH_LAYOUT.id, tslib_1.__assign({ graphData: graphData }, graphLayout))];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2: return [4 /*yield*/, appRef.commandService.executeCommand(command_contributions_1.XFlowGraphCommands.GRAPH_RENDER.id, {
                            graphData: graphData,
                        })
                        /** 自动居中画布内容 */
                    ];
                    case 3:
                        _a.sent();
                        if (!isAutoCenter) return [3 /*break*/, 5];
                        return [4 /*yield*/, appRef.getGraphInstance()];
                    case 4:
                        x6Graph = _a.sent();
                        x6Graph.centerContent();
                        _a.label = 5;
                    case 5: return [2 /*return*/];
                }
            });
        }); };
        fn();
        /** 不要删 只和graphData联动 */
        // eslint-disable-next-line
    }, [graphData]);
    /** classname */
    var appClzName = classnames_1.default.apply(void 0, tslib_1.__spreadArray(['xflow-app-workspace',
        className], extensionRegistry.getContainerClassNames(), false));
    /** 判断是否需要自动渲染画布组件, 坐标相对于xflow-graph-root */
    var hasCanvasComponent = (Array.isArray(children) ? children : [children]).some(function (child) {
        return child && child.props.isXFlowCanvas;
    });
    return (react_1.default.createElement(app_context_1.XFlowAppInternalProvider, { app: appRef },
        react_1.default.createElement(extension_context_1.ExtensionRegistryContext.Provider, { value: extensionRegistry },
            react_1.default.createElement("div", { className: appClzName, id: extensionRegistry.getInstanceId(), ref: appContainerRef, style: style },
                !hasCanvasComponent && (react_1.default.createElement(canvas_1.XFlowCanvas, { config: graphConfig, position: { top: 0, bottom: 0, left: 0, right: 0 } })),
                children,
                react_1.default.createElement(model_service_1.ModelServiceRegistry, { config: modelServiceConfig }),
                react_1.default.createElement(command_contributions_1.CommandsRegistry, { config: commandConfig }),
                react_1.default.createElement(hooks_1.HookRegistry, { config: hookConfig }),
                react_1.default.createElement(toolbar_1.ToolbarRegistry, null),
                react_1.default.createElement(menu_1.MenuRegistry, null)))));
};
exports.XFlow = XFlow;
exports.XFlow.defaultProps = {};
exports.default = exports.XFlow;
