// antd-token-previewer/lib/component-demos/radio/button.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _antd = require("antd");

var Demo = function Demo() {
  return /*#__PURE__*/_react.default.createElement(_antd.Space, null, /*#__PURE__*/_react.default.createElement(_antd.Radio.Group, {
    defaultValue: 'a',
    buttonStyle: "solid"
  }, /*#__PURE__*/_react.default.createElement(_antd.Radio.Button, {
    value: 'a',
    checked: true
  }, "Hangzhou"), /*#__PURE__*/_react.default.createElement(_antd.Radio.Button, {
    value: 'b'
  }, "Shanghai")), /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(_antd.Radio.Button, null, "Apple"), /*#__PURE__*/_react.default.createElement(_antd.Radio.Button, {
    checked: true
  }, "Orange")));
};

var componentDemo = {
  demo: /*#__PURE__*/_react.default.createElement(Demo, null),
  tokens: ['colorPrimaryActive', 'colorPrimaryHover'],
  key: 'button'
};
var _default = componentDemo;
exports.default = _default;
