// antd-token-previewer/lib/component-demos/button/disabled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _antd = require("antd");

var Demo = function Demo() {
  return /*#__PURE__*/_react.default.createElement(_antd.Space, null, /*#__PURE__*/_react.default.createElement(_antd.Button, {
    disabled: true,
    type: "primary"
  }, "Primary"), /*#__PURE__*/_react.default.createElement(_antd.Button, {
    disabled: true
  }, "Default"), /*#__PURE__*/_react.default.createElement(_antd.Button, {
    disabled: true,
    type: "dashed"
  }, "Dashed"), /*#__PURE__*/_react.default.createElement("br", null), /*#__PURE__*/_react.default.createElement(_antd.Button, {
    disabled: true,
    type: "text"
  }, "Text"), /*#__PURE__*/_react.default.createElement(_antd.Button, {
    disabled: true,
    ghost: true
  }, "Ghost"), /*#__PURE__*/_react.default.createElement(_antd.Button, {
    disabled: true,
    type: "link"
  }, "Link"));
};

var componentDemo = {
  demo: /*#__PURE__*/_react.default.createElement(Demo, null),
  tokens: ['colorTextDisabled', 'colorBgContainerDisabled'],
  key: 'disabled'
};
var _default = componentDemo;
exports.default = _default;
