// @antv/l7-maps/lib/utils/BaseMapWrapper.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));
var _initializerDefineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/initializerDefineProperty"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _applyDecoratedDescriptor2 = _interopRequireDefault(require("@babel/runtime/helpers/applyDecoratedDescriptor"));
var _initializerWarningHelper2 = _interopRequireDefault(require("@babel/runtime/helpers/initializerWarningHelper"));
var _l7Core = require("@antv/l7-core");
var _dec, _class, _descriptor;
var BaseMapWrapper = (_dec = (0, _l7Core.lazyInject)(_l7Core.TYPES.IGlobalConfigService), (_class = /*#__PURE__*/function () {
  function BaseMapWrapper(config) {
    (0, _classCallCheck2.default)(this, BaseMapWrapper);
    (0, _initializerDefineProperty2.default)(this, "configService", _descriptor, this);
    this.config = config;
  }
  (0, _createClass2.default)(BaseMapWrapper, [{
    key: "setContainer",
    value: function setContainer(sceneContainer, id, canvas, hasBaseMap) {
      // 绑定用户传入的原始地图参数
      sceneContainer.bind(_l7Core.TYPES.MapConfig).toConstantValue((0, _objectSpread2.default)((0, _objectSpread2.default)({}, this.config), {}, {
        id: id,
        canvas: canvas,
        hasBaseMap: hasBaseMap
      }));
      sceneContainer.bind(_l7Core.TYPES.IMapService).to(this.getServiceConstructor()).inSingletonScope();
    }
  }, {
    key: "getServiceConstructor",
    value: function getServiceConstructor() {
      throw new Error('Method not implemented.');
    }
  }]);
  return BaseMapWrapper;
}(), (_descriptor = (0, _applyDecoratedDescriptor2.default)(_class.prototype, "configService", [_dec], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
})), _class));
exports.default = BaseMapWrapper;
