// @antv/xflow-extension/lib/canvas-collapse-panel/service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useCollapsePanelData = exports.executeCollapsePanelCommand = void 0;
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importDefault(require("react"));
var xflow_core_1 = require("@antv/xflow-core");
var interface_1 = require("./interface");
/** 方便其他组件执行Command改变Panel内部状态 */
var executeCollapsePanelCommand = function (cmds, updateModel) {
    cmds.executeCommand(xflow_core_1.XFlowModelCommands.UPDATE_MODEL.id, {
        getModel: function (modelService) { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                // eslint-disable-next-line react-hooks/rules-of-hooks
                return [2 /*return*/, interface_1.NsCollapsePanelModel.useModel(modelService)];
            });
        }); },
        updateModel: updateModel,
    });
};
exports.executeCollapsePanelCommand = executeCollapsePanelCommand;
/** 方便其他组件执行Command改变Panel内部状态 */
var useCollapsePanelData = function (props) {
    var collapsible = props.collapsible, nodeDataService = props.nodeDataService, searchService = props.searchService, onCollapseChange = props.onCollapseChange;
    var modelService = (0, xflow_core_1.useXFlowApp)().modelService;
    /** 创建model */
    var _a = (0, xflow_core_1.createComponentModel)({
        /** service: onKeywordChange */
        keyword: '',
        /** 展开的key id */
        collapseData: [],
        /** 搜索结果 */
        searchResult: [],
    }), state = _a[0], setState = _a[1], panelModel = _a[2];
    /** 注册model成为全局状态，方便其他组件联动 */
    react_1.default.useEffect(function () {
        if (modelService.findDeferredModel(interface_1.NsCollapsePanelModel.id)) {
            return;
        }
        modelService.registerModel({
            id: interface_1.NsCollapsePanelModel.id,
            modelFactory: function () { return panelModel; },
            watchChange: function (self) { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
                var graphMetaModel, fetch, graphMetaDisposable;
                return tslib_1.__generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, xflow_core_1.MODELS.GRAPH_META.getModel(modelService)]; //useContext(MODELS.GRAPH_META.id)
                        case 1:
                            graphMetaModel = _a.sent() //useContext(MODELS.GRAPH_META.id)
                            ;
                            fetch = function (meta) { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
                                var data;
                                return tslib_1.__generator(this, function (_a) {
                                    switch (_a.label) {
                                        case 0: return [4 /*yield*/, nodeDataService(meta, modelService)];
                                        case 1:
                                            data = _a.sent();
                                            return [2 /*return*/, { collapseData: data }];
                                    }
                                });
                            }); };
                            graphMetaDisposable = graphMetaModel.watch(function (meta) { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
                                var collapseData;
                                return tslib_1.__generator(this, function (_a) {
                                    switch (_a.label) {
                                        case 0: return [4 /*yield*/, fetch(meta)];
                                        case 1:
                                            collapseData = (_a.sent()).collapseData;
                                            self.setValue(function (m) {
                                                m.keyword = '';
                                                m.collapseData = collapseData;
                                                m.searchResult = [];
                                            });
                                            return [2 /*return*/];
                                    }
                                });
                            }); });
                            return [2 /*return*/, xflow_core_1.Disposable.create(function () {
                                    graphMetaDisposable.dispose();
                                })];
                    }
                });
            }); },
        });
        /* eslint-disable-next-line  */
    }, []);
    /** 折叠文件夹 */
    var onActiveKeyChange = react_1.default.useCallback(function (activeKey) {
        setState(function (modelState) {
            modelState.collapseData.forEach(function (item) {
                if (item.id === activeKey && item.collapsible !== false) {
                    item.isCollapsed = !item.isCollapsed;
                }
            });
        });
    }, [setState]);
    /** 搜索 */
    var onKeywordChange = react_1.default.useCallback(function (keyword, panelNodes) { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
        var searchResult_1;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!searchService) {
                        return [2 /*return*/, console.warn('searchService is not defined')];
                    }
                    if (!keyword) return [3 /*break*/, 2];
                    return [4 /*yield*/, searchService(panelNodes, keyword)];
                case 1:
                    searchResult_1 = _a.sent();
                    setState(function (modelState) {
                        modelState.keyword = keyword;
                        modelState.searchResult = searchResult_1;
                    });
                    return [3 /*break*/, 3];
                case 2:
                    setState(function (modelState) {
                        modelState.keyword = '';
                        modelState.searchResult = [];
                    });
                    _a.label = 3;
                case 3: return [2 /*return*/];
            }
        });
    }); }, 
    // eslint-disable-next-line
    [searchService]);
    return {
        state: state,
        collapsible: collapsible,
        setState: setState,
        onKeywordChange: onKeywordChange,
        onActiveKeyChange: onActiveKeyChange,
        onCollapseChange: onCollapseChange,
    };
};
exports.useCollapsePanelData = useCollapsePanelData;
