// @antv/l7-utils/lib/anchor.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/anchor.ts
var anchor_exports = {};
__export(anchor_exports, {
  anchorTranslate: () => anchorTranslate,
  anchorType: () => anchorType,
  applyAnchorClass: () => applyAnchorClass
});
module.exports = __toCommonJS(anchor_exports);
var anchorType = /* @__PURE__ */ ((anchorType2) => {
  anchorType2["CENTER"] = "center";
  anchorType2["TOP"] = "top";
  anchorType2["TOP-LEFT"] = "top-left";
  anchorType2["TOP-RIGHT"] = "top-right";
  anchorType2["BOTTOM"] = "bottom";
  anchorType2["BOTTOM-LEFT"] = "bottom-left";
  anchorType2["BOTTOM-RIGHT"] = "bottom-right";
  anchorType2["LEFT"] = "left";
  anchorType2["RIGHT"] = "right";
  return anchorType2;
})(anchorType || {});
var anchorTranslate = {
  center: "translate(-50%,-50%)",
  top: "translate(-50%,0)",
  "top-left": "translate(0,0)",
  "top-right": "translate(-100%,0)",
  bottom: "translate(-50%,-100%)",
  "bottom-left": "translate(0,-100%)",
  "bottom-right": "translate(-100%,-100%)",
  left: "translate(0,-50%)",
  right: "translate(-100%,-50%)"
};
function applyAnchorClass(element, anchor, prefix) {
  const classList = element.classList;
  for (const key in anchorTranslate) {
    if (anchorTranslate.hasOwnProperty(key)) {
      classList.remove(`l7-${prefix}-anchor-${key}`);
    }
  }
  classList.add(`l7-${prefix}-anchor-${anchor}`);
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  anchorTranslate,
  anchorType,
  applyAnchorClass
});
