// @antv/l7-utils/lib/env.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/env.ts
var env_exports = {};
__export(env_exports, {
  getReferrer: () => getReferrer,
  isAndroid: () => isAndroid,
  isImageBitmap: () => isImageBitmap,
  isPC: () => isPC,
  isWorker: () => isWorker,
  isiOS: () => isiOS
});
module.exports = __toCommonJS(env_exports);
function isImageBitmap(image) {
  return typeof ImageBitmap !== "undefined" && image instanceof ImageBitmap;
}
function isWorker() {
  return typeof importScripts === "function";
}
var getReferrer = isWorker() ? () => self.worker && self.worker.referrer : () => (window.location.protocol === "blob:" ? window.parent : window).location.href;
var userAgent = navigator.userAgent;
var isiOS = !!userAgent.match(/\(i[^;]+;( U;)? CPU.+Mac OS X/);
var isAndroid = userAgent.indexOf("Android") > -1 || userAgent.indexOf("Adr") > -1;
function isPC() {
  const userAgentInfo = navigator.userAgent;
  const Agents = [
    "Android",
    "iPhone",
    "SymbianOS",
    "Windows Phone",
    "iPad",
    "iPod"
  ];
  let flag = true;
  for (const v of Agents) {
    if (userAgentInfo.indexOf(v) > 0) {
      flag = false;
      break;
    }
  }
  return flag;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  getReferrer,
  isAndroid,
  isImageBitmap,
  isPC,
  isWorker,
  isiOS
});
