// @antv/xflow-core/lib/command-contributions/graph/graph-save-data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GraphSaveDataCommand = exports.NsGraphSaveData = void 0;
var tslib_1 = require("tslib");
var mana_syringe_1 = require("mana-syringe");
var constant_1 = require("../constant");
var interface_1 = require("../../command/interface");
var NsGraphSaveData;
(function (NsGraphSaveData) {
    /** Command Id: 用于注册named factory */
    NsGraphSaveData.command = constant_1.XFlowGraphCommands.SAVE_GRAPH_DATA;
    /** hookName */
    NsGraphSaveData.hookKey = 'saveGraphData';
})(NsGraphSaveData = exports.NsGraphSaveData || (exports.NsGraphSaveData = {}));
var GraphSaveDataCommand = /** @class */ (function () {
    /** 创建节点命令 */
    function GraphSaveDataCommand() {
        var _this = this;
        /** 执行Cmd */
        this.execute = function () { return tslib_1.__awaiter(_this, void 0, void 0, function () {
            var ctx, args, hooks;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        ctx = this.ctx;
                        args = ctx.getArgs();
                        hooks = ctx.getHooks();
                        /** 执行hooks */
                        return [4 /*yield*/, hooks.saveGraphData.call(
                            /** 执行hooks pipeline处理args */
                            args.args, 
                            /** 执行 callback */
                            function (handlerArgs) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                                var saveGraphDataService, includeAttrs, x6Graph, x6Nodes, x6Edges, nodes, edges, graphData, graphMeta, result;
                                return tslib_1.__generator(this, function (_a) {
                                    switch (_a.label) {
                                        case 0:
                                            saveGraphDataService = handlerArgs.saveGraphDataService, includeAttrs = handlerArgs.includeAttrs;
                                            return [4 /*yield*/, ctx.getX6Graph()];
                                        case 1:
                                            x6Graph = _a.sent();
                                            x6Nodes = x6Graph.getNodes();
                                            x6Edges = x6Graph.getEdges();
                                            nodes = x6Nodes.map(function (node) {
                                                var data = node.getData();
                                                var position = node.position();
                                                var size = node.size();
                                                var model = tslib_1.__assign(tslib_1.__assign(tslib_1.__assign({ id: node.id }, data), position), size);
                                                if (includeAttrs) {
                                                    model.attrs = node.getAttrs();
                                                }
                                                return model;
                                            });
                                            edges = x6Edges.map(function (edge) {
                                                var data = edge.getData();
                                                var model = tslib_1.__assign({ id: edge.id }, data);
                                                if (includeAttrs) {
                                                    model.attrs = edge.getAttrs();
                                                }
                                                return model;
                                            });
                                            graphData = { nodes: nodes, edges: edges };
                                            return [4 /*yield*/, this.ctx.getGraphMeta()
                                                /** 执行 service */
                                            ];
                                        case 2:
                                            graphMeta = _a.sent();
                                            if (!saveGraphDataService) return [3 /*break*/, 4];
                                            return [4 /*yield*/, saveGraphDataService(graphMeta, graphData)
                                                /** 设置结果 */
                                            ];
                                        case 3:
                                            result = _a.sent();
                                            /** 设置结果 */
                                            if (result) {
                                                this.ctx.setResult(result);
                                            }
                                            _a.label = 4;
                                        case 4: return [2 /*return*/, {}];
                                    }
                                });
                            }); }, 
                            /** 外部的 hook */
                            args.hooks)];
                    case 1:
                        /** 执行hooks */
                        _a.sent();
                        return [2 /*return*/, this];
                }
            });
        }); };
        /** undo cmd */
        this.undo = function () { return tslib_1.__awaiter(_this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                this.ctx.undo();
                return [2 /*return*/, this];
            });
        }); };
        /** redo cmd */
        this.redo = function () { return tslib_1.__awaiter(_this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!!this.ctx.isUndoable) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.execute()];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2: return [2 /*return*/, this];
                }
            });
        }); };
    }
    GraphSaveDataCommand.prototype.init = function () {
        this.ctx = this.contextProvider();
    };
    /** isUndoable */
    GraphSaveDataCommand.prototype.isUndoable = function () {
        return this.ctx.isUndoable();
    };
    tslib_1.__decorate([
        (0, mana_syringe_1.inject)(interface_1.ICommandContextProvider),
        tslib_1.__metadata("design:type", Object)
    ], GraphSaveDataCommand.prototype, "contextProvider", void 0);
    tslib_1.__decorate([
        (0, mana_syringe_1.postConstruct)(),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], GraphSaveDataCommand.prototype, "init", null);
    GraphSaveDataCommand = tslib_1.__decorate([
        (0, mana_syringe_1.injectable)({
            token: { token: interface_1.ICommandHandler, named: NsGraphSaveData.command.id },
        })
        /** 创建节点命令 */
    ], GraphSaveDataCommand);
    return GraphSaveDataCommand;
}());
exports.GraphSaveDataCommand = GraphSaveDataCommand;
