// @antv/xflow-core/lib/command-contributions/node/node-update.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpdateNodeCommand = exports.NsUpdateNode = void 0;
var tslib_1 = require("tslib");
var mana_syringe_1 = require("mana-syringe");
var x6_1 = require("@antv/x6");
var isBoolean_1 = tslib_1.__importDefault(require("lodash/isBoolean"));
var constant_1 = require("../constant");
var interface_1 = require("../../command/interface");
var x6_react_shape_1 = require("@antv/x6-react-shape");
var NsUpdateNode;
(function (NsUpdateNode) {
    NsUpdateNode.command = constant_1.XFlowNodeCommands.UPDATE_NODE;
    NsUpdateNode.hookKey = 'updateNode';
    NsUpdateNode.XFlowNodeSetOptions = { overwrite: true };
    NsUpdateNode.NODE_WIDTH = 200;
    NsUpdateNode.NODE_HEIGHT = 40;
})(NsUpdateNode = exports.NsUpdateNode || (exports.NsUpdateNode = {}));
var UpdateNodeCommand = /** @class */ (function () {
    /** 节点更新命令 */
    function UpdateNodeCommand() {
        var _this = this;
        this.setNodeConfig = function (x6Node, nodeConfig, options) {
            x6Node.setData(nodeConfig, options);
            x6Node.setPosition((nodeConfig === null || nodeConfig === void 0 ? void 0 : nodeConfig.x) || 0, (nodeConfig === null || nodeConfig === void 0 ? void 0 : nodeConfig.y) || 0);
            x6Node.setSize((nodeConfig === null || nodeConfig === void 0 ? void 0 : nodeConfig.width) || NsUpdateNode.NODE_WIDTH, (nodeConfig === null || nodeConfig === void 0 ? void 0 : nodeConfig.height) || NsUpdateNode.NODE_HEIGHT);
            x6Node.angle((nodeConfig === null || nodeConfig === void 0 ? void 0 : nodeConfig.angle) || 0, {
                absolute: true,
            });
            if ((0, isBoolean_1.default)(nodeConfig === null || nodeConfig === void 0 ? void 0 : nodeConfig.visible)) {
                x6Node.setVisible(nodeConfig === null || nodeConfig === void 0 ? void 0 : nodeConfig.visible);
            }
            // SVG 元素更新label
            if (!(x6Node instanceof x6_react_shape_1.ReactShape) && !!x6Node.getAttrByPath('text/text')) {
                x6Node.setAttrByPath('text/text', nodeConfig.label);
            }
            // 支持nodeAttrs
            if (nodeConfig.attrs) {
                x6Node.setAttrs(nodeConfig.attrs);
            }
            // 更新ports
            if (Array.isArray(nodeConfig.ports)) {
                x6Node.setPropByPath('ports/items', nodeConfig.ports, tslib_1.__assign({ rewrite: true }, options));
            }
        };
        this.getNodeConfig = function (x6Node) {
            var data = x6Node.getData();
            var position = x6Node.getPosition();
            var size = x6Node.getSize();
            return tslib_1.__assign(tslib_1.__assign(tslib_1.__assign({}, data), position), size);
        };
        this.getNodeCell = function (x6Graph, handlerArgs) {
            var nodeConfig = handlerArgs.nodeConfig, setNodeConfig = handlerArgs.setNodeConfig;
            var nodeId = '';
            if (setNodeConfig && setNodeConfig.node && typeof setNodeConfig.node === 'string') {
                nodeId = setNodeConfig.node;
            }
            else if (nodeConfig && nodeConfig.id && typeof nodeConfig.id === 'string') {
                nodeId = nodeConfig.id;
            }
            if (nodeId) {
                return x6Graph === null || x6Graph === void 0 ? void 0 : x6Graph.getCellById(nodeId);
            }
            if (setNodeConfig && setNodeConfig.node instanceof x6_1.Node) {
                return setNodeConfig.node;
            }
        };
        this.getNextNodeConfig = function (handlerArgs, x6Node) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
            var nodeData;
            return tslib_1.__generator(this, function (_a) {
                if (handlerArgs && handlerArgs.setNodeConfig && handlerArgs.setNodeConfig.callback) {
                    nodeData = this.getNodeConfig(x6Node);
                    return [2 /*return*/, handlerArgs.setNodeConfig.callback(nodeData)];
                }
                return [2 /*return*/, handlerArgs.nodeConfig];
            });
        }); };
        this.execute = function () { return tslib_1.__awaiter(_this, void 0, void 0, function () {
            var _a, args, runtimeHook, hooks, result;
            var _this = this;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = this.ctx.getArgs(), args = _a.args, runtimeHook = _a.hooks;
                        hooks = this.ctx.getHooks();
                        return [4 /*yield*/, hooks.updateNode.call(args, function (handlerArgs) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                                var _a, options, graph, x6Node, nextNodeConfig;
                                return tslib_1.__generator(this, function (_b) {
                                    switch (_b.label) {
                                        case 0:
                                            _a = handlerArgs.options, options = _a === void 0 ? NsUpdateNode.XFlowNodeSetOptions : _a;
                                            return [4 /*yield*/, this.ctx.getX6Graph()];
                                        case 1:
                                            graph = _b.sent();
                                            x6Node = this.getNodeCell(graph, handlerArgs);
                                            return [4 /*yield*/, this.getNextNodeConfig(handlerArgs, x6Node)];
                                        case 2:
                                            nextNodeConfig = _b.sent();
                                            this.setNodeConfig(x6Node, nextNodeConfig, options);
                                            return [2 /*return*/, {
                                                    nodeConfig: nextNodeConfig,
                                                    nodeCell: x6Node,
                                                }];
                                    }
                                });
                            }); }, runtimeHook)];
                    case 1:
                        result = _b.sent();
                        this.ctx.setResult(result);
                        return [2 /*return*/, this];
                }
            });
        }); };
        this.undo = function () { return tslib_1.__awaiter(_this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                this.ctx.undo();
                return [2 /*return*/, this];
            });
        }); };
        this.redo = function () { return tslib_1.__awaiter(_this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!!this.ctx.isUndoable) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.execute()];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2: return [2 /*return*/, this];
                }
            });
        }); };
    }
    UpdateNodeCommand.prototype.init = function () {
        this.ctx = this.contextProvider();
    };
    UpdateNodeCommand.prototype.isUndoable = function () {
        return this.ctx.isUndoable();
    };
    tslib_1.__decorate([
        (0, mana_syringe_1.inject)(interface_1.ICommandContextProvider),
        tslib_1.__metadata("design:type", Object)
    ], UpdateNodeCommand.prototype, "contextProvider", void 0);
    tslib_1.__decorate([
        (0, mana_syringe_1.postConstruct)(),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], UpdateNodeCommand.prototype, "init", null);
    UpdateNodeCommand = tslib_1.__decorate([
        (0, mana_syringe_1.injectable)({
            token: { token: interface_1.ICommandHandler, named: NsUpdateNode.command.id },
        })
        /** 节点更新命令 */
    ], UpdateNodeCommand);
    return UpdateNodeCommand;
}());
exports.UpdateNodeCommand = UpdateNodeCommand;
