// @antv/l7-utils/lib/workers/index.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/workers/index.ts
var workers_exports = {};
__export(workers_exports, {
  WorkerSourceMap: () => WorkerSourceMap,
  createWorker: () => import_web_worker_helper.createWorker,
  worker: () => worker
});
module.exports = __toCommonJS(workers_exports);
var import_web_worker_helper = require("web-worker-helper");
var import_lineModel = require("./lineModel");
var import_pointFillModel = require("./pointFillModel");
var import_polygonFillModel = require("./polygonFillModel");
var WorkerSourceMap = {
  pointFill: import_pointFillModel.pointFillModel,
  line: import_lineModel.lineModel,
  polygonFill: import_polygonFillModel.polygonFillModel
};
var workerTypes = {
  ...WorkerSourceMap
};
async function worker({ workerType, data }) {
  if (workerTypes[workerType]) {
    return workerTypes[workerType](data);
  }
  return Promise.reject(
    new Error(`Worker with type "${workerType}" non-existent.`)
  );
}
(0, import_web_worker_helper.createWorker)(worker);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  WorkerSourceMap,
  createWorker,
  worker
});
