// antd-token-previewer/es/ThemeSelect.js
import _toConsumableArray from "@babel/runtime/helpers/esm/toConsumableArray";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import React, { useMemo } from 'react';
import makeStyle from "./utils/makeStyle";
import classNames from 'classnames';
import { Button, Dropdown, Menu } from 'antd';
import { CloseOutlined, PlusOutlined } from '@ant-design/icons';
var useStyle = makeStyle('ThemeSelect', function (token) {
  var _previewerThemeSel;

  return {
    '.previewer-theme-select': (_previewerThemeSel = {
      padding: "".concat(token.paddingXXS, "px ").concat(token.paddingXS, "px"),
      borderRadius: 4,
      backgroundColor: 'rgba(0, 0, 0, 0.02)',
      height: token.controlHeight,
      display: 'flex',
      alignItems: 'center',
      overflow: 'hidden'
    }, _defineProperty(_previewerThemeSel, "".concat(token.rootCls, "-btn.previewer-theme-select-add-btn"), {
      minWidth: 0,
      width: 16,
      height: 16,
      fontSize: 8,
      display: 'inline-flex',
      alignItems: 'center',
      justifyContent: 'center',
      marginInlineStart: token.marginSM,
      boxShadow: 'none'
    }), _defineProperty(_previewerThemeSel, '.previewer-theme-select-tag', {
      height: 22,
      display: 'inline-flex',
      alignItems: 'center',
      justifyContent: 'center',
      boxSizing: 'border-box',
      borderRadius: 4,
      backgroundColor: token.colorBgContainer,
      border: "".concat(token.lineWidth, "px ").concat(token.lineType, " ").concat(token.colorBorder),
      paddingInline: 10,
      fontSize: token.fontSizeSM,
      position: 'relative',
      cursor: 'pointer',
      // transition: `all ${token.motionDurationMid}`,
      '&:not(:last-child)': {
        marginInlineEnd: token.marginXS
      },
      '&.previewer-theme-select-tag-active': {
        border: "".concat(token.lineWidth, "px ").concat(token.lineType, " ").concat(token['blue-1']),
        backgroundColor: 'rgba(22,119,255,0.10)',
        color: token.colorPrimary,
        '&::after': {
          content: '""',
          borderStartEndRadius: 2,
          position: 'absolute',
          insetInlineEnd: 2,
          top: 2,
          width: 6,
          height: 6,
          background: "linear-gradient(to right top, transparent, transparent 50%, ".concat(token.colorPrimary, " 50%, ").concat(token.colorPrimary, " 100%)")
        }
      },
      '.previewer-theme-select-tag-close-btn': {
        position: 'absolute',
        top: -2,
        insetInlineEnd: -2,
        width: 12,
        height: 12,
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center',
        background: token.colorBgContainer,
        boxShadow: '0 2px 8px -2px rgba(0,0,0,0.05), 0 1px 4px -1px rgba(25,15,15,0.07), 0 0 1px 0 rgba(0,0,0,0.08)',
        borderRadius: '50%',
        opacity: 0,
        pointerEvents: 'none',
        zIndex: 2,
        color: token.colorIcon,
        '> .anticon': {
          fontSize: 6
        }
      },
      '&:hover': {
        '.previewer-theme-select-tag-close-btn': {
          opacity: 1,
          pointerEvents: 'initial'
        }
      }
    }), _previewerThemeSel),
    '.previewer-theme-select-dropdown': {
      '.previewer-theme-select-dropdown-title': _defineProperty({}, "".concat(token.rootCls, "-dropdown-menu-item-group-title"), {
        fontSize: token.fontSizeSM,
        paddingBottom: token.padding,
        paddingTop: 10
      })
    }
  };
});

var ThemeSelect = function ThemeSelect(props) {
  var onEnabledThemeChange = props.onEnabledThemeChange,
      onShownThemeChange = props.onShownThemeChange,
      enabledThemes = props.enabledThemes,
      shownThemes = props.shownThemes,
      themes = props.themes,
      showAddTheme = props.showAddTheme;

  var _useStyle = useStyle(),
      _useStyle2 = _slicedToArray(_useStyle, 2),
      wrapSSR = _useStyle2[0],
      hashId = _useStyle2[1];

  var dropdownItems = useMemo(function () {
    return [{
      disabled: true,
      label: '添加主题即可预览',
      className: 'previewer-theme-select-dropdown-title',
      type: 'group',
      key: 'add-theme-to-preview'
    }].concat(_toConsumableArray(themes.filter(function (theme) {
      return !shownThemes.includes(theme.key);
    }).map(function (theme) {
      return {
        icon: theme.icon,
        value: theme.key,
        label: theme.name,
        key: theme.key,
        onClick: function onClick() {
          onShownThemeChange([].concat(_toConsumableArray(shownThemes), [theme.key]), theme.key, {
            type: 'select'
          });
        }
      };
    })));
  }, [themes, shownThemes, onShownThemeChange]);
  var shownThemeEntities = useMemo(function () {
    return themes.filter(function (theme) {
      return shownThemes.includes(theme.key);
    });
  }, [themes, shownThemes]);
  return wrapSSR( /*#__PURE__*/React.createElement("div", {
    className: classNames('previewer-theme-select', hashId)
  }, shownThemeEntities.map(function (theme) {
    return /*#__PURE__*/React.createElement("span", {
      onClick: function onClick() {
        if (theme.fixed) {
          return;
        }

        onEnabledThemeChange(enabledThemes.includes(theme.key) ? enabledThemes.filter(function (item) {
          return item !== theme.key;
        }) : [].concat(_toConsumableArray(enabledThemes), [theme.key]));
      },
      key: theme.key,
      className: classNames('previewer-theme-select-tag', {
        'previewer-theme-select-tag-active': enabledThemes.includes(theme.key)
      })
    }, theme.name, theme.closable && /*#__PURE__*/React.createElement("span", {
      className: "previewer-theme-select-tag-close-btn",
      onClick: function onClick(e) {
        e.stopPropagation();
        onEnabledThemeChange(enabledThemes.filter(function (item) {
          return item !== theme.key;
        }));
        onShownThemeChange(shownThemes.filter(function (item) {
          return item !== theme.key;
        }), theme.key, {
          type: 'deselect'
        });
      }
    }, /*#__PURE__*/React.createElement(CloseOutlined, null)));
  }), showAddTheme && /*#__PURE__*/React.createElement(Dropdown, {
    placement: "bottomRight",
    trigger: ['click'],
    overlay: /*#__PURE__*/React.createElement(Menu, {
      items: dropdownItems
    }),
    overlayClassName: classNames('previewer-theme-select-dropdown', hashId)
  }, /*#__PURE__*/React.createElement(Button, {
    type: "primary",
    shape: "circle",
    className: "previewer-theme-select-add-btn",
    icon: /*#__PURE__*/React.createElement(PlusOutlined, null)
  }))));
};

export default ThemeSelect;
