// antd-token-previewer/es/component-demos/grid/grid.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import React from 'react';
import { Row, Col } from 'antd';
import makeStyle from "../../utils/makeStyle";
import classNames from 'classnames';
var useStyle = makeStyle('GridDemo', function (token) {
  var _previewerGridDemo;

  return {
    '.previewer-grid-demo': (_previewerGridDemo = {}, _defineProperty(_previewerGridDemo, "".concat(token.rootCls, "-row"), {
      marginBottom: 16
    }), _defineProperty(_previewerGridDemo, "".concat(token.rootCls, "-row > div:not(.gutter-row)"), {
      padding: '16px 0',
      background: '#0092ff',
      color: '#fff',
      display: 'inline-flex',
      alignItems: 'center',
      justifyContent: 'center',
      '&:nth-child(2n + 1)': {
        background: 'rgba(0,146,255,.75)'
      }
    }), _previewerGridDemo)
  };
});

var Demo = function Demo() {
  var _useStyle = useStyle(),
      _useStyle2 = _slicedToArray(_useStyle, 2),
      hashId = _useStyle2[1];

  return /*#__PURE__*/React.createElement("div", {
    className: classNames('previewer-grid-demo', hashId)
  }, /*#__PURE__*/React.createElement(Row, null, /*#__PURE__*/React.createElement(Col, {
    span: 24
  }, "col")), /*#__PURE__*/React.createElement(Row, null, /*#__PURE__*/React.createElement(Col, {
    span: 12
  }, "col-12"), " ", /*#__PURE__*/React.createElement(Col, {
    span: 12
  }, "col-12")), /*#__PURE__*/React.createElement(Row, null, /*#__PURE__*/React.createElement(Col, {
    span: 8
  }, "col-8"), " ", /*#__PURE__*/React.createElement(Col, {
    span: 8
  }, "col-8"), /*#__PURE__*/React.createElement(Col, {
    span: 8
  }, "col-8")), /*#__PURE__*/React.createElement(Row, null, /*#__PURE__*/React.createElement(Col, {
    span: 6
  }, "col-6"), " ", /*#__PURE__*/React.createElement(Col, {
    span: 6
  }, "col-6"), /*#__PURE__*/React.createElement(Col, {
    span: 6
  }, "col-6"), /*#__PURE__*/React.createElement(Col, {
    span: 6
  }, "col-6")));
};

var componentDemo = {
  demo: /*#__PURE__*/React.createElement(Demo, null),
  key: 'default'
};
export default componentDemo;
