// @ant-design/graphs/lib/obsolescent/graphs/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultEdgeCfg = exports.defaultNodeCfg = exports.defaultMinimapCfg = exports.defaultLabelStyle = exports.cardTitlePadding = exports.defaultTitleLabelCfg = exports.Margin = exports.defaultNodeAnchorPoints = exports.defaultNodeStyle = exports.defaultNodeSize = exports.defaultStateStyles = void 0;
// 默认交互状态
exports.defaultStateStyles = {
    hover: {
        stroke: '#1890ff',
        lineWidth: 2,
    },
};
// card 默认节点大小
exports.defaultNodeSize = [120, 40];
// 默认节点样式
exports.defaultNodeStyle = {
    stroke: '#40a9ff',
};
// 默认 anchor 连接点
exports.defaultNodeAnchorPoints = [
    [0, 0.5],
    [1, 0.5],
];
// card 内部 padding | margin
exports.Margin = 6;
// card title 默认样式
exports.defaultTitleLabelCfg = {
    fill: '#fff',
    fontSize: 12,
};
exports.cardTitlePadding = 2;
// card body|footer  默认样式
exports.defaultLabelStyle = {
    fill: '#000',
    fontSize: 12,
};
exports.defaultMinimapCfg = {
    show: false,
    size: [150, 100],
    type: 'keyShape',
};
exports.defaultNodeCfg = {
    style: {
        fill: '#000',
        fontSize: 12,
    },
};
exports.defaultEdgeCfg = {
    style: {
        stroke: '#ccc',
    },
};
